// nativescript-ui-chart/ui-chart.android.js
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var commonModule = require("./ui-chart.common");
var publicEnumModule = require("./misc/chart-public-enum");
var trackBallCommonModule = require("./visualization/parts/chart-track-ball.common");
var view_1 = require("tns-core-modules/ui/core/view");
__export(require("./ui-chart.common"));
__export(require("./misc/chart-public-enum"));
__export(require("./visualization/views/chart-legend-view"));
__export(require("./visualization/views/chart-axis"));
__export(require("./misc/chart-event-data.common"));
__export(require("./visualization/parts/chart-label-style.common"));
__export(require("./visualization/views/chart-series"));
__export(require("./visualization/parts/chart-palette.common"));
__export(require("./visualization/parts/chart-annotation"));
__export(require("./visualization/parts/chart-cartesian-chart-grid.common"));
var chart_cartesian_chart_grid_1 = require("./visualization/parts/chart-cartesian-chart-grid");
exports.RadCartesianChartGrid = chart_cartesian_chart_grid_1.RadCartesianChartGrid;
__export(require("./visualization/parts/chart-track-ball"));
var applicationModule = require("tns-core-modules/application");
var enums_1 = require("tns-core-modules/ui/enums");
function getSdkApiLevel() {
    return android.os.Build.VERSION.SDK_INT;
}
var ChartSelectionChangeListener;
var ChartPanZoomListener;
function initializeListeners() {
    if (!ChartSelectionChangeListener) {
        var ChartSelectionChangeListenerImpl = /** @class */ (function (_super) {
            __extends(ChartSelectionChangeListenerImpl, _super);
            function ChartSelectionChangeListenerImpl(owner) {
                var _this = _super.call(this) || this;
                _this.owner = owner;
                return global.__native(_this);
            }
            ChartSelectionChangeListenerImpl.prototype.onSelectionChanged = function (selectionContext) {
                var isCartesianChart = this.owner instanceof RadCartesianChart;
                if (selectionContext.selectedSeries()) {
                    var args = {
                        eventName: commonModule.RadChartBase.seriesSelectedEvent,
                        object: this.owner,
                        series: isCartesianChart ? this.owner.series.getItem(selectionContext.selectedSeries().getCollectionIndex()) : this.owner.series,
                        pointIndex: null,
                        pointData: null
                    };
                    this.owner.notify(args);
                }
                if (selectionContext.deselectedSeries()) {
                    var args = {
                        eventName: commonModule.RadChartBase.seriesDeselectedEvent,
                        object: this.owner,
                        series: isCartesianChart ? this.owner.series.getItem(selectionContext.deselectedSeries().getCollectionIndex()) : this.owner.series,
                        pointIndex: null,
                        pointData: null
                    };
                    this.owner.notify(args);
                }
                if (selectionContext.selectedDataPoint()) {
                    var args = {
                        eventName: commonModule.RadChartBase.pointSelectedEvent,
                        object: this.owner,
                        series: isCartesianChart ? this.owner.series.getItem(selectionContext.selectedDataPoint().getPresenter().getCollectionIndex()) : this.owner.series,
                        pointIndex: selectionContext.selectedDataPoint().index(),
                        pointData: selectionContext.selectedDataPoint()
                    };
                    this.owner.notify(args);
                }
                if (selectionContext.deselectedDataPoint()) {
                    var args = {
                        eventName: commonModule.RadChartBase.pointDeselectedEvent,
                        object: this.owner,
                        series: isCartesianChart ? this.owner.series.getItem(selectionContext.deselectedDataPoint().getPresenter().getCollectionIndex()) : this.owner.series,
                        pointIndex: selectionContext.deselectedDataPoint().index(),
                        pointData: selectionContext.deselectedDataPoint()
                    };
                    this.owner.notify(args);
                }
            };
            ChartSelectionChangeListenerImpl = __decorate([
                Interfaces([com.telerik.widget.chart.visualization.behaviors.ChartSelectionChangeListener]),
                __metadata("design:paramtypes", [commonModule.RadChartBase])
            ], ChartSelectionChangeListenerImpl);
            return ChartSelectionChangeListenerImpl;
        }(java.lang.Object));
        ChartSelectionChangeListener = ChartSelectionChangeListenerImpl;
    }
    if (!ChartPanZoomListener) {
        var ChartPanZoomListenerImpl = /** @class */ (function (_super) {
            __extends(ChartPanZoomListenerImpl, _super);
            function ChartPanZoomListenerImpl(owner) {
                var _this = _super.call(this) || this;
                _this.owner = owner;
                return global.__native(_this);
            }
            ChartPanZoomListenerImpl.prototype.onPan = function (panX, panY) {
                var args = {
                    eventName: commonModule.RadChartBase.chartPannedEvent,
                    object: this.owner,
                    pointData: null,
                    pointIndex: null,
                    series: null
                };
                this.owner.notify(args);
            };
            ChartPanZoomListenerImpl.prototype.onZoom = function (zoomX, zoomY) {
                var args = {
                    eventName: commonModule.RadChartBase.chartZoomedEvent,
                    object: this.owner,
                    pointData: null,
                    pointIndex: null,
                    series: null
                };
                this.owner.notify(args);
            };
            ChartPanZoomListenerImpl = __decorate([
                Interfaces([com.telerik.widget.chart.visualization.behaviors.PanZoomListener]),
                __metadata("design:paramtypes", [commonModule.RadChartBase])
            ], ChartPanZoomListenerImpl);
            return ChartPanZoomListenerImpl;
        }(java.lang.Object));
        ChartPanZoomListener = ChartPanZoomListenerImpl;
    }
}
var RadPieChart = /** @class */ (function (_super) {
    __extends(RadPieChart, _super);
    function RadPieChart() {
        return _super.call(this) || this;
    }
    Object.defineProperty(RadPieChart.prototype, "androidView", {
        get: function () {
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadPieChart.prototype, "rootLayout", {
        get: function () {
            return this._rootLayout;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadPieChart.prototype, "_nativeView", {
        get: function () {
            return this._rootLayout;
        },
        enumerable: true,
        configurable: true
    });
    RadPieChart.prototype.createNativeView = function () {
        var _this = this;
        initializeListeners();
        this._android = new com.telerik.widget.chart.visualization.pieChart.RadPieChartView(this._context, applicationModule.systemAppearance());
        this._rootLayout = new android.widget.RelativeLayout(this._context);
        this._rootLayout.addView(this._android);
        if (this.legend) {
            this.legend.updateLegendView(this);
        }
        this.initializer.loadSeries(this);
        this.initSelectionBehavior();
        this.pieChartSystemAppearanceChangedHandler = function (args) {
            if (!_this._android) {
                return;
            }
            _this.updatePaletteStyle(args.newValue);
            if (_this.legend) {
                _this.legend.updateLegendTitle();
                _this.legend.updateListViewForTheme();
            }
        };
        applicationModule.on(applicationModule.systemAppearanceChangedEvent, this.pieChartSystemAppearanceChangedHandler);
        return this._rootLayout;
    };
    RadPieChart.prototype.initNativeView = function () {
        if (getSdkApiLevel() && getSdkApiLevel() >= 29) {
            this._android.setForceDarkAllowed(false);
            this.nativeViewProtected.setForceDarkAllowed(false);
        }
        if (applicationModule.systemAppearance() === enums_1.SystemAppearance.dark && this.legend) {
            this.legend.updateLegendTitle();
        }
    };
    RadPieChart.prototype.disposeNativeView = function () {
        applicationModule.off(applicationModule.systemAppearanceChangedEvent, this.pieChartSystemAppearanceChangedHandler);
        this.detachSeries();
        if (this._selectionBehavior) {
            this._android.getBehaviors().remove(this._selectionBehavior);
            this._selectionBehavior = null;
        }
        if (this._android._selectionChangeListener) {
            this._android._selectionChangeListener.owner = null;
        }
    };
    RadPieChart.prototype._onDetach = function (force) {
        this.detachSeries();
        _super.prototype['_onDetach'].call(this, force);
    };
    RadPieChart.prototype.updatePaletteStyle = function (appearance) {
        this._android.setPalette(this.getPaletteForSystemAppearance(appearance));
        this._android.setSelectionPalette(this.getSelectedPaletteForSystemAppearance(appearance));
        if (this.palettes) {
            this.initializer.reloadPalettes(this);
        }
    };
    RadPieChart.prototype.getPaletteForSystemAppearance = function (appearance) {
        if (appearance === enums_1.SystemAppearance.dark) {
            return com.telerik.widget.palettes.ChartPalettes.dark(this._android.getContext());
        }
        return com.telerik.widget.palettes.ChartPalettes.light(this._android.getContext());
    };
    RadPieChart.prototype.getSelectedPaletteForSystemAppearance = function (appearance) {
        if (appearance === enums_1.SystemAppearance.dark) {
            return com.telerik.widget.palettes.ChartPalettes.darkSelected(this._android.getContext());
        }
        return com.telerik.widget.palettes.ChartPalettes.lightSelected(this._android.getContext());
    };
    RadPieChart.prototype.detachSeries = function () {
        var thisAndroid = this._android;
        if (thisAndroid) {
            this._rootLayout = undefined;
            if (this.series) {
                for (var i = 0; i < this.series.length; i++) {
                    if (thisAndroid.getSeries().indexOf(this.series.getItem(i).android) !== -1) {
                        thisAndroid.getSeries().remove(this.series.getItem(i).android);
                    }
                }
            }
        }
    };
    RadPieChart.prototype.initSelectionBehavior = function () {
        if (!this._selectionBehavior) {
            this._selectionBehavior = new com.telerik.widget.chart.visualization.behaviors.ChartSelectionBehavior();
            this._selectionBehavior.setDataPointsSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
            this._selectionBehavior.setSeriesSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
            this._android._selectionChangeListener = new ChartSelectionChangeListener(this);
            this._selectionBehavior.setSelectionChangeListener(this._android._selectionChangeListener);
            this._android.getBehaviors().add(this._selectionBehavior);
        }
    };
    RadPieChart.prototype.onLegendChanged = function (oldValue, newValue) {
        if (this.legend) {
            this.legend.updateLegendView(this);
        }
        this.initializer.onLegendChanged(oldValue, newValue, this);
    };
    return RadPieChart;
}(commonModule.RadPieChart));
exports.RadPieChart = RadPieChart;
var RadCartesianChart = /** @class */ (function (_super) {
    __extends(RadCartesianChart, _super);
    function RadCartesianChart() {
        return _super.call(this) || this;
    }
    Object.defineProperty(RadCartesianChart.prototype, "android", {
        get: function () {
            return this._rootLayout;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadCartesianChart.prototype, "androidView", {
        get: function () {
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadCartesianChart.prototype, "rootLayout", {
        get: function () {
            return this._rootLayout;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadCartesianChart.prototype, "_nativeView", {
        get: function () {
            return this._rootLayout;
        },
        enumerable: true,
        configurable: true
    });
    RadCartesianChart.prototype.createNativeView = function () {
        var _this = this;
        initializeListeners();
        this._android = new com.telerik.widget.chart.visualization.cartesianChart.RadCartesianChartView(this._context, applicationModule.systemAppearance());
        if (this._selectionBehavior) {
            this._android.getBehaviors().add(this._selectionBehavior);
        }
        else {
            this.initSelectionBehavior();
        }
        this._android._selectionChangeListener = new ChartSelectionChangeListener(this);
        this._selectionBehavior.setSelectionChangeListener(this._android._selectionChangeListener);
        this._rootLayout = new android.widget.RelativeLayout(this._context);
        this._rootLayout.addView(this._android);
        if (this.grid) {
            this._android.setGrid(null);
            this._android.setGrid(this.grid._android);
        }
        this._ngKey = Object.keys(this).find(function (key) { return key.startsWith('_ngcontent'); });
        this._vueKey = Object.keys(this).find(function (key) { return key.startsWith('data-v'); });
        if (this.horizontalAxis) {
            this._android.setHorizontalAxis(null);
            this._android.setHorizontalAxis(this.horizontalAxis.android);
            this.horizontalAxis.owner = this;
        }
        if (this.verticalAxis) {
            this._android.setVerticalAxis(null);
            this._android.setVerticalAxis(this.verticalAxis.android);
            this.verticalAxis.owner = this;
        }
        if (this.series) {
            this.initializer.loadSeries(this);
        }
        if (this.legend) {
            this.legend.updateLegendView(this);
        }
        if (this.annotations) {
            for (var i = 0; i < this.annotations.length; i++) {
                this.initNativeAnnotation(i);
                this.annotations.getItem(i)._onOwnerUICreated();
            }
            this.initializer.loadAnnotations(this);
        }
        this.updatePanZoomBehavior();
        this._syncZoomValues();
        this.setNativeTrackballBehavior(this.trackball);
        this.cartesianChartSystemAppearanceChangedHandler = function (args) {
            if (!_this._android) {
                return;
            }
            _this.updatePaletteStyle(args.newValue);
            if (_this.legend) {
                _this.legend.updateLegendTitle();
                _this.legend.updateListViewForTheme();
            }
        };
        applicationModule.on(applicationModule.systemAppearanceChangedEvent, this.cartesianChartSystemAppearanceChangedHandler);
        return this._rootLayout;
    };
    RadCartesianChart.prototype.initNativeView = function () {
        if (getSdkApiLevel() && getSdkApiLevel() >= 29) {
            this._android.setForceDarkAllowed(false);
            this.nativeViewProtected.setForceDarkAllowed(false);
        }
        if (applicationModule.systemAppearance() === enums_1.SystemAppearance.dark && this.legend) {
            this.legend.updateLegendTitle();
        }
    };
    RadCartesianChart.prototype.disposeNativeView = function () {
        applicationModule.off(applicationModule.systemAppearanceChangedEvent, this.cartesianChartSystemAppearanceChangedHandler);
        this.detachSeries();
        if (this._panZoomBehavior) {
            this._android.getBehaviors().remove(this._panZoomBehavior);
            this._panZoomBehavior = undefined;
        }
        if (this._selectionBehavior) {
            this._android.getBehaviors().remove(this._selectionBehavior);
            this._selectionBehavior = null;
        }
        if (this._android._panZoomListener) {
            this._android._panZoomListener.owner = null;
        }
        if (this._android._selectionChangeListener) {
            this._android._selectionChangeListener.owner = null;
        }
        _super.prototype.disposeNativeView.call(this);
    };
    RadCartesianChart.prototype[view_1.paddingLeftProperty.getDefault] = function () {
        return this._android.getChartPaddingLeft();
    };
    RadCartesianChart.prototype[view_1.paddingLeftProperty.setNative] = function (value) {
        var paddingTop = this._android.getChartPaddingTop();
        var paddingRight = this._android.getChartPaddingRight();
        var paddingBottom = this._android.getChartPaddingBottom();
        this._android.setChartPadding(0, paddingTop, paddingRight, paddingBottom);
    };
    RadCartesianChart.prototype[view_1.paddingTopProperty.getDefault] = function () {
        return this._android.getChartPaddingTop();
    };
    RadCartesianChart.prototype[view_1.paddingTopProperty.setNative] = function (value) {
        var paddingLeft = this._android.getChartPaddingLeft();
        var paddingRight = this._android.getChartPaddingRight();
        var paddingBottom = this._android.getChartPaddingBottom();
        this._android.setChartPadding(paddingLeft, 0, paddingRight, paddingBottom);
    };
    RadCartesianChart.prototype[view_1.paddingRightProperty.getDefault] = function () {
        return this._android.getChartPaddingRight();
    };
    RadCartesianChart.prototype[view_1.paddingRightProperty.setNative] = function (value) {
        var paddingLeft = this._android.getChartPaddingLeft();
        var paddingTop = this._android.getChartPaddingTop();
        var paddingBottom = this._android.getChartPaddingBottom();
        this._android.setChartPadding(paddingLeft, paddingTop, 0, paddingBottom);
    };
    RadCartesianChart.prototype[view_1.paddingBottomProperty.getDefault] = function () {
        return this._android.getChartPaddingBottom();
    };
    RadCartesianChart.prototype[view_1.paddingBottomProperty.setNative] = function (value) {
        var paddingLeft = this._android.getChartPaddingLeft();
        var paddingTop = this._android.getChartPaddingTop();
        var paddingRight = this._android.getChartPaddingRight();
        this._android.setChartPadding(paddingLeft, paddingTop, paddingRight, 0);
    };
    RadCartesianChart.prototype.onLegendChanged = function (oldValue, newValue) {
        if (this.legend) {
            this.legend.updateLegendView(this);
        }
        this.initializer.onLegendChanged(oldValue, newValue, this);
    };
    RadCartesianChart.prototype.getAxisByID = function (axisID) {
        if (this.horizontalAxis && this.horizontalAxis.id === axisID) {
            return this.horizontalAxis;
        }
        if (this.verticalAxis && this.verticalAxis.id === axisID) {
            return this.verticalAxis;
        }
        if (this.series) {
            var axis = void 0;
            for (var i = 0; i < this.series.length; ++i) {
                axis = this.series.getItem(i).horizontalAxis;
                if (axis && axis.id === axisID) {
                    return axis;
                }
                axis = this.series.getItem(i).verticalAxis;
                if (axis && axis.id === axisID) {
                    return axis;
                }
            }
        }
        return null;
    };
    RadCartesianChart.prototype._onDetach = function (force) {
        this.detachSeries();
        _super.prototype['_onDetach'].call(this, force);
    };
    RadCartesianChart.prototype.updatePaletteStyle = function (appearance) {
        this._android.setPalette(this.getPaletteForSystemAppearance(appearance));
        this._android.setSelectionPalette(this.getSelectedPaletteForSystemAppearance(appearance));
        if (this.palettes) {
            this.initializer.reloadPalettes(this);
        }
    };
    RadCartesianChart.prototype.getPaletteForSystemAppearance = function (appearance) {
        if (appearance === enums_1.SystemAppearance.dark) {
            return com.telerik.widget.palettes.ChartPalettes.dark(this._android.getContext());
        }
        return com.telerik.widget.palettes.ChartPalettes.light(this._android.getContext());
    };
    RadCartesianChart.prototype.getSelectedPaletteForSystemAppearance = function (appearance) {
        if (appearance === enums_1.SystemAppearance.dark) {
            return com.telerik.widget.palettes.ChartPalettes.darkSelected(this._android.getContext());
        }
        return com.telerik.widget.palettes.ChartPalettes.lightSelected(this._android.getContext());
    };
    RadCartesianChart.prototype.detachSeries = function () {
        var thisAndroid = this._android;
        if (thisAndroid) {
            this._rootLayout = undefined;
            if (this.series) {
                for (var i = 0; i < this.series.length; i++) {
                    if (thisAndroid.getSeries().indexOf(this.series.getItem(i).android) !== -1) {
                        thisAndroid.getSeries().remove(this.series.getItem(i).android);
                    }
                }
            }
            thisAndroid.setHorizontalAxis(undefined);
            thisAndroid.setVerticalAxis(undefined);
            thisAndroid.setGrid(undefined);
        }
    };
    RadCartesianChart.prototype.initNativeAnnotation = function (index) {
        if (this.androidView) {
            this.annotations.getItem(index)._init(this);
            this.annotations.getItem(index)._createNative();
        }
    };
    // get allowZoom/allowPan from axes and init the chart behavior
    RadCartesianChart.prototype.updatePanZoomBehavior = function () {
        if (!this._android) {
            return;
        }
        var zoomHorizontal = false;
        var zoomVertical = false;
        var panHorizontal = false;
        var panVertical = false;
        var i = 0;
        if (this.series) {
            while ((i < this.series.length) && !(panHorizontal && panVertical && zoomHorizontal && zoomVertical)) {
                if (this.series.getItem(i).horizontalAxis) {
                    panHorizontal = panHorizontal || this.series.getItem(i).horizontalAxis.allowPan;
                    zoomHorizontal = zoomHorizontal || this.series.getItem(i).horizontalAxis.allowZoom;
                }
                if (this.series.getItem(i).verticalAxis) {
                    panVertical = panVertical || this.series.getItem(i).verticalAxis.allowPan;
                    zoomVertical = zoomVertical || this.series.getItem(i).verticalAxis.allowZoom;
                }
                i++;
            }
        }
        if (this.horizontalAxis) {
            panHorizontal = panHorizontal || this.horizontalAxis.allowPan;
            zoomHorizontal = zoomHorizontal || this.horizontalAxis.allowZoom;
        }
        if (this.verticalAxis) {
            panVertical = panVertical || this.verticalAxis.allowPan;
            zoomVertical = zoomVertical || this.verticalAxis.allowZoom;
        }
        if (panHorizontal || panVertical || zoomHorizontal || zoomVertical) {
            if (!this._panZoomBehavior) {
                this._panZoomBehavior = new com.telerik.widget.chart.visualization.behaviors.ChartPanAndZoomBehavior();
                this._android._panZoomListener = new ChartPanZoomListener(this);
                this._panZoomBehavior.addPanZoomListener(this._android._panZoomListener);
            }
            if (panHorizontal || panVertical) {
                var panMode = com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.NONE;
                if (panHorizontal && panVertical) {
                    panMode = com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.BOTH;
                }
                else {
                    panMode = panHorizontal ? com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.HORIZONTAL : com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.VERTICAL;
                }
                this._panZoomBehavior.setPanMode(panMode);
            }
            if (zoomHorizontal || zoomVertical) {
                var zoomMode = com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.NONE;
                if (zoomHorizontal && zoomVertical) {
                    zoomMode = com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.BOTH;
                }
                else {
                    zoomMode = zoomHorizontal ? com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.HORIZONTAL : com.telerik.widget.chart.visualization.behaviors.ChartPanZoomMode.VERTICAL;
                }
                this._panZoomBehavior.setZoomMode(zoomMode);
            }
            this._android.getBehaviors().add(this._panZoomBehavior);
        }
        else {
            if (this._panZoomBehavior) {
                this._android.getBehaviors().remove(this._panZoomBehavior);
                this._panZoomBehavior = null;
            }
        }
    };
    RadCartesianChart.prototype._syncZoomValues = function () {
        if (this._android) {
            if (this.horizontalZoom || this.verticalZoom) {
                this._android.setZoom(this.horizontalZoom ? this.horizontalZoom : 1, this.verticalZoom ? this.verticalZoom : 1);
            }
        }
    };
    RadCartesianChart.prototype.initSelectionBehavior = function () {
        if (!this._selectionBehavior) {
            this._selectionBehavior = new com.telerik.widget.chart.visualization.behaviors.ChartSelectionBehavior();
            this._selectionBehavior.setDataPointsSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
            this._selectionBehavior.setSeriesSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
            if (this._android) {
                this._android.getBehaviors().add(this._selectionBehavior);
            }
        }
    };
    RadCartesianChart.prototype.onSeriesChanged = function (oldValue, newValue) {
        _super.prototype.onSeriesChanged.call(this, oldValue, newValue);
        if (newValue) {
            this.updatePanZoomBehavior();
            this._syncZoomValues();
        }
    };
    RadCartesianChart.prototype.onSeriesSelectionModeChanged = function (oldValue, newValue) {
        this.initSelectionBehavior();
        if (newValue === publicEnumModule.ChartSelectionMode.None) {
            this._selectionBehavior.setSeriesSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
        }
        if (newValue === publicEnumModule.ChartSelectionMode.Single) {
            this._selectionBehavior.setSeriesSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.SINGLE);
        }
        if (newValue === publicEnumModule.ChartSelectionMode.Multiple) {
            this._selectionBehavior.setSeriesSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.MULTIPLE);
        }
    };
    RadCartesianChart.prototype.onPointSelectionModeChanged = function (oldValue, newValue) {
        this.initSelectionBehavior();
        if (newValue === publicEnumModule.ChartSelectionMode.None) {
            this._selectionBehavior.setDataPointsSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.NONE);
        }
        if (newValue === publicEnumModule.ChartSelectionMode.Single) {
            this._selectionBehavior.setDataPointsSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.SINGLE);
        }
        if (newValue === publicEnumModule.ChartSelectionMode.Multiple) {
            this._selectionBehavior.setDataPointsSelectionMode(com.telerik.widget.chart.visualization.behaviors.ChartSelectionMode.MULTIPLE);
        }
    };
    RadCartesianChart.prototype.onHorizontalAxisChanged = function (oldValue, newValue) {
        _super.prototype.onHorizontalAxisChanged.call(this, oldValue, newValue);
        if (oldValue) {
            this._removeView(oldValue);
        }
        if (newValue) {
            this._addView(newValue);
        }
    };
    RadCartesianChart.prototype[commonModule.RadCartesianChart.horizontalAxisProperty.setNative] = function (value) {
        this._android.setHorizontalAxis(null);
        this._android.setHorizontalAxis(value.android);
        this.updatePanZoomBehavior();
    };
    RadCartesianChart.prototype.onVerticalAxisChanged = function (oldValue, newValue) {
        _super.prototype.onVerticalAxisChanged.call(this, oldValue, newValue);
        if (oldValue) {
            this._removeView(oldValue);
        }
        if (newValue) {
            this._addView(newValue);
        }
    };
    RadCartesianChart.prototype[commonModule.RadCartesianChart.verticalAxisProperty.setNative] = function (value) {
        this._android.setVerticalAxis(null);
        this._android.setVerticalAxis(value.android);
        this.updatePanZoomBehavior();
    };
    RadCartesianChart.prototype.onGridChanged = function (oldValue, newValue) {
        if (this._android && newValue) {
            this._android.setGrid(null);
            this._android.setGrid(newValue._android);
        }
    };
    RadCartesianChart.prototype.onAnnotationsChanged = function (oldValue, newValue) {
        _super.prototype.onAnnotationsChanged.call(this, oldValue, newValue);
        if (newValue) {
            var newArray = newValue;
            for (var i = 0; i < newArray.length; i++) {
                newArray.getItem(i)._init(this);
            }
        }
    };
    RadCartesianChart.prototype.onHorizontalZoomChanged = function (oldValue, newValue) {
        if (!this._android) {
            return;
        }
        if (newValue !== oldValue) {
            this._syncZoomValues();
        }
        else {
            console.log("WARNING: Vertical zoom must be a number greater or equal to 1");
        }
    };
    RadCartesianChart.prototype.onVerticalZoomChanged = function (oldValue, newValue) {
        if (!this._android) {
            return;
        }
        if (newValue !== oldValue) {
            this._syncZoomValues();
        }
        else {
            console.log("WARNING: Vertical zoom must be a number greater or equal to 1");
        }
    };
    RadCartesianChart.prototype.onTrackballChanged = function (oldValue, newValue) {
        _super.prototype.onTrackballChanged.call(this, oldValue, newValue);
        if (!this._android) {
            return;
        }
        this.removeTrackball(oldValue, newValue);
        if (newValue && (newValue instanceof trackBallCommonModule.Trackball)) {
            newValue.android = new com.telerik.widget.chart.visualization.behaviors.ChartTrackBallBehavior(this._context);
            this._android.getBehaviors().add(newValue.android);
        }
    };
    RadCartesianChart.prototype.removeTrackball = function (oldValue, newValue) {
        if (oldValue && this._android.getBehaviors().indexOf(oldValue.android) !== -1) {
            this._android.getBehaviors().remove(oldValue.android);
        }
    };
    RadCartesianChart.prototype.setNativeTrackballBehavior = function (trackball) {
        if (!trackball) {
            return;
        }
        trackball.android = new com.telerik.widget.chart.visualization.behaviors.ChartTrackBallBehavior(this._context);
        this._android.getBehaviors().add(trackball.android);
    };
    return RadCartesianChart;
}(commonModule.RadCartesianChart));
exports.RadCartesianChart = RadCartesianChart;
