// kinvey-nativescript-sdk/lib/angular/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var nativescript_1 = require("../nativescript");
exports.getAppVersion = nativescript_1.getAppVersion;
exports.setAppVersion = nativescript_1.setAppVersion;
exports.logger = nativescript_1.logger;
exports.Acl = nativescript_1.Acl;
exports.Aggregation = nativescript_1.Aggregation;
exports.StorageProvider = nativescript_1.StorageProvider;
exports.DataStoreType = nativescript_1.DataStoreType;
exports.Errors = nativescript_1.Errors;
exports.Kmd = nativescript_1.Kmd;
exports.Metadata = nativescript_1.Kmd;
exports.Query = nativescript_1.Query;
exports.AuthorizationGrant = nativescript_1.AuthorizationGrant;
var datastore_service_1 = require("./datastore.service");
exports.DataStoreService = datastore_service_1.DataStoreService;
var endpoint_service_1 = require("./endpoint.service");
exports.EndpointService = endpoint_service_1.EndpointService;
var files_service_1 = require("./files.service");
exports.FilesService = files_service_1.FilesService;
var kinvey_module_1 = require("./kinvey.module");
exports.KinveyModule = kinvey_module_1.KinveyModule;
var ping_service_1 = require("./ping.service");
exports.PingService = ping_service_1.PingService;
var user_service_1 = require("./user.service");
exports.UserService = user_service_1.UserService;
