// nativescript-ui-autocomplete/angular/autocomplete-directives.js
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var element_registry_1 = require("nativescript-angular/element-registry");
var __1 = require("./../");
var ListItemContext = /** @class */ (function (_super) {
    __extends(ListItemContext, _super);
    function ListItemContext($implicit, item, index, even, odd) {
        var _this = _super.call(this, item) || this;
        _this.$implicit = $implicit;
        _this.item = item;
        _this.index = index;
        _this.even = even;
        _this.odd = odd;
        return _this;
    }
    return ListItemContext;
}(core_1.ElementRef));
exports.ListItemContext = ListItemContext;
var NG_VIEW = "ng_view";
var RadAutoCompleteTextViewComponent = /** @class */ (function () {
    function RadAutoCompleteTextViewComponent(_elementRef, loader) {
        var _this = this;
        this._elementRef = _elementRef;
        this.loader = loader;
        this.setupItemView = new core_1.EventEmitter();
        this._autoCompleteTextView = _elementRef.nativeElement;
        var component = this;
        this._autoCompleteTextView.itemViewLoader = function (viewType) {
            switch (viewType) {
                case __1.AutoCompleteViewType.ItemView:
                    if (component._itemTemplate && _this.loader) {
                        var nativeItem = _this.loader.createEmbeddedView(component._itemTemplate, new ListItemContext(), 0);
                        var typedView = getItemViewRoot(nativeItem);
                        typedView[NG_VIEW] = nativeItem;
                        return typedView;
                    }
                    return null;
                default:
                    return null;
            }
        };
    }
    Object.defineProperty(RadAutoCompleteTextViewComponent.prototype, "nativeElement", {
        get: function () {
            return this._autoCompleteTextView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadAutoCompleteTextViewComponent.prototype, "autoCompleteTextView", {
        get: function () {
            return this._autoCompleteTextView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadAutoCompleteTextViewComponent.prototype, "itemTemplate", {
        set: function (value) {
            this._itemTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    RadAutoCompleteTextViewComponent.prototype.onItemLoading = function (args) {
        var index = args.index;
        var currentItem = args.data;
        var ngView = args.view[NG_VIEW];
        if (ngView) {
            this.setupViewRef(ngView, currentItem, index);
            this.detectChangesOnChild(ngView, index);
        }
    };
    RadAutoCompleteTextViewComponent.prototype.setupViewRef = function (viewRef, data, index) {
        var context = viewRef.context;
        context.$implicit = data;
        context.item = data;
        context.index = index;
        context.even = (index % 2 === 0);
        context.odd = !context.even;
        this.setupItemView.next({ view: viewRef, data: data, index: index, context: context });
    };
    RadAutoCompleteTextViewComponent.prototype.detectChangesOnChild = function (viewRef, index) {
        // Manually detect changes in child view ref
        // TODO: Is there a better way of getting viewRef's change detector
        var childChangeDetector = viewRef;
        childChangeDetector.markForCheck();
        childChangeDetector.detectChanges();
    };
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], RadAutoCompleteTextViewComponent.prototype, "setupItemView", void 0);
    __decorate([
        core_1.HostListener("itemLoading", ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [__1.CollectionViewEventData]),
        __metadata("design:returntype", void 0)
    ], RadAutoCompleteTextViewComponent.prototype, "onItemLoading", null);
    RadAutoCompleteTextViewComponent = __decorate([
        core_1.Component({
            selector: "RadAutoCompleteTextView",
            template: ""
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __param(1, core_1.Inject(core_1.ViewContainerRef)),
        __metadata("design:paramtypes", [core_1.ElementRef,
            core_1.ViewContainerRef])
    ], RadAutoCompleteTextViewComponent);
    return RadAutoCompleteTextViewComponent;
}());
exports.RadAutoCompleteTextViewComponent = RadAutoCompleteTextViewComponent;
var TKAutoCompleteSuggestionViewDirective = /** @class */ (function () {
    function TKAutoCompleteSuggestionViewDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKAutoCompleteSuggestionViewDirective.prototype.ngOnInit = function () {
        this._suggestionView = this._elementRef.nativeElement;
        this.owner.autoCompleteTextView.suggestionView = this._suggestionView;
    };
    Object.defineProperty(TKAutoCompleteSuggestionViewDirective.prototype, "nativeElement", {
        get: function () {
            return this._suggestionView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TKAutoCompleteSuggestionViewDirective.prototype, "autoCompleteTextView", {
        get: function () {
            return this._suggestionView;
        },
        enumerable: true,
        configurable: true
    });
    TKAutoCompleteSuggestionViewDirective = __decorate([
        core_1.Directive({
            selector: "[tkAutoCompleteSuggestionView]"
        }),
        __param(0, core_1.Inject(RadAutoCompleteTextViewComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadAutoCompleteTextViewComponent,
            core_1.ElementRef])
    ], TKAutoCompleteSuggestionViewDirective);
    return TKAutoCompleteSuggestionViewDirective;
}());
exports.TKAutoCompleteSuggestionViewDirective = TKAutoCompleteSuggestionViewDirective;
var TKSuggestionItemTemplateDirective = /** @class */ (function () {
    function TKSuggestionItemTemplateDirective(owner, template) {
        this.owner = owner;
        this.template = template;
    }
    TKSuggestionItemTemplateDirective.prototype.ngOnInit = function () {
        this.owner.itemTemplate = this.template;
    };
    TKSuggestionItemTemplateDirective = __decorate([
        core_1.Directive({
            selector: "[tkSuggestionItemTemplate]"
        }),
        __param(0, core_1.Inject(RadAutoCompleteTextViewComponent)),
        __param(1, core_1.Inject(core_1.TemplateRef)),
        __metadata("design:paramtypes", [RadAutoCompleteTextViewComponent,
            core_1.TemplateRef])
    ], TKSuggestionItemTemplateDirective);
    return TKSuggestionItemTemplateDirective;
}());
exports.TKSuggestionItemTemplateDirective = TKSuggestionItemTemplateDirective;
function getItemViewRoot(viewRef, rootLocator) {
    if (rootLocator === void 0) { rootLocator = element_registry_1.getSingleViewRecursive; }
    var rootView = rootLocator(viewRef.rootNodes, 0);
    rootView.on("unloaded", function () {
        viewRef.destroy();
        delete rootView[NG_VIEW];
    });
    return rootView;
}
exports.AUTOCOMPLETETEXTVIEW_DIRECTIVES = [RadAutoCompleteTextViewComponent, TKAutoCompleteSuggestionViewDirective, TKSuggestionItemTemplateDirective];
element_registry_1.registerElement("RadAutoCompleteTextView", function () { return __1.RadAutoCompleteTextView; });
element_registry_1.registerElement("SuggestionView", function () { return __1.SuggestionView; });
var NativeScriptUIAutoCompleteTextViewModule = /** @class */ (function () {
    function NativeScriptUIAutoCompleteTextViewModule() {
    }
    NativeScriptUIAutoCompleteTextViewModule = __decorate([
        core_1.NgModule({
            declarations: [exports.AUTOCOMPLETETEXTVIEW_DIRECTIVES],
            exports: [exports.AUTOCOMPLETETEXTVIEW_DIRECTIVES]
        })
    ], NativeScriptUIAutoCompleteTextViewModule);
    return NativeScriptUIAutoCompleteTextViewModule;
}());
exports.NativeScriptUIAutoCompleteTextViewModule = NativeScriptUIAutoCompleteTextViewModule;
