// kinvey-js-sdk/lib/http/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var url_1 = require("url");
var url_join_1 = __importDefault(require("url-join"));
var isPlainObject_1 = __importDefault(require("lodash/isPlainObject"));
var kinvey_1 = require("../kinvey");
function clean(value) {
    return Object.keys(value)
        .reduce(function (cleanVal, key) {
        var objVal = value[key];
        if (isPlainObject_1.default(objVal)) {
            objVal = clean(objVal);
        }
        if (typeof objVal !== 'undefined' && objVal !== null) {
            cleanVal[key] = objVal;
        }
        return cleanVal;
    }, {});
}
var KinveyBaasNamespace;
(function (KinveyBaasNamespace) {
    KinveyBaasNamespace["AppData"] = "appdata";
    KinveyBaasNamespace["Blob"] = "blob";
    KinveyBaasNamespace["Push"] = "push";
    KinveyBaasNamespace["Rpc"] = "rpc";
    KinveyBaasNamespace["User"] = "user";
})(KinveyBaasNamespace = exports.KinveyBaasNamespace || (exports.KinveyBaasNamespace = {}));
function formatKinveyBaasUrl(namespace, path, query) {
    return url_1.format({
        protocol: kinvey_1.getBaasProtocol(),
        host: kinvey_1.getBaasHost(),
        pathname: path ? url_join_1.default(namespace, kinvey_1.getAppKey(), path) : url_join_1.default(namespace, kinvey_1.getAppKey()),
        query: query ? clean(query) : undefined
    });
}
exports.formatKinveyBaasUrl = formatKinveyBaasUrl;
function formatKinveyAuthUrl(path, query) {
    return url_1.format({
        protocol: kinvey_1.getAuthProtocol(),
        host: kinvey_1.getAuthHost(),
        pathname: path,
        query: query ? clean(query) : undefined
    });
}
exports.formatKinveyAuthUrl = formatKinveyAuthUrl;
