// nativescript-ui-gauge/angular/gauges-directives.js
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var element_registry_1 = require("nativescript-angular/element-registry");
var __1 = require("./../");
var observable_array_1 = require("tns-core-modules/data/observable-array");
var RadRadialGaugeComponent = /** @class */ (function () {
    function RadRadialGaugeComponent(_elementRef) {
        this._elementRef = _elementRef;
        this._radialGauge = _elementRef.nativeElement;
    }
    Object.defineProperty(RadRadialGaugeComponent.prototype, "nativeElement", {
        get: function () {
            return this._radialGauge;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadRadialGaugeComponent.prototype, "radialGauge", {
        get: function () {
            return this._radialGauge;
        },
        enumerable: true,
        configurable: true
    });
    RadRadialGaugeComponent = __decorate([
        core_1.Component({
            selector: "RadRadialGauge",
            template: ""
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], RadRadialGaugeComponent);
    return RadRadialGaugeComponent;
}());
exports.RadRadialGaugeComponent = RadRadialGaugeComponent;
var TKRadialScaleDirective = /** @class */ (function () {
    function TKRadialScaleDirective(_elementRef) {
        this._elementRef = _elementRef;
        this._radialScale = this._elementRef.nativeElement;
    }
    Object.defineProperty(TKRadialScaleDirective.prototype, "radialScale", {
        get: function () {
            return this._radialScale;
        },
        enumerable: true,
        configurable: true
    });
    TKRadialScaleDirective = __decorate([
        core_1.Directive({
            selector: "RadialScale"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], TKRadialScaleDirective);
    return TKRadialScaleDirective;
}());
exports.TKRadialScaleDirective = TKRadialScaleDirective;
var TKRadialBarIndicatorDirective = /** @class */ (function () {
    function TKRadialBarIndicatorDirective(_elementRef) {
        this._elementRef = _elementRef;
        this._barIndicator = this._elementRef.nativeElement;
    }
    Object.defineProperty(TKRadialBarIndicatorDirective.prototype, "radialScale", {
        get: function () {
            return this._barIndicator;
        },
        enumerable: true,
        configurable: true
    });
    TKRadialBarIndicatorDirective = __decorate([
        core_1.Directive({
            selector: "RadialBarIndicator"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], TKRadialBarIndicatorDirective);
    return TKRadialBarIndicatorDirective;
}());
exports.TKRadialBarIndicatorDirective = TKRadialBarIndicatorDirective;
var TKRadialNeedleDirective = /** @class */ (function () {
    function TKRadialNeedleDirective(_elementRef) {
        this._elementRef = _elementRef;
        this._radialNeedle = this._elementRef.nativeElement;
    }
    Object.defineProperty(TKRadialNeedleDirective.prototype, "radialNeedle", {
        get: function () {
            return this._radialNeedle;
        },
        enumerable: true,
        configurable: true
    });
    TKRadialNeedleDirective = __decorate([
        core_1.Directive({
            selector: "RadialNeedle",
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], TKRadialNeedleDirective);
    return TKRadialNeedleDirective;
}());
exports.TKRadialNeedleDirective = TKRadialNeedleDirective;
var TKRadialGaugeScalesDirective = /** @class */ (function () {
    function TKRadialGaugeScalesDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialGaugeScalesDirective.prototype.ngOnInit = function () {
        var scale = this._elementRef.nativeElement;
        var gauge = this.owner.radialGauge;
        if (gauge.scales) {
            gauge.scales.push(scale);
        }
        else {
            gauge.scales = new observable_array_1.ObservableArray([scale]);
        }
    };
    TKRadialGaugeScalesDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialGaugeScales]"
        }),
        __param(0, core_1.Inject(RadRadialGaugeComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadRadialGaugeComponent,
            core_1.ElementRef])
    ], TKRadialGaugeScalesDirective);
    return TKRadialGaugeScalesDirective;
}());
exports.TKRadialGaugeScalesDirective = TKRadialGaugeScalesDirective;
var TKRadialScaleStyleDirective = /** @class */ (function () {
    function TKRadialScaleStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialScaleStyleDirective.prototype.ngOnInit = function () {
        var radialScaleStyle = this._elementRef.nativeElement;
        this.owner.radialScale.scaleStyle = radialScaleStyle;
    };
    TKRadialScaleStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialScaleStyle]"
        }),
        __param(0, core_1.Inject(TKRadialScaleDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKRadialScaleDirective,
            core_1.ElementRef])
    ], TKRadialScaleStyleDirective);
    return TKRadialScaleStyleDirective;
}());
exports.TKRadialScaleStyleDirective = TKRadialScaleStyleDirective;
var TKRadialScaleIndicatorsDirective = /** @class */ (function () {
    function TKRadialScaleIndicatorsDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialScaleIndicatorsDirective.prototype.ngOnInit = function () {
        var barIndicator = this._elementRef.nativeElement;
        var scale = this.owner.radialScale;
        if (scale.indicators) {
            scale.indicators.push(barIndicator);
        }
        else {
            scale.indicators = new observable_array_1.ObservableArray([barIndicator]);
        }
    };
    TKRadialScaleIndicatorsDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialScaleIndicators]"
        }),
        __param(0, core_1.Inject(TKRadialScaleDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKRadialScaleDirective,
            core_1.ElementRef])
    ], TKRadialScaleIndicatorsDirective);
    return TKRadialScaleIndicatorsDirective;
}());
exports.TKRadialScaleIndicatorsDirective = TKRadialScaleIndicatorsDirective;
var TKRadialBarIndicatorStyleDirective = /** @class */ (function () {
    function TKRadialBarIndicatorStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialBarIndicatorStyleDirective.prototype.ngOnInit = function () {
        var indicatorStyle = this._elementRef.nativeElement;
        this.owner.radialScale.indicatorStyle = indicatorStyle;
    };
    TKRadialBarIndicatorStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialBarIndicatorStyle]"
        }),
        __param(0, core_1.Inject(TKRadialBarIndicatorDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKRadialBarIndicatorDirective,
            core_1.ElementRef])
    ], TKRadialBarIndicatorStyleDirective);
    return TKRadialBarIndicatorStyleDirective;
}());
exports.TKRadialBarIndicatorStyleDirective = TKRadialBarIndicatorStyleDirective;
var TKRadialGaugeTitleStyleDirective = /** @class */ (function () {
    function TKRadialGaugeTitleStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialGaugeTitleStyleDirective.prototype.ngOnInit = function () {
        var style = this._elementRef.nativeElement;
        this.owner.radialGauge.titleStyle = style;
    };
    TKRadialGaugeTitleStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialGaugeTitleStyle]"
        }),
        __param(0, core_1.Inject(RadRadialGaugeComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadRadialGaugeComponent,
            core_1.ElementRef])
    ], TKRadialGaugeTitleStyleDirective);
    return TKRadialGaugeTitleStyleDirective;
}());
exports.TKRadialGaugeTitleStyleDirective = TKRadialGaugeTitleStyleDirective;
var TKRadialGaugeSubtitleStyleDirective = /** @class */ (function () {
    function TKRadialGaugeSubtitleStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialGaugeSubtitleStyleDirective.prototype.ngOnInit = function () {
        var style = this._elementRef.nativeElement;
        this.owner.radialGauge.subtitleStyle = style;
    };
    TKRadialGaugeSubtitleStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialGaugeSubtitleStyle]"
        }),
        __param(0, core_1.Inject(RadRadialGaugeComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadRadialGaugeComponent,
            core_1.ElementRef])
    ], TKRadialGaugeSubtitleStyleDirective);
    return TKRadialGaugeSubtitleStyleDirective;
}());
exports.TKRadialGaugeSubtitleStyleDirective = TKRadialGaugeSubtitleStyleDirective;
var TKRadialNeedleStyleDirective = /** @class */ (function () {
    function TKRadialNeedleStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRadialNeedleStyleDirective.prototype.ngOnInit = function () {
        var style = this._elementRef.nativeElement;
        this.owner.radialNeedle.needleStyle = style;
    };
    TKRadialNeedleStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRadialNeedleStyle]"
        }),
        __param(0, core_1.Inject(TKRadialNeedleDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKRadialNeedleDirective,
            core_1.ElementRef])
    ], TKRadialNeedleStyleDirective);
    return TKRadialNeedleStyleDirective;
}());
exports.TKRadialNeedleStyleDirective = TKRadialNeedleStyleDirective;
exports.GAUGES_DIRECTIVES = [RadRadialGaugeComponent, TKRadialScaleDirective, TKRadialGaugeScalesDirective, TKRadialScaleStyleDirective, TKRadialScaleIndicatorsDirective, TKRadialBarIndicatorStyleDirective, TKRadialBarIndicatorDirective, TKRadialGaugeTitleStyleDirective, TKRadialNeedleStyleDirective, TKRadialNeedleDirective, TKRadialGaugeSubtitleStyleDirective];
element_registry_1.registerElement("RadRadialGauge", function () { return __1.RadRadialGauge; });
element_registry_1.registerElement("RadialScale", function () { return __1.RadialScale; });
element_registry_1.registerElement("ScaleStyle", function () { return __1.ScaleStyle; });
element_registry_1.registerElement("RadialBarIndicator", function () { return __1.RadialBarIndicator; });
element_registry_1.registerElement("BarIndicatorStyle", function () { return __1.BarIndicatorStyle; });
element_registry_1.registerElement("RadialNeedle", function () { return __1.RadialNeedle; });
element_registry_1.registerElement("TitleStyle", function () { return __1.TitleStyle; });
element_registry_1.registerElement("SubtitleStyle", function () { return __1.SubtitleStyle; });
element_registry_1.registerElement("NeedleStyle", function () { return __1.NeedleStyle; });
var NativeScriptUIGaugeModule = /** @class */ (function () {
    function NativeScriptUIGaugeModule() {
    }
    NativeScriptUIGaugeModule = __decorate([
        core_1.NgModule({
            declarations: [exports.GAUGES_DIRECTIVES],
            exports: [exports.GAUGES_DIRECTIVES]
        })
    ], NativeScriptUIGaugeModule);
    return NativeScriptUIGaugeModule;
}());
exports.NativeScriptUIGaugeModule = NativeScriptUIGaugeModule;
