// nativescript-secure-storage/secure-storage.android.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var secure_storage_common_1 = require("./secure-storage.common");
var utils = require("tns-core-modules/utils/utils");
var SecureStorage = (function (_super) {
    __extends(SecureStorage, _super);
    function SecureStorage() {
        var _this = _super.call(this) || this;
        _this.hawk = com.orhanobut.hawk.Hawk.init(utils.ad.getApplicationContext()).build();
        return _this;
    }
    SecureStorage.prototype.get = function (arg) {
        return new Promise(function (resolve, reject) {
            resolve(com.orhanobut.hawk.Hawk.get(arg.key));
        });
    };
    SecureStorage.prototype.getSync = function (arg) {
        return com.orhanobut.hawk.Hawk.get(arg.key);
    };
    SecureStorage.prototype.set = function (arg) {
        return new Promise(function (resolve, reject) {
            resolve(com.orhanobut.hawk.Hawk.put(arg.key, arg.value));
        });
    };
    SecureStorage.prototype.setSync = function (arg) {
        return com.orhanobut.hawk.Hawk.put(arg.key, arg.value);
    };
    SecureStorage.prototype.remove = function (arg) {
        return new Promise(function (resolve, reject) {
            resolve(com.orhanobut.hawk.Hawk.delete(arg.key));
        });
    };
    SecureStorage.prototype.removeSync = function (arg) {
        return com.orhanobut.hawk.Hawk.delete(arg.key);
    };
    SecureStorage.prototype.removeAll = function (arg) {
        return new Promise(function (resolve, reject) {
            resolve(com.orhanobut.hawk.Hawk.deleteAll());
        });
    };
    SecureStorage.prototype.removeAllSync = function (arg) {
        return com.orhanobut.hawk.Hawk.deleteAll();
    };
    return SecureStorage;
}(secure_storage_common_1.SecureStorageCommon));
exports.SecureStorage = SecureStorage;
