// kinvey-js-sdk/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var acl_1 = require("./acl");
exports.Acl = acl_1.Acl;
var aggregation_1 = require("./aggregation");
exports.Aggregation = aggregation_1.Aggregation;
var datastore_1 = require("./datastore");
exports.DataStoreType = datastore_1.DataStoreType;
var endpoint_1 = require("./endpoint");
var Errors = __importStar(require("./errors"));
exports.Errors = Errors;
var Files = __importStar(require("./files"));
exports.Files = Files;
var init_1 = require("./init");
exports.init = init_1.init;
exports.initialize = init_1.initialize;
var kmd_1 = require("./kmd");
exports.Kmd = kmd_1.Kmd;
exports.Metadata = kmd_1.Kmd;
var log_1 = require("./log");
exports.logger = log_1.logger;
var ping_1 = require("./ping");
exports.ping = ping_1.ping;
var query_1 = require("./query");
exports.Query = query_1.Query;
var user_1 = require("./user");
exports.User = user_1.User;
exports.AuthorizationGrant = user_1.AuthorizationGrant;
var http_1 = require("./http");
exports.getAppVersion = http_1.getAppVersion;
exports.setAppVersion = http_1.setAppVersion;
var CustomEndpoint = { execute: endpoint_1.endpoint };
exports.CustomEndpoint = CustomEndpoint;
var DataStore = { collection: datastore_1.collection, getInstance: datastore_1.getInstance, clearCache: datastore_1.clearCache };
exports.DataStore = DataStore;
