// kinvey-js-sdk/lib/kinvey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var isNumber = require("lodash/isNumber");
var config_1 = require("./config");
function getAppKey() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    return config.appKey;
}
exports.getAppKey = getAppKey;
function getAppSecret() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    return config.appSecret;
}
exports.getAppSecret = getAppSecret;
function getMasterSecret() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    return config.masterSecret;
}
exports.getMasterSecret = getMasterSecret;
function getInstanceId() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    return config.instanceId;
}
exports.getInstanceId = getInstanceId;
function getBaasProtocol() {
    return 'https:';
}
exports.getBaasProtocol = getBaasProtocol;
function getBaasHost() {
    var instanceId = getInstanceId();
    if (instanceId) {
        return instanceId + "-baas.kinvey.com";
    }
    return 'baas.kinvey.com';
}
exports.getBaasHost = getBaasHost;
function getAuthProtocol() {
    return 'https:';
}
exports.getAuthProtocol = getAuthProtocol;
function getAuthHost() {
    var instanceId = getInstanceId();
    if (instanceId) {
        return instanceId + "-auth.kinvey.com";
    }
    return 'auth.kinvey.com';
}
exports.getAuthHost = getAuthHost;
function getDefaultTimeout() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    if (isNumber(config.defaultTimeout)) {
        return config.defaultTimeout;
    }
    return 60000; // 1 minute
}
exports.getDefaultTimeout = getDefaultTimeout;
function getEncryptionKey() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    return config.encryptionKey;
}
exports.getEncryptionKey = getEncryptionKey;
function getApiVersion() {
    var config = config_1.getConfig(config_1.ConfigKey.KinveyConfig);
    if (isNumber(config.apiVersion) && config.apiVersion >= 3) {
        return config.apiVersion;
    }
    return 4;
}
exports.getApiVersion = getApiVersion;
