// kinvey-js-sdk/lib/kmd.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var isPlainObject_1 = __importDefault(require("lodash/isPlainObject"));
var kinvey_1 = require("./errors/kinvey");
/**
 * This class provides a way to access the KMD (Kinvey Metadata)
 * information for an entity.
 */
var Kmd = /** @class */ (function () {
    function Kmd(entity) {
        if (!isPlainObject_1.default(entity)) {
            throw new kinvey_1.KinveyError('entity argument must be an object');
        }
        entity._kmd = entity._kmd || {}; // eslint-disable-line no-param-reassign
        this.entity = entity;
    }
    Object.defineProperty(Kmd.prototype, "authtoken", {
        /**
         * Get the auth token.
         *
         * @returns {string} _kmd.authtoken
         */
        get: function () {
            return (this.entity._kmd && this.entity._kmd.authtoken) || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "ect", {
        /**
         * Get created at time.
         *
         * @returns {Date?} _kmd.ect
         */
        get: function () {
            return this.createdAt;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "createdAt", {
        /**
         * Get created at time.
         *
         * @returns {Date?} _kmd.ect
         */
        get: function () {
            if (this.entity._kmd && this.entity._kmd.ect) {
                return new Date(this.entity._kmd.ect);
            }
            return undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "lmt", {
        /**
         * Get last modified time.
         *
         * @returns {Date?} _kmd.lmt
         */
        get: function () {
            return this.updatedAt;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "lastModified", {
        /**
         * Get last modified time.
         *
         * @returns {Date?} _kmd.lmt
         */
        get: function () {
            return this.updatedAt;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "updatedAt", {
        /**
         * Get last modified time.
         *
         * @returns {Date?} _kmd.lmt
         */
        get: function () {
            if (this.entity._kmd && this.entity._kmd.lmt) {
                return new Date(this.entity._kmd.lmt);
            }
            return undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Kmd.prototype, "emailVerification", {
        /**
         * Get the email verification details.
         *
         * @returns {Object} _kmd.emailVerification
         */
        get: function () {
            if (this.entity._kmd && this.entity._kmd.emailVerification) {
                return this.entity._kmd.emailVerification;
            }
            return undefined;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Checks if an email for a user has been confirmed.
     *
     * @returns {boolean} True if the email has been confirmed otherwise false
     */
    Kmd.prototype.isEmailConfirmed = function () {
        if (this.emailVerification) {
            return this.emailVerification.status === 'confirmed';
        }
        return false;
    };
    /**
     * Checks if the entity has been created locally.
     *
     * @returns {boolean} True if the entity has been created locally otherwise false
     */
    Kmd.prototype.isLocal = function () {
        return (this.entity._kmd && this.entity._kmd.local === true) || false;
    };
    Kmd.prototype.toPlainObject = function () {
        return this.entity._kmd;
    };
    return Kmd;
}());
exports.Kmd = Kmd;
