// kinvey-js-sdk/lib/errors/base.js
"use strict";
// export default class BaseError extends Error {
//   constructor(message = 'An error occurred.', debug?: string, code?: number, kinveyRequestId?: string, ...args) {
//     // Pass remaining arguments (including vendor specific ones) to parent constructor
//     super(...args);
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
//     // Maintains proper stack trace for where our error was thrown (only available on V8)
//     if (Error.captureStackTrace) {
//       Error.captureStackTrace(this, BaseError);
//     }
//     // Custom debugging information
//     this.name = 'BaseError';
//     this.message = message;
//     this.debug = debug;
//     this.code = code;
//     this.kinveyRequestId = kinveyRequestId;
//   }
// }
var BaseError = /** @class */ (function (_super) {
    __extends(BaseError, _super);
    function BaseError(message) {
        var _newTarget = this.constructor;
        if (message === void 0) { message = 'An error occurred.'; }
        var _this = _super.call(this, message) || this;
        Object.setPrototypeOf(_this, _newTarget.prototype);
        _this.name = 'BaseError';
        return _this;
    }
    return BaseError;
}(Error));
exports.BaseError = BaseError;
