// nativescript-ui-chart/visualization/parts/chart-annotation.common.js
Object.defineProperty(exports, "__esModule", { value: true });
var view_1 = require("tns-core-modules/ui/core/view");
var chart_public_enum_1 = require("../../misc/chart-public-enum");
var color_1 = require("tns-core-modules/color");
var CartesianChartAnnotation = /** @class */ (function (_super) {
    __extends(CartesianChartAnnotation, _super);
    function CartesianChartAnnotation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CartesianChartAnnotation.prototype.onAxisIdPropertyChanged = function (oldValue, newValue) {
        this.onAxisIdChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onZPositionPropertyChanged = function (oldValue, newValue) {
        this.onZPositionChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onHiddenPropertyChanged = function (oldValue, newValue) {
        this.onHiddenChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onStrokeWidthPropertyChanged = function (oldValue, newValue) {
        this.onStrokeWidthChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onStrokeColorPropertyChanged = function (oldValue, newValue) {
        this.onStrokeColorChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onStrokeDashPatternPropertyChanged = function (oldValue, newValue) {
        this.onStrokeDashPatternChanged(oldValue, newValue);
    };
    CartesianChartAnnotation.prototype.onAxisIdChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onZPositionChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onHiddenChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onStrokeWidthChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onStrokeColorChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onStrokeDashPatternChanged = function (oldValue, newValue) {
    };
    CartesianChartAnnotation.prototype.onOwnerChanged = function () {
    };
    Object.defineProperty(CartesianChartAnnotation.prototype, "owner", {
        get: function () {
            return this._owner;
        },
        set: function (value) {
            this._owner = value;
            this.onOwnerChanged();
        },
        enumerable: true,
        configurable: true
    });
    CartesianChartAnnotation.axisIdProperty = new view_1.Property({
        name: "axisId",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onAxisIdPropertyChanged(oldValue, newValue);
        },
    });
    CartesianChartAnnotation.zPositionProperty = new view_1.Property({
        name: "zPosition",
        defaultValue: undefined,
        valueConverter: function (value) { return chart_public_enum_1.ChartAnnotationZPosition[chart_public_enum_1.toPascalCase(value)]; },
        valueChanged: function (target, oldValue, newValue) {
            target.onZPositionPropertyChanged(oldValue, newValue);
        },
    });
    CartesianChartAnnotation.hiddenProperty = new view_1.Property({
        name: "hidden",
        defaultValue: undefined,
        valueConverter: view_1.booleanConverter,
        valueChanged: function (target, oldValue, newValue) {
            target.onHiddenPropertyChanged(oldValue, newValue);
        },
    });
    CartesianChartAnnotation.strokeWidthProperty = new view_1.Property({
        name: "strokeWidth",
        defaultValue: undefined,
        valueConverter: parseFloat,
        valueChanged: function (target, oldValue, newValue) {
            target.onStrokeWidthPropertyChanged(oldValue, newValue);
        },
    });
    CartesianChartAnnotation.strokeColorProperty = new view_1.Property({
        name: "strokeColor",
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) {
            target.onStrokeColorPropertyChanged(oldValue, newValue);
        },
    });
    CartesianChartAnnotation.strokeDashPatternProperty = new view_1.Property({
        name: "strokeDashPattern",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onStrokeDashPatternPropertyChanged(oldValue, newValue);
        },
    });
    return CartesianChartAnnotation;
}(view_1.ViewBase));
exports.CartesianChartAnnotation = CartesianChartAnnotation;
CartesianChartAnnotation.axisIdProperty.register(CartesianChartAnnotation);
CartesianChartAnnotation.zPositionProperty.register(CartesianChartAnnotation);
CartesianChartAnnotation.hiddenProperty.register(CartesianChartAnnotation);
CartesianChartAnnotation.strokeWidthProperty.register(CartesianChartAnnotation);
CartesianChartAnnotation.strokeColorProperty.register(CartesianChartAnnotation);
CartesianChartAnnotation.strokeDashPatternProperty.register(CartesianChartAnnotation);
var ChartGridLineAnnotation = /** @class */ (function (_super) {
    __extends(ChartGridLineAnnotation, _super);
    function ChartGridLineAnnotation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChartGridLineAnnotation.prototype.onValuePropertyChanged = function (oldValue, newValue) {
        this.onValueChanged(oldValue, newValue);
    };
    ChartGridLineAnnotation.prototype.onValueChanged = function (oldValue, newValue) {
    };
    ChartGridLineAnnotation.valueProperty = new view_1.Property({
        name: "value",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onValuePropertyChanged(oldValue, newValue);
        },
    });
    return ChartGridLineAnnotation;
}(CartesianChartAnnotation));
exports.ChartGridLineAnnotation = ChartGridLineAnnotation;
ChartGridLineAnnotation.valueProperty.register(ChartGridLineAnnotation);
var ChartPlotBandAnnotation = /** @class */ (function (_super) {
    __extends(ChartPlotBandAnnotation, _super);
    function ChartPlotBandAnnotation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChartPlotBandAnnotation.prototype.onMinValuePropertyChanged = function (oldValue, newValue) {
        this.onMinValueChanged(oldValue, newValue);
    };
    ChartPlotBandAnnotation.prototype.onMaxValuePropertyChanged = function (oldValue, newValue) {
        this.onMaxValueChanged(oldValue, newValue);
    };
    ChartPlotBandAnnotation.prototype.onFillColorPropertyChanged = function (oldValue, newValue) {
        this.onFillColorChanged(oldValue, newValue);
    };
    ChartPlotBandAnnotation.prototype.onMinValueChanged = function (oldValue, newValue) {
    };
    ChartPlotBandAnnotation.prototype.onMaxValueChanged = function (oldValue, newValue) {
    };
    ChartPlotBandAnnotation.prototype.onFillColorChanged = function (oldValue, newValue) {
    };
    ChartPlotBandAnnotation.minValueProperty = new view_1.Property({
        name: "minValue",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onMinValuePropertyChanged(oldValue, newValue);
        },
    });
    ChartPlotBandAnnotation.maxValueProperty = new view_1.Property({
        name: "maxValue",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onMaxValuePropertyChanged(oldValue, newValue);
        },
    });
    ChartPlotBandAnnotation.fillColorProperty = new view_1.Property({
        name: "fillColor",
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) {
            target.onFillColorPropertyChanged(oldValue, newValue);
        },
    });
    return ChartPlotBandAnnotation;
}(CartesianChartAnnotation));
exports.ChartPlotBandAnnotation = ChartPlotBandAnnotation;
ChartPlotBandAnnotation.minValueProperty.register(ChartPlotBandAnnotation);
ChartPlotBandAnnotation.maxValueProperty.register(ChartPlotBandAnnotation);
ChartPlotBandAnnotation.fillColorProperty.register(ChartPlotBandAnnotation);
