// nativescript-ui-chart/visualization/parts/chart-label-style.common.js
Object.defineProperty(exports, "__esModule", { value: true });
var view_1 = require("tns-core-modules/ui/core/view");
var color_1 = require("tns-core-modules/color");
var PointLabelStyle = /** @class */ (function (_super) {
    __extends(PointLabelStyle, _super);
    function PointLabelStyle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PointLabelStyle.strokeWidthProperty = new view_1.Property({
        name: "strokeWidth",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.strokeColorProperty = new view_1.Property({
        name: "strokeColor",
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.fillColorProperty = new view_1.Property({
        name: "fillColor",
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.textColorProperty = new view_1.Property({
        name: "textColor",
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.textSizeProperty = new view_1.Property({
        name: "textSize",
        defaultValue: 10,
        valueConverter: parseFloat,
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.marginProperty = new view_1.Property({
        name: "margin",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.textFormatProperty = new view_1.Property({
        name: "textFormat",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
        },
    });
    PointLabelStyle.fontNameProperty = new view_1.Property({
        name: "fontName",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) { },
    });
    PointLabelStyle.fontStyleProperty = new view_1.Property({
        name: "fontStyle",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) { },
    });
    return PointLabelStyle;
}(view_1.ViewBase));
exports.PointLabelStyle = PointLabelStyle;
PointLabelStyle.strokeWidthProperty.register(PointLabelStyle);
PointLabelStyle.strokeColorProperty.register(PointLabelStyle);
PointLabelStyle.fillColorProperty.register(PointLabelStyle);
PointLabelStyle.textColorProperty.register(PointLabelStyle);
PointLabelStyle.textSizeProperty.register(PointLabelStyle);
PointLabelStyle.marginProperty.register(PointLabelStyle);
PointLabelStyle.textFormatProperty.register(PointLabelStyle);
PointLabelStyle.fontNameProperty.register(PointLabelStyle);
PointLabelStyle.fontStyleProperty.register(PointLabelStyle);
