// nativescript-ui-chart/visualization/parts/chart-track-ball.common.js
Object.defineProperty(exports, "__esModule", { value: true });
var chart_public_enum_1 = require("../../misc/chart-public-enum");
var view_1 = require("tns-core-modules/ui/core/view");
var Trackball = /** @class */ (function (_super) {
    __extends(Trackball, _super);
    function Trackball() {
        return _super.call(this) || this;
    }
    Object.defineProperty(Trackball.prototype, "android", {
        get: function () {
            return this._android;
        },
        set: function (value) {
            this._android = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Trackball.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    Trackball.prototype.onSnapModePropertyChanged = function (oldValue, newValue) {
        this.onSnapModeChanged(oldValue, newValue);
    };
    Trackball.prototype.onShowIntersectionPointsPropertyChanged = function (oldValue, newValue) {
        this.onShowIntersectionPointsChanged(oldValue, newValue);
    };
    Trackball.prototype.onSnapModeChanged = function (oldValue, newValue) { };
    Trackball.prototype.onShowIntersectionPointsChanged = function (oldValue, newValue) { };
    Trackball.prototype.onOwnerChanged = function () { };
    Trackball.trackBallContentRequestedEvent = "trackBallContentRequested";
    Trackball.snapModeProperty = new view_1.Property({
        name: "snapMode",
        defaultValue: chart_public_enum_1.ChartTrackballSnapMode.ClosestPoint,
        valueConverter: function (value) { return chart_public_enum_1.ChartTrackballSnapMode[chart_public_enum_1.toPascalCase(value)]; },
        valueChanged: function (target, oldValue, newValue) {
            target.onSnapModePropertyChanged(oldValue, newValue);
        },
    });
    Trackball.showIntersectionPointsProperty = new view_1.Property({
        name: "showIntersectionPoints",
        defaultValue: false,
        valueConverter: view_1.booleanConverter,
        valueChanged: function (target, oldValue, newValue) {
            target.onShowIntersectionPointsPropertyChanged(oldValue, newValue);
        },
    });
    return Trackball;
}(view_1.ViewBase));
exports.Trackball = Trackball;
Trackball.snapModeProperty.register(Trackball);
Trackball.showIntersectionPointsProperty.register(Trackball);
