// nativescript-ui-chart/angular/chart-directives.js
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var elementRegistryModule = require("nativescript-angular/element-registry");
var chartModule = require("./../");
var observable_array_1 = require("tns-core-modules/data/observable-array");
var SeriesComponentBase = /** @class */ (function () {
    function SeriesComponentBase() {
    }
    Object.defineProperty(SeriesComponentBase.prototype, "nativeSeries", {
        get: function () {
            return this._nativeSeries;
        },
        enumerable: true,
        configurable: true
    });
    return SeriesComponentBase;
}());
exports.SeriesComponentBase = SeriesComponentBase;
var RadCartesianChartComponent = /** @class */ (function () {
    function RadCartesianChartComponent(_elementRef) {
        this._elementRef = _elementRef;
        this._nativeChart = _elementRef.nativeElement;
    }
    Object.defineProperty(RadCartesianChartComponent.prototype, "nativeElement", {
        get: function () {
            return this._nativeChart;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadCartesianChartComponent.prototype, "cartesianChart", {
        get: function () {
            return this._nativeChart;
        },
        enumerable: true,
        configurable: true
    });
    RadCartesianChartComponent = __decorate([
        core_1.Component({
            selector: "RadCartesianChart",
            template: "",
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], RadCartesianChartComponent);
    return RadCartesianChartComponent;
}());
exports.RadCartesianChartComponent = RadCartesianChartComponent;
var RadPieChartComponent = /** @class */ (function () {
    function RadPieChartComponent(_elementRef) {
        this._elementRef = _elementRef;
        this._nativeChart = _elementRef.nativeElement;
    }
    Object.defineProperty(RadPieChartComponent.prototype, "nativeElement", {
        get: function () {
            return this._nativeChart;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadPieChartComponent.prototype, "pieChart", {
        get: function () {
            return this._nativeChart;
        },
        enumerable: true,
        configurable: true
    });
    RadPieChartComponent = __decorate([
        core_1.Component({
            selector: "RadPieChart",
            template: "",
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], RadPieChartComponent);
    return RadPieChartComponent;
}());
exports.RadPieChartComponent = RadPieChartComponent;
var RadLegendViewDirective = /** @class */ (function () {
    function RadLegendViewDirective() {
    }
    RadLegendViewDirective = __decorate([
        core_1.Directive({
            selector: "RadLegendView"
        }),
        __metadata("design:paramtypes", [])
    ], RadLegendViewDirective);
    return RadLegendViewDirective;
}());
exports.RadLegendViewDirective = RadLegendViewDirective;
var CategoricalAxisDirective = /** @class */ (function () {
    function CategoricalAxisDirective() {
    }
    CategoricalAxisDirective = __decorate([
        core_1.Directive({
            selector: "CategoricalAxis"
        }),
        __metadata("design:paramtypes", [])
    ], CategoricalAxisDirective);
    return CategoricalAxisDirective;
}());
exports.CategoricalAxisDirective = CategoricalAxisDirective;
var LinearAxisDirective = /** @class */ (function () {
    function LinearAxisDirective() {
    }
    LinearAxisDirective = __decorate([
        core_1.Directive({
            selector: "LinearAxis"
        }),
        __metadata("design:paramtypes", [])
    ], LinearAxisDirective);
    return LinearAxisDirective;
}());
exports.LinearAxisDirective = LinearAxisDirective;
var DateTimeCategoricalAxisDirective = /** @class */ (function () {
    function DateTimeCategoricalAxisDirective() {
    }
    DateTimeCategoricalAxisDirective = __decorate([
        core_1.Directive({
            selector: "DateTimeCategoricalAxis"
        }),
        __metadata("design:paramtypes", [])
    ], DateTimeCategoricalAxisDirective);
    return DateTimeCategoricalAxisDirective;
}());
exports.DateTimeCategoricalAxisDirective = DateTimeCategoricalAxisDirective;
var DateTimeContinuousAxisDirective = /** @class */ (function () {
    function DateTimeContinuousAxisDirective() {
    }
    DateTimeContinuousAxisDirective = __decorate([
        core_1.Directive({
            selector: "DateTimeContinuousAxis"
        }),
        __metadata("design:paramtypes", [])
    ], DateTimeContinuousAxisDirective);
    return DateTimeContinuousAxisDirective;
}());
exports.DateTimeContinuousAxisDirective = DateTimeContinuousAxisDirective;
var LogarithmicAxisDirective = /** @class */ (function () {
    function LogarithmicAxisDirective() {
    }
    LogarithmicAxisDirective = __decorate([
        core_1.Directive({
            selector: "LogarithmicAxis"
        }),
        __metadata("design:paramtypes", [])
    ], LogarithmicAxisDirective);
    return LogarithmicAxisDirective;
}());
exports.LogarithmicAxisDirective = LogarithmicAxisDirective;
var LineSeriesDirective = /** @class */ (function (_super) {
    __extends(LineSeriesDirective, _super);
    function LineSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    LineSeriesDirective = __decorate([
        core_1.Directive({
            selector: "LineSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], LineSeriesDirective);
    return LineSeriesDirective;
}(SeriesComponentBase));
exports.LineSeriesDirective = LineSeriesDirective;
var AreaSeriesDirective = /** @class */ (function (_super) {
    __extends(AreaSeriesDirective, _super);
    function AreaSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    AreaSeriesDirective = __decorate([
        core_1.Directive({
            selector: "AreaSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], AreaSeriesDirective);
    return AreaSeriesDirective;
}(SeriesComponentBase));
exports.AreaSeriesDirective = AreaSeriesDirective;
var SplineSeriesDirective = /** @class */ (function (_super) {
    __extends(SplineSeriesDirective, _super);
    function SplineSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    SplineSeriesDirective = __decorate([
        core_1.Directive({
            selector: "SplineSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], SplineSeriesDirective);
    return SplineSeriesDirective;
}(SeriesComponentBase));
exports.SplineSeriesDirective = SplineSeriesDirective;
var SplineAreaSeriesDirective = /** @class */ (function (_super) {
    __extends(SplineAreaSeriesDirective, _super);
    function SplineAreaSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    SplineAreaSeriesDirective = __decorate([
        core_1.Directive({
            selector: "SplineAreaSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], SplineAreaSeriesDirective);
    return SplineAreaSeriesDirective;
}(SeriesComponentBase));
exports.SplineAreaSeriesDirective = SplineAreaSeriesDirective;
var BarSeriesDirective = /** @class */ (function (_super) {
    __extends(BarSeriesDirective, _super);
    function BarSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    BarSeriesDirective = __decorate([
        core_1.Directive({
            selector: "BarSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], BarSeriesDirective);
    return BarSeriesDirective;
}(SeriesComponentBase));
exports.BarSeriesDirective = BarSeriesDirective;
var RangeBarSeriesDirective = /** @class */ (function (_super) {
    __extends(RangeBarSeriesDirective, _super);
    function RangeBarSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    RangeBarSeriesDirective = __decorate([
        core_1.Directive({
            selector: "RangeBarSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], RangeBarSeriesDirective);
    return RangeBarSeriesDirective;
}(SeriesComponentBase));
exports.RangeBarSeriesDirective = RangeBarSeriesDirective;
var BubbleSeriesDirective = /** @class */ (function (_super) {
    __extends(BubbleSeriesDirective, _super);
    function BubbleSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    BubbleSeriesDirective = __decorate([
        core_1.Directive({
            selector: "BubbleSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], BubbleSeriesDirective);
    return BubbleSeriesDirective;
}(SeriesComponentBase));
exports.BubbleSeriesDirective = BubbleSeriesDirective;
var ScatterBubbleSeriesDirective = /** @class */ (function (_super) {
    __extends(ScatterBubbleSeriesDirective, _super);
    function ScatterBubbleSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    ScatterBubbleSeriesDirective = __decorate([
        core_1.Directive({
            selector: "ScatterBubbleSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], ScatterBubbleSeriesDirective);
    return ScatterBubbleSeriesDirective;
}(SeriesComponentBase));
exports.ScatterBubbleSeriesDirective = ScatterBubbleSeriesDirective;
var ScatterSeriesDirective = /** @class */ (function (_super) {
    __extends(ScatterSeriesDirective, _super);
    function ScatterSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    ScatterSeriesDirective = __decorate([
        core_1.Directive({
            selector: "ScatterSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], ScatterSeriesDirective);
    return ScatterSeriesDirective;
}(SeriesComponentBase));
exports.ScatterSeriesDirective = ScatterSeriesDirective;
var PaletteDirective = /** @class */ (function (_super) {
    __extends(PaletteDirective, _super);
    function PaletteDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    PaletteDirective = __decorate([
        core_1.Directive({
            selector: "Palette"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], PaletteDirective);
    return PaletteDirective;
}(SeriesComponentBase));
exports.PaletteDirective = PaletteDirective;
var PieSeriesDirective = /** @class */ (function (_super) {
    __extends(PieSeriesDirective, _super);
    function PieSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    PieSeriesDirective = __decorate([
        core_1.Directive({
            selector: "PieSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], PieSeriesDirective);
    return PieSeriesDirective;
}(SeriesComponentBase));
exports.PieSeriesDirective = PieSeriesDirective;
var DonutSeriesDirective = /** @class */ (function (_super) {
    __extends(DonutSeriesDirective, _super);
    function DonutSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    DonutSeriesDirective = __decorate([
        core_1.Directive({
            selector: "DonutSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], DonutSeriesDirective);
    return DonutSeriesDirective;
}(SeriesComponentBase));
exports.DonutSeriesDirective = DonutSeriesDirective;
var CandlestickSeriesDirective = /** @class */ (function (_super) {
    __extends(CandlestickSeriesDirective, _super);
    function CandlestickSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    CandlestickSeriesDirective = __decorate([
        core_1.Directive({
            selector: "CandlestickSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], CandlestickSeriesDirective);
    return CandlestickSeriesDirective;
}(SeriesComponentBase));
exports.CandlestickSeriesDirective = CandlestickSeriesDirective;
var OhlcSeriesDirective = /** @class */ (function (_super) {
    __extends(OhlcSeriesDirective, _super);
    function OhlcSeriesDirective(_elementRef) {
        var _this = _super.call(this) || this;
        _this._elementRef = _elementRef;
        _this._nativeSeries = _this._elementRef.nativeElement;
        return _this;
    }
    OhlcSeriesDirective = __decorate([
        core_1.Directive({
            selector: "OhlcSeries"
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], OhlcSeriesDirective);
    return OhlcSeriesDirective;
}(SeriesComponentBase));
exports.OhlcSeriesDirective = OhlcSeriesDirective;
var RadCartesianChartGridDirective = /** @class */ (function () {
    function RadCartesianChartGridDirective() {
    }
    RadCartesianChartGridDirective = __decorate([
        core_1.Directive({
            selector: "RadCartesianChartGrid"
        }),
        __metadata("design:paramtypes", [])
    ], RadCartesianChartGridDirective);
    return RadCartesianChartGridDirective;
}());
exports.RadCartesianChartGridDirective = RadCartesianChartGridDirective;
var ChartGridLineAnnotationDirective = /** @class */ (function () {
    function ChartGridLineAnnotationDirective() {
    }
    ChartGridLineAnnotationDirective = __decorate([
        core_1.Directive({
            selector: "ChartGridLineAnnotation"
        }),
        __metadata("design:paramtypes", [])
    ], ChartGridLineAnnotationDirective);
    return ChartGridLineAnnotationDirective;
}());
exports.ChartGridLineAnnotationDirective = ChartGridLineAnnotationDirective;
var ChartPlotBandAnnotationDirective = /** @class */ (function () {
    function ChartPlotBandAnnotationDirective() {
    }
    ChartPlotBandAnnotationDirective = __decorate([
        core_1.Directive({
            selector: "ChartPlotBandAnnotation"
        }),
        __metadata("design:paramtypes", [])
    ], ChartPlotBandAnnotationDirective);
    return ChartPlotBandAnnotationDirective;
}());
exports.ChartPlotBandAnnotationDirective = ChartPlotBandAnnotationDirective;
var TrackballDirective = /** @class */ (function () {
    function TrackballDirective() {
    }
    TrackballDirective = __decorate([
        core_1.Directive({
            selector: "Trackball"
        }),
        __metadata("design:paramtypes", [])
    ], TrackballDirective);
    return TrackballDirective;
}());
exports.TrackballDirective = TrackballDirective;
var PointLabelStyleDirective = /** @class */ (function () {
    function PointLabelStyleDirective() {
    }
    PointLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "PointLabelStyle"
        }),
        __metadata("design:paramtypes", [])
    ], PointLabelStyleDirective);
    return PointLabelStyleDirective;
}());
exports.PointLabelStyleDirective = PointLabelStyleDirective;
var TKPieLabelStyleDirective = /** @class */ (function () {
    function TKPieLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKPieLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKPieLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkPieLabelStyle]"
        }),
        __param(0, core_1.Inject(PieSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [PieSeriesDirective,
            core_1.ElementRef])
    ], TKPieLabelStyleDirective);
    return TKPieLabelStyleDirective;
}());
exports.TKPieLabelStyleDirective = TKPieLabelStyleDirective;
var TKDonutLabelStyleDirective = /** @class */ (function () {
    function TKDonutLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKDonutLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKDonutLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkDonutLabelStyle]"
        }),
        __param(0, core_1.Inject(DonutSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [DonutSeriesDirective,
            core_1.ElementRef])
    ], TKDonutLabelStyleDirective);
    return TKDonutLabelStyleDirective;
}());
exports.TKDonutLabelStyleDirective = TKDonutLabelStyleDirective;
var TKLineLabelStyleDirective = /** @class */ (function () {
    function TKLineLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKLineLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKLineLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkLineLabelStyle]"
        }),
        __param(0, core_1.Inject(LineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [LineSeriesDirective,
            core_1.ElementRef])
    ], TKLineLabelStyleDirective);
    return TKLineLabelStyleDirective;
}());
exports.TKLineLabelStyleDirective = TKLineLabelStyleDirective;
var TKBarLabelStyleDirective = /** @class */ (function () {
    function TKBarLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBarLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKBarLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkBarLabelStyle]"
        }),
        __param(0, core_1.Inject(BarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BarSeriesDirective,
            core_1.ElementRef])
    ], TKBarLabelStyleDirective);
    return TKBarLabelStyleDirective;
}());
exports.TKBarLabelStyleDirective = TKBarLabelStyleDirective;
var TKRangeBarLabelStyleDirective = /** @class */ (function () {
    function TKRangeBarLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRangeBarLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKRangeBarLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkRangeBarLabelStyle]"
        }),
        __param(0, core_1.Inject(RangeBarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RangeBarSeriesDirective,
            core_1.ElementRef])
    ], TKRangeBarLabelStyleDirective);
    return TKRangeBarLabelStyleDirective;
}());
exports.TKRangeBarLabelStyleDirective = TKRangeBarLabelStyleDirective;
var TKAreaLabelStyleDirective = /** @class */ (function () {
    function TKAreaLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKAreaLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKAreaLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkAreaLabelStyle]"
        }),
        __param(0, core_1.Inject(AreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [AreaSeriesDirective,
            core_1.ElementRef])
    ], TKAreaLabelStyleDirective);
    return TKAreaLabelStyleDirective;
}());
exports.TKAreaLabelStyleDirective = TKAreaLabelStyleDirective;
var TKSplineLabelStyleDirective = /** @class */ (function () {
    function TKSplineLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKSplineLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineLabelStyle]",
        }),
        __param(0, core_1.Inject(SplineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineSeriesDirective,
            core_1.ElementRef])
    ], TKSplineLabelStyleDirective);
    return TKSplineLabelStyleDirective;
}());
exports.TKSplineLabelStyleDirective = TKSplineLabelStyleDirective;
var TKSplineAreaLabelStyleDirective = /** @class */ (function () {
    function TKSplineAreaLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineAreaLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKSplineAreaLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineAreaLabelStyle]",
        }),
        __param(0, core_1.Inject(SplineAreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineAreaSeriesDirective,
            core_1.ElementRef])
    ], TKSplineAreaLabelStyleDirective);
    return TKSplineAreaLabelStyleDirective;
}());
exports.TKSplineAreaLabelStyleDirective = TKSplineAreaLabelStyleDirective;
var TKBubbleLabelStyleDirective = /** @class */ (function () {
    function TKBubbleLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBubbleLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKBubbleLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkBubbleLabelStyle]"
        }),
        __param(0, core_1.Inject(BubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BubbleSeriesDirective,
            core_1.ElementRef])
    ], TKBubbleLabelStyleDirective);
    return TKBubbleLabelStyleDirective;
}());
exports.TKBubbleLabelStyleDirective = TKBubbleLabelStyleDirective;
var TKScatterBubbleLabelStyleDirective = /** @class */ (function () {
    function TKScatterBubbleLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterBubbleLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKScatterBubbleLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterBubbleLabelStyle]"
        }),
        __param(0, core_1.Inject(ScatterBubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterBubbleSeriesDirective,
            core_1.ElementRef])
    ], TKScatterBubbleLabelStyleDirective);
    return TKScatterBubbleLabelStyleDirective;
}());
exports.TKScatterBubbleLabelStyleDirective = TKScatterBubbleLabelStyleDirective;
var TKCandlestickLabelStyleDirective = /** @class */ (function () {
    function TKCandlestickLabelStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCandlestickLabelStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKCandlestickLabelStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkCandlestickLabelStyle]"
        }),
        __param(0, core_1.Inject(CandlestickSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [CandlestickSeriesDirective,
            core_1.ElementRef])
    ], TKCandlestickLabelStyleDirective);
    return TKCandlestickLabelStyleDirective;
}());
exports.TKCandlestickLabelStyleDirective = TKCandlestickLabelStyleDirective;
var TKOhlcStyleDirective = /** @class */ (function () {
    function TKOhlcStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKOhlcStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKOhlcStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkOhlcLabelStyle]"
        }),
        __param(0, core_1.Inject(OhlcSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [OhlcSeriesDirective,
            core_1.ElementRef])
    ], TKOhlcStyleDirective);
    return TKOhlcStyleDirective;
}());
exports.TKOhlcStyleDirective = TKOhlcStyleDirective;
var TKScatterStyleDirective = /** @class */ (function () {
    function TKScatterStyleDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterStyleDirective.prototype.ngOnInit = function () {
        var labelStyle = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.labelStyle = labelStyle;
    };
    TKScatterStyleDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterLabelStyle]"
        }),
        __param(0, core_1.Inject(ScatterSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterSeriesDirective,
            core_1.ElementRef])
    ], TKScatterStyleDirective);
    return TKScatterStyleDirective;
}());
exports.TKScatterStyleDirective = TKScatterStyleDirective;
var TKCartesianGridDirective = /** @class */ (function () {
    function TKCartesianGridDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianGridDirective.prototype.ngOnInit = function () {
        var grid = this._elementRef.nativeElement;
        var cartesianChart = this.owner.cartesianChart;
        cartesianChart.grid = grid;
    };
    TKCartesianGridDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianGrid]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianGridDirective);
    return TKCartesianGridDirective;
}());
exports.TKCartesianGridDirective = TKCartesianGridDirective;
var TKCartesianSeriesDirective = /** @class */ (function () {
    function TKCartesianSeriesDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianSeriesDirective.prototype.ngOnInit = function () {
        var series = this._elementRef.nativeElement;
        if (this.owner.cartesianChart.series) {
            this.owner.cartesianChart.series.push(series);
        }
        else {
            this.owner.cartesianChart.series = new observable_array_1.ObservableArray([series]);
        }
    };
    TKCartesianSeriesDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianSeries]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianSeriesDirective);
    return TKCartesianSeriesDirective;
}());
exports.TKCartesianSeriesDirective = TKCartesianSeriesDirective;
var TKPieChartSeriesDirective = /** @class */ (function () {
    function TKPieChartSeriesDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKPieChartSeriesDirective.prototype.ngOnInit = function () {
        var series = this._elementRef.nativeElement;
        if (this.owner.pieChart.series) {
            this.owner.pieChart.series.push(series);
        }
        else {
            this.owner.pieChart.series = new observable_array_1.ObservableArray([series]);
        }
    };
    TKPieChartSeriesDirective = __decorate([
        core_1.Directive({
            selector: "[tkPieSeries]"
        }),
        __param(0, core_1.Inject(RadPieChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadPieChartComponent,
            core_1.ElementRef])
    ], TKPieChartSeriesDirective);
    return TKPieChartSeriesDirective;
}());
exports.TKPieChartSeriesDirective = TKPieChartSeriesDirective;
var TKCartesianPaletteDirective = /** @class */ (function () {
    function TKCartesianPaletteDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianPaletteDirective.prototype.ngOnInit = function () {
        this._nativePalette = this._elementRef.nativeElement;
        if (this.owner.cartesianChart.palettes) {
            this.owner.cartesianChart.palettes.push(this._nativePalette);
        }
        else {
            this.owner.cartesianChart.palettes = new observable_array_1.ObservableArray([this._nativePalette]);
        }
    };
    Object.defineProperty(TKCartesianPaletteDirective.prototype, "nativePalette", {
        get: function () {
            return this._nativePalette;
        },
        enumerable: true,
        configurable: true
    });
    TKCartesianPaletteDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianPalette]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianPaletteDirective);
    return TKCartesianPaletteDirective;
}());
exports.TKCartesianPaletteDirective = TKCartesianPaletteDirective;
var TKPiePaletteDirective = /** @class */ (function () {
    function TKPiePaletteDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKPiePaletteDirective.prototype.ngOnInit = function () {
        this._nativePalette = this._elementRef.nativeElement;
        if (this.owner.pieChart.palettes) {
            this.owner.pieChart.palettes.push(this._nativePalette);
        }
        else {
            this.owner.pieChart.palettes = new observable_array_1.ObservableArray([this._nativePalette]);
        }
    };
    Object.defineProperty(TKPiePaletteDirective.prototype, "nativePalette", {
        get: function () {
            return this._nativePalette;
        },
        enumerable: true,
        configurable: true
    });
    TKPiePaletteDirective = __decorate([
        core_1.Directive({
            selector: "[tkPiePalette]"
        }),
        __param(0, core_1.Inject(RadPieChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadPieChartComponent,
            core_1.ElementRef])
    ], TKPiePaletteDirective);
    return TKPiePaletteDirective;
}());
exports.TKPiePaletteDirective = TKPiePaletteDirective;
var TKCartesianPaletteEntryDirective = /** @class */ (function () {
    function TKCartesianPaletteEntryDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianPaletteEntryDirective.prototype.ngOnInit = function () {
        var entry = this._elementRef.nativeElement;
        if (this.owner.nativePalette.entries) {
            this.owner.nativePalette.entries.push(entry);
            this.owner.nativePalette.updateOwner();
        }
        else {
            this.owner.nativePalette.entries = new observable_array_1.ObservableArray([entry]);
        }
    };
    TKCartesianPaletteEntryDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianPaletteEntry]"
        }),
        __param(0, core_1.Inject(TKCartesianPaletteDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKCartesianPaletteDirective,
            core_1.ElementRef])
    ], TKCartesianPaletteEntryDirective);
    return TKCartesianPaletteEntryDirective;
}());
exports.TKCartesianPaletteEntryDirective = TKCartesianPaletteEntryDirective;
var TKPiePaletteEntryDirective = /** @class */ (function () {
    function TKPiePaletteEntryDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKPiePaletteEntryDirective.prototype.ngOnInit = function () {
        var entry = this._elementRef.nativeElement;
        if (this.owner.nativePalette.entries) {
            this.owner.nativePalette.entries.push(entry);
            this.owner.nativePalette.updateOwner();
        }
        else {
            this.owner.nativePalette.entries = new observable_array_1.ObservableArray([entry]);
        }
    };
    TKPiePaletteEntryDirective = __decorate([
        core_1.Directive({
            selector: "[tkPiePaletteEntry]"
        }),
        __param(0, core_1.Inject(TKPiePaletteDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [TKPiePaletteDirective,
            core_1.ElementRef])
    ], TKPiePaletteEntryDirective);
    return TKPiePaletteEntryDirective;
}());
exports.TKPiePaletteEntryDirective = TKPiePaletteEntryDirective;
var TKCartesianHorizontalAxisDirective = /** @class */ (function () {
    function TKCartesianHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianHorizontalAxisDirective.prototype.ngOnInit = function () {
        var axis = this._elementRef.nativeElement;
        var cartesianChart = this.owner.cartesianChart;
        cartesianChart.horizontalAxis = axis;
    };
    TKCartesianHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianHorizontalAxis]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianHorizontalAxisDirective);
    return TKCartesianHorizontalAxisDirective;
}());
exports.TKCartesianHorizontalAxisDirective = TKCartesianHorizontalAxisDirective;
var TKCartesianVerticalAxisDirective = /** @class */ (function () {
    function TKCartesianVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianVerticalAxisDirective.prototype.ngOnInit = function () {
        var axis = this._elementRef.nativeElement;
        var cartesianChart = this.owner.cartesianChart;
        cartesianChart.verticalAxis = axis;
    };
    TKCartesianVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianVerticalAxis]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianVerticalAxisDirective);
    return TKCartesianVerticalAxisDirective;
}());
exports.TKCartesianVerticalAxisDirective = TKCartesianVerticalAxisDirective;
var TKLineVerticalAxisDirective = /** @class */ (function () {
    function TKLineVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKLineVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKLineVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkLineVerticalAxis]"
        }),
        __param(0, core_1.Inject(LineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [LineSeriesDirective,
            core_1.ElementRef])
    ], TKLineVerticalAxisDirective);
    return TKLineVerticalAxisDirective;
}());
exports.TKLineVerticalAxisDirective = TKLineVerticalAxisDirective;
var TKLineHorizontalAxisDirective = /** @class */ (function () {
    function TKLineHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKLineHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKLineHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkLineHorizontalAxis]"
        }),
        __param(0, core_1.Inject(LineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [LineSeriesDirective,
            core_1.ElementRef])
    ], TKLineHorizontalAxisDirective);
    return TKLineHorizontalAxisDirective;
}());
exports.TKLineHorizontalAxisDirective = TKLineHorizontalAxisDirective;
var TKBarVerticalAxisDirective = /** @class */ (function () {
    function TKBarVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBarVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKBarVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkBarVerticalAxis]"
        }),
        __param(0, core_1.Inject(BarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BarSeriesDirective,
            core_1.ElementRef])
    ], TKBarVerticalAxisDirective);
    return TKBarVerticalAxisDirective;
}());
exports.TKBarVerticalAxisDirective = TKBarVerticalAxisDirective;
var TKBarHorizontalAxisDirective = /** @class */ (function () {
    function TKBarHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBarHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKBarHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkBarHorizontalAxis]"
        }),
        __param(0, core_1.Inject(BarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BarSeriesDirective,
            core_1.ElementRef])
    ], TKBarHorizontalAxisDirective);
    return TKBarHorizontalAxisDirective;
}());
exports.TKBarHorizontalAxisDirective = TKBarHorizontalAxisDirective;
var TKRangeBarVerticalAxisDirective = /** @class */ (function () {
    function TKRangeBarVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRangeBarVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKRangeBarVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkRangeBarVerticalAxis]"
        }),
        __param(0, core_1.Inject(RangeBarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RangeBarSeriesDirective,
            core_1.ElementRef])
    ], TKRangeBarVerticalAxisDirective);
    return TKRangeBarVerticalAxisDirective;
}());
exports.TKRangeBarVerticalAxisDirective = TKRangeBarVerticalAxisDirective;
var TKRangeBarHorizontalAxisDirective = /** @class */ (function () {
    function TKRangeBarHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKRangeBarHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKRangeBarHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkRangeBarHorizontalAxis]"
        }),
        __param(0, core_1.Inject(RangeBarSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RangeBarSeriesDirective,
            core_1.ElementRef])
    ], TKRangeBarHorizontalAxisDirective);
    return TKRangeBarHorizontalAxisDirective;
}());
exports.TKRangeBarHorizontalAxisDirective = TKRangeBarHorizontalAxisDirective;
var TKAreaVerticalAxisDirective = /** @class */ (function () {
    function TKAreaVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKAreaVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKAreaVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkAreaVerticalAxis]"
        }),
        __param(0, core_1.Inject(AreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [AreaSeriesDirective,
            core_1.ElementRef])
    ], TKAreaVerticalAxisDirective);
    return TKAreaVerticalAxisDirective;
}());
exports.TKAreaVerticalAxisDirective = TKAreaVerticalAxisDirective;
var TKAreaHorizontalAxisDirective = /** @class */ (function () {
    function TKAreaHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKAreaHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKAreaHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkAreaHorizontalAxis]"
        }),
        __param(0, core_1.Inject(AreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [AreaSeriesDirective,
            core_1.ElementRef])
    ], TKAreaHorizontalAxisDirective);
    return TKAreaHorizontalAxisDirective;
}());
exports.TKAreaHorizontalAxisDirective = TKAreaHorizontalAxisDirective;
var TKSplineVerticalAxisDirective = /** @class */ (function () {
    function TKSplineVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKSplineVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineVerticalAxis]"
        }),
        __param(0, core_1.Inject(SplineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineSeriesDirective,
            core_1.ElementRef])
    ], TKSplineVerticalAxisDirective);
    return TKSplineVerticalAxisDirective;
}());
exports.TKSplineVerticalAxisDirective = TKSplineVerticalAxisDirective;
var TKSplineHorizontalAxisDirective = /** @class */ (function () {
    function TKSplineHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKSplineHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineHorizontalAxis]"
        }),
        __param(0, core_1.Inject(SplineSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineSeriesDirective,
            core_1.ElementRef])
    ], TKSplineHorizontalAxisDirective);
    return TKSplineHorizontalAxisDirective;
}());
exports.TKSplineHorizontalAxisDirective = TKSplineHorizontalAxisDirective;
var TKSplineAreaVerticalAxisDirective = /** @class */ (function () {
    function TKSplineAreaVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineAreaVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKSplineAreaVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineAreaVerticalAxis]"
        }),
        __param(0, core_1.Inject(SplineAreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineAreaSeriesDirective,
            core_1.ElementRef])
    ], TKSplineAreaVerticalAxisDirective);
    return TKSplineAreaVerticalAxisDirective;
}());
exports.TKSplineAreaVerticalAxisDirective = TKSplineAreaVerticalAxisDirective;
var TKSplineAreaHorizontalAxisDirective = /** @class */ (function () {
    function TKSplineAreaHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKSplineAreaHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKSplineAreaHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkSplineAreaHorizontalAxis]"
        }),
        __param(0, core_1.Inject(SplineAreaSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [SplineAreaSeriesDirective,
            core_1.ElementRef])
    ], TKSplineAreaHorizontalAxisDirective);
    return TKSplineAreaHorizontalAxisDirective;
}());
exports.TKSplineAreaHorizontalAxisDirective = TKSplineAreaHorizontalAxisDirective;
var TKBubbleVerticalAxisDirective = /** @class */ (function () {
    function TKBubbleVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBubbleVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKBubbleVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkBubbleVerticalAxis]"
        }),
        __param(0, core_1.Inject(BubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BubbleSeriesDirective,
            core_1.ElementRef])
    ], TKBubbleVerticalAxisDirective);
    return TKBubbleVerticalAxisDirective;
}());
exports.TKBubbleVerticalAxisDirective = TKBubbleVerticalAxisDirective;
var TKBubbleHorizontalAxisDirective = /** @class */ (function () {
    function TKBubbleHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKBubbleHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKBubbleHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkBubbleHorizontalAxis]"
        }),
        __param(0, core_1.Inject(BubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [BubbleSeriesDirective,
            core_1.ElementRef])
    ], TKBubbleHorizontalAxisDirective);
    return TKBubbleHorizontalAxisDirective;
}());
exports.TKBubbleHorizontalAxisDirective = TKBubbleHorizontalAxisDirective;
var TKScatterBubbleVerticalAxisDirective = /** @class */ (function () {
    function TKScatterBubbleVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterBubbleVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKScatterBubbleVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterBubbleVerticalAxis]"
        }),
        __param(0, core_1.Inject(ScatterBubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterBubbleSeriesDirective,
            core_1.ElementRef])
    ], TKScatterBubbleVerticalAxisDirective);
    return TKScatterBubbleVerticalAxisDirective;
}());
exports.TKScatterBubbleVerticalAxisDirective = TKScatterBubbleVerticalAxisDirective;
var TKScatterBubbleHorizontalAxisDirective = /** @class */ (function () {
    function TKScatterBubbleHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterBubbleHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKScatterBubbleHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterBubbleHorizontalAxis]"
        }),
        __param(0, core_1.Inject(ScatterBubbleSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterBubbleSeriesDirective,
            core_1.ElementRef])
    ], TKScatterBubbleHorizontalAxisDirective);
    return TKScatterBubbleHorizontalAxisDirective;
}());
exports.TKScatterBubbleHorizontalAxisDirective = TKScatterBubbleHorizontalAxisDirective;
var TKCandlestickVerticalAxisDirective = /** @class */ (function () {
    function TKCandlestickVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCandlestickVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKCandlestickVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkCandlestickVerticalAxis]"
        }),
        __param(0, core_1.Inject(CandlestickSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [CandlestickSeriesDirective,
            core_1.ElementRef])
    ], TKCandlestickVerticalAxisDirective);
    return TKCandlestickVerticalAxisDirective;
}());
exports.TKCandlestickVerticalAxisDirective = TKCandlestickVerticalAxisDirective;
var TKCandlestickHorizontalAxisDirective = /** @class */ (function () {
    function TKCandlestickHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCandlestickHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKCandlestickHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkCandlestickHorizontalAxis]"
        }),
        __param(0, core_1.Inject(CandlestickSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [CandlestickSeriesDirective,
            core_1.ElementRef])
    ], TKCandlestickHorizontalAxisDirective);
    return TKCandlestickHorizontalAxisDirective;
}());
exports.TKCandlestickHorizontalAxisDirective = TKCandlestickHorizontalAxisDirective;
var TKOhlcVerticalAxisDirective = /** @class */ (function () {
    function TKOhlcVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKOhlcVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKOhlcVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkOhlcVerticalAxis]"
        }),
        __param(0, core_1.Inject(OhlcSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [OhlcSeriesDirective,
            core_1.ElementRef])
    ], TKOhlcVerticalAxisDirective);
    return TKOhlcVerticalAxisDirective;
}());
exports.TKOhlcVerticalAxisDirective = TKOhlcVerticalAxisDirective;
var TKOhlcHorizontalAxisDirective = /** @class */ (function () {
    function TKOhlcHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKOhlcHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKOhlcHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkOhlcHorizontalAxis]"
        }),
        __param(0, core_1.Inject(OhlcSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [OhlcSeriesDirective,
            core_1.ElementRef])
    ], TKOhlcHorizontalAxisDirective);
    return TKOhlcHorizontalAxisDirective;
}());
exports.TKOhlcHorizontalAxisDirective = TKOhlcHorizontalAxisDirective;
var TKScatterVerticalAxisDirective = /** @class */ (function () {
    function TKScatterVerticalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterVerticalAxisDirective.prototype.ngOnInit = function () {
        var verticalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.verticalAxis = verticalAxis;
    };
    TKScatterVerticalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterVerticalAxis]"
        }),
        __param(0, core_1.Inject(ScatterSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterSeriesDirective,
            core_1.ElementRef])
    ], TKScatterVerticalAxisDirective);
    return TKScatterVerticalAxisDirective;
}());
exports.TKScatterVerticalAxisDirective = TKScatterVerticalAxisDirective;
var TKScatterHorizontalAxisDirective = /** @class */ (function () {
    function TKScatterHorizontalAxisDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKScatterHorizontalAxisDirective.prototype.ngOnInit = function () {
        var horizontalAxis = this._elementRef.nativeElement;
        var series = this.owner.nativeSeries;
        series.horizontalAxis = horizontalAxis;
    };
    TKScatterHorizontalAxisDirective = __decorate([
        core_1.Directive({
            selector: "[tkScatterHorizontalAxis]"
        }),
        __param(0, core_1.Inject(ScatterSeriesDirective)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [ScatterSeriesDirective,
            core_1.ElementRef])
    ], TKScatterHorizontalAxisDirective);
    return TKScatterHorizontalAxisDirective;
}());
exports.TKScatterHorizontalAxisDirective = TKScatterHorizontalAxisDirective;
var TKPieLegendDirective = /** @class */ (function () {
    function TKPieLegendDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKPieLegendDirective.prototype.ngOnInit = function () {
        var legend = this._elementRef.nativeElement;
        var pieChart = this.owner.pieChart;
        pieChart.legend = legend;
    };
    TKPieLegendDirective = __decorate([
        core_1.Directive({
            selector: "[tkPieLegend]",
        }),
        __param(0, core_1.Inject(RadPieChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadPieChartComponent,
            core_1.ElementRef])
    ], TKPieLegendDirective);
    return TKPieLegendDirective;
}());
exports.TKPieLegendDirective = TKPieLegendDirective;
var TKCartesianLegendDirective = /** @class */ (function () {
    function TKCartesianLegendDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianLegendDirective.prototype.ngOnInit = function () {
        var legend = this._elementRef.nativeElement;
        var cartesianChart = this.owner.cartesianChart;
        cartesianChart.legend = legend;
    };
    TKCartesianLegendDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianLegend]",
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianLegendDirective);
    return TKCartesianLegendDirective;
}());
exports.TKCartesianLegendDirective = TKCartesianLegendDirective;
var TKCartesianTrackballDirective = /** @class */ (function () {
    function TKCartesianTrackballDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianTrackballDirective.prototype.ngOnInit = function () {
        var trackball = this._elementRef.nativeElement;
        var cartesianChart = this.owner.cartesianChart;
        cartesianChart.trackball = trackball;
    };
    TKCartesianTrackballDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianTrackball]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianTrackballDirective);
    return TKCartesianTrackballDirective;
}());
exports.TKCartesianTrackballDirective = TKCartesianTrackballDirective;
var TKCartesianAnnotationsDirective = /** @class */ (function () {
    function TKCartesianAnnotationsDirective(owner, _elementRef) {
        this.owner = owner;
        this._elementRef = _elementRef;
    }
    TKCartesianAnnotationsDirective.prototype.ngOnInit = function () {
        var annotation = this._elementRef.nativeElement;
        if (this.owner.cartesianChart.annotations) {
            this.owner.cartesianChart.annotations.push(annotation);
        }
        else {
            this.owner.cartesianChart.annotations = new observable_array_1.ObservableArray([annotation]);
        }
    };
    TKCartesianAnnotationsDirective = __decorate([
        core_1.Directive({
            selector: "[tkCartesianAnnotations]"
        }),
        __param(0, core_1.Inject(RadCartesianChartComponent)),
        __param(1, core_1.Inject(core_1.ElementRef)),
        __metadata("design:paramtypes", [RadCartesianChartComponent,
            core_1.ElementRef])
    ], TKCartesianAnnotationsDirective);
    return TKCartesianAnnotationsDirective;
}());
exports.TKCartesianAnnotationsDirective = TKCartesianAnnotationsDirective;
exports.CHART_DIRECTIVES = [RadCartesianChartComponent, RadPieChartComponent, LineSeriesDirective, AreaSeriesDirective, SplineSeriesDirective, SplineAreaSeriesDirective, BarSeriesDirective, RangeBarSeriesDirective, BubbleSeriesDirective, ScatterBubbleSeriesDirective, ScatterSeriesDirective, PieSeriesDirective, TKPieChartSeriesDirective, TKCartesianSeriesDirective, PieSeriesDirective, TKCartesianHorizontalAxisDirective, TKCartesianVerticalAxisDirective, LinearAxisDirective, DateTimeCategoricalAxisDirective, CategoricalAxisDirective, RadCartesianChartGridDirective, TKCartesianGridDirective, PaletteDirective, TKCartesianPaletteDirective, CandlestickSeriesDirective, OhlcSeriesDirective, RadLegendViewDirective, DonutSeriesDirective, TKPieLegendDirective, TKCartesianLegendDirective, ChartGridLineAnnotationDirective, TKCartesianAnnotationsDirective, ChartPlotBandAnnotationDirective, TKCartesianTrackballDirective, TrackballDirective, PointLabelStyleDirective, TKPieLabelStyleDirective, TKLineLabelStyleDirective, TKBarLabelStyleDirective, TKRangeBarLabelStyleDirective, TKAreaLabelStyleDirective, TKSplineLabelStyleDirective, TKBubbleLabelStyleDirective, TKScatterBubbleLabelStyleDirective, TKCandlestickLabelStyleDirective, TKOhlcStyleDirective, TKScatterStyleDirective, TKDonutLabelStyleDirective, DateTimeContinuousAxisDirective, TKCartesianPaletteEntryDirective, LogarithmicAxisDirective, TKLineVerticalAxisDirective, TKLineHorizontalAxisDirective, TKBarVerticalAxisDirective, TKBarHorizontalAxisDirective, TKRangeBarVerticalAxisDirective, TKRangeBarHorizontalAxisDirective, TKAreaVerticalAxisDirective, TKAreaHorizontalAxisDirective, TKSplineVerticalAxisDirective, TKSplineHorizontalAxisDirective, TKSplineAreaVerticalAxisDirective, TKSplineAreaHorizontalAxisDirective, TKBubbleVerticalAxisDirective, TKBubbleHorizontalAxisDirective, TKScatterBubbleVerticalAxisDirective, TKScatterBubbleHorizontalAxisDirective, TKCandlestickVerticalAxisDirective, TKCandlestickHorizontalAxisDirective, TKOhlcVerticalAxisDirective, TKOhlcHorizontalAxisDirective, TKScatterVerticalAxisDirective, TKScatterHorizontalAxisDirective, TKPiePaletteDirective, TKPiePaletteEntryDirective, TKSplineAreaLabelStyleDirective];
elementRegistryModule.registerElement("RadCartesianChart", function () { return chartModule.RadCartesianChart; });
elementRegistryModule.registerElement("RadPieChart", function () { return chartModule.RadPieChart; });
elementRegistryModule.registerElement("RadLegendView", function () { return chartModule.RadLegendView; });
elementRegistryModule.registerElement("LineSeries", function () { return chartModule.LineSeries; });
elementRegistryModule.registerElement("PieSeries", function () { return chartModule.PieSeries; });
elementRegistryModule.registerElement("DonutSeries", function () { return chartModule.DonutSeries; });
elementRegistryModule.registerElement("AreaSeries", function () { return chartModule.AreaSeries; });
elementRegistryModule.registerElement("CategoricalAxis", function () { return chartModule.CategoricalAxis; });
elementRegistryModule.registerElement("LinearAxis", function () { return chartModule.LinearAxis; });
elementRegistryModule.registerElement("DateTimeCategoricalAxis", function () { return chartModule.DateTimeCategoricalAxis; });
elementRegistryModule.registerElement("SplineSeries", function () { return chartModule.SplineSeries; });
elementRegistryModule.registerElement("BarSeries", function () { return chartModule.BarSeries; });
elementRegistryModule.registerElement("RangeBarSeries", function () { return chartModule.RangeBarSeries; });
elementRegistryModule.registerElement("BubbleSeries", function () { return chartModule.BubbleSeries; });
elementRegistryModule.registerElement("ScatterBubbleSeries", function () { return chartModule.ScatterBubbleSeries; });
elementRegistryModule.registerElement("ScatterSeries", function () { return chartModule.ScatterSeries; });
elementRegistryModule.registerElement("OhlcSeries", function () { return chartModule.OhlcSeries; });
elementRegistryModule.registerElement("CandlestickSeries", function () { return chartModule.CandlestickSeries; });
elementRegistryModule.registerElement("RadCartesianChartGrid", function () { return chartModule.RadCartesianChartGrid; });
elementRegistryModule.registerElement("Palette", function () { return chartModule.Palette; });
elementRegistryModule.registerElement("PaletteEntry", function () { return chartModule.PaletteEntry; });
elementRegistryModule.registerElement("ChartGridLineAnnotation", function () { return chartModule.ChartGridLineAnnotation; });
elementRegistryModule.registerElement("ChartPlotBandAnnotation", function () { return chartModule.ChartPlotBandAnnotation; });
elementRegistryModule.registerElement("Trackball", function () { return chartModule.Trackball; });
elementRegistryModule.registerElement("PointLabelStyle", function () { return chartModule.PointLabelStyle; });
elementRegistryModule.registerElement("DateTimeContinuousAxis", function () { return chartModule.DateTimeContinuousAxis; });
elementRegistryModule.registerElement("LogarithmicAxis", function () { return chartModule.LogarithmicAxis; });
elementRegistryModule.registerElement("SplineAreaSeries", function () { return chartModule.SplineAreaSeries; });
var NativeScriptUIChartModule = /** @class */ (function () {
    function NativeScriptUIChartModule() {
    }
    NativeScriptUIChartModule = __decorate([
        core_1.NgModule({
            declarations: [exports.CHART_DIRECTIVES],
            exports: [exports.CHART_DIRECTIVES]
        })
    ], NativeScriptUIChartModule);
    return NativeScriptUIChartModule;
}());
exports.NativeScriptUIChartModule = NativeScriptUIChartModule;
