// nativescript-image/image-common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var view_1 = require("@nativescript/core/ui/core/view");
var platform_1 = require("@nativescript/core/platform");
var CLogTypes;
(function (CLogTypes) {
    CLogTypes[CLogTypes["info"] = 0] = "info";
    CLogTypes[CLogTypes["warning"] = 1] = "warning";
    CLogTypes[CLogTypes["error"] = 2] = "error";
})(CLogTypes = exports.CLogTypes || (exports.CLogTypes = {}));
exports.debug = false;
function setDebug(value) {
    exports.debug = value;
}
exports.setDebug = setDebug;
exports.CLog = function (type) {
    if (type === void 0) { type = 0; }
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    if (exports.debug) {
        if (type === 0) {
            console.log.apply(console, ['[nativescript-image]'].concat(args));
        }
        else if (type === 1) {
            console.warn.apply(console, ['[nativescript-image]'].concat(args));
        }
        else if (type === 2) {
            console.error.apply(console, ['[nativescript-image]'].concat(args));
        }
    }
};
var ScaleType;
(function (ScaleType) {
    ScaleType["None"] = "none";
    ScaleType["Fill"] = "fill";
    ScaleType["AspectFill"] = "aspectFill";
    ScaleType["AspectFit"] = "aspectFit";
    ScaleType["Center"] = "center";
    ScaleType["CenterCrop"] = "centerCrop";
    ScaleType["CenterInside"] = "centerInside";
    ScaleType["FitCenter"] = "fitCenter";
    ScaleType["FitEnd"] = "fitEnd";
    ScaleType["FitStart"] = "fitStart";
    ScaleType["FitXY"] = "fitXY";
    ScaleType["FocusCrop"] = "focusCrop";
})(ScaleType = exports.ScaleType || (exports.ScaleType = {}));
var EventData = (function () {
    function EventData() {
    }
    Object.defineProperty(EventData.prototype, "eventName", {
        get: function () {
            return this._eventName;
        },
        set: function (value) {
            this._eventName = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EventData.prototype, "object", {
        get: function () {
            return this._object;
        },
        set: function (value) {
            this._object = value;
        },
        enumerable: true,
        configurable: true
    });
    return EventData;
}());
exports.EventData = EventData;
var ImageBase = (function (_super) {
    __extends(ImageBase, _super);
    function ImageBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ImageBase.prototype.handleImageProgress = function (value, totalSize) {
    };
    ImageBase.finalImageSetEvent = 'finalImageSet';
    ImageBase.failureEvent = 'failure';
    ImageBase.intermediateImageFailedEvent = 'intermediateImageFailed';
    ImageBase.intermediateImageSetEvent = 'intermediateImageSet';
    ImageBase.releaseEvent = 'release';
    ImageBase.submitEvent = 'submit';
    ImageBase.srcProperty = new view_1.Property({ name: 'src' });
    ImageBase.lowerResSrcProperty = new view_1.Property({ name: 'lowerResSrc' });
    ImageBase.placeholderImageUriProperty = new view_1.Property({ name: 'placeholderImageUri' });
    ImageBase.failureImageUriProperty = new view_1.Property({ name: 'failureImageUri' });
    ImageBase.stretchProperty = new view_1.Property({ name: 'stretch' });
    ImageBase.backgroundUriProperty = new view_1.Property({ name: 'backgroundUri' });
    ImageBase.progressiveRenderingEnabledProperty = new view_1.Property({ name: 'progressiveRenderingEnabled', valueConverter: view_1.booleanConverter });
    ImageBase.localThumbnailPreviewsEnabledProperty = new view_1.Property({ name: 'localThumbnailPreviewsEnabled', valueConverter: view_1.booleanConverter });
    ImageBase.showProgressBarProperty = new view_1.Property({ name: 'showProgressBar', valueConverter: view_1.booleanConverter });
    ImageBase.progressBarColorProperty = new view_1.Property({ name: 'progressBarColor', defaultValue: undefined });
    ImageBase.roundAsCircleProperty = new view_1.Property({ name: 'roundAsCircle', valueConverter: view_1.booleanConverter, affectsLayout: platform_1.isAndroid });
    ImageBase.roundTopLeftProperty = new view_1.Property({ name: 'roundTopLeft', defaultValue: undefined, valueConverter: view_1.booleanConverter, affectsLayout: platform_1.isAndroid });
    ImageBase.roundTopRightProperty = new view_1.Property({ name: 'roundTopRight', valueConverter: view_1.booleanConverter, affectsLayout: platform_1.isAndroid });
    ImageBase.roundBottomLeftProperty = new view_1.Property({ name: 'roundBottomLeft', valueConverter: view_1.booleanConverter, affectsLayout: platform_1.isAndroid });
    ImageBase.roundBottomRightProperty = new view_1.Property({ name: 'roundBottomRight', valueConverter: view_1.booleanConverter, affectsLayout: platform_1.isAndroid });
    ImageBase.roundedCornerRadiusProperty = new view_1.Property({ name: 'roundedCornerRadius', valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.blurRadiusProperty = new view_1.Property({ name: 'blurRadius', valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.blurDownSamplingProperty = new view_1.Property({ name: 'blurDownSampling', valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.autoPlayAnimationsProperty = new view_1.Property({ name: 'autoPlayAnimations', valueConverter: view_1.booleanConverter });
    ImageBase.tapToRetryEnabledProperty = new view_1.Property({ name: 'tapToRetryEnabled', valueConverter: view_1.booleanConverter });
    ImageBase.aspectRatioProperty = new view_1.Property({ name: 'aspectRatio', affectsLayout: true, valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.decodeWidthProperty = new view_1.Property({ name: 'decodeWidth', valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.decodeHeightProperty = new view_1.Property({ name: 'decodeHeight', valueConverter: function (v) { return parseFloat(v); } });
    ImageBase.tintColorProperty = new view_1.Property({ name: 'tintColor' });
    ImageBase.alwaysFadeProperty = new view_1.Property({ name: 'alwaysFade', valueConverter: view_1.booleanConverter, defaultValue: false });
    ImageBase.fadeDurationProperty = new view_1.Property({ name: 'fadeDuration', valueConverter: function (v) { return parseFloat(v); } });
    return ImageBase;
}(view_1.View));
exports.ImageBase = ImageBase;
ImageBase.srcProperty.register(ImageBase);
ImageBase.lowerResSrcProperty.register(ImageBase);
ImageBase.placeholderImageUriProperty.register(ImageBase);
ImageBase.failureImageUriProperty.register(ImageBase);
ImageBase.stretchProperty.register(ImageBase);
ImageBase.fadeDurationProperty.register(ImageBase);
ImageBase.backgroundUriProperty.register(ImageBase);
ImageBase.progressiveRenderingEnabledProperty.register(ImageBase);
ImageBase.localThumbnailPreviewsEnabledProperty.register(ImageBase);
ImageBase.showProgressBarProperty.register(ImageBase);
ImageBase.progressBarColorProperty.register(ImageBase);
ImageBase.roundAsCircleProperty.register(ImageBase);
ImageBase.roundTopLeftProperty.register(ImageBase);
ImageBase.roundTopRightProperty.register(ImageBase);
ImageBase.roundBottomLeftProperty.register(ImageBase);
ImageBase.roundBottomRightProperty.register(ImageBase);
ImageBase.roundedCornerRadiusProperty.register(ImageBase);
ImageBase.blurRadiusProperty.register(ImageBase);
ImageBase.blurDownSamplingProperty.register(ImageBase);
ImageBase.autoPlayAnimationsProperty.register(ImageBase);
ImageBase.tapToRetryEnabledProperty.register(ImageBase);
ImageBase.aspectRatioProperty.register(ImageBase);
ImageBase.decodeWidthProperty.register(ImageBase);
ImageBase.decodeHeightProperty.register(ImageBase);
ImageBase.alwaysFadeProperty.register(ImageBase);
