// kinvey-nativescript-sdk/lib/nativescript/init.js
Object.defineProperty(exports, "__esModule", { value: true });
var kinvey_js_sdk_1 = require("kinvey-js-sdk");
var nativescript_1 = __importDefault(require("pubnub/lib/nativescript"));
var HttpAdapter = __importStar(require("./httpAdapter"));
var SessionStore = __importStar(require("./sessionStore"));
var Popup = __importStar(require("./popup"));
var storage_1 = require("./storage");
var utils_1 = require("./utils");
function init(config) {
    var configFromPackageJson = utils_1.getDataFromPackageJson();
    var mergedConfig = Object.assign({}, configFromPackageJson, config);
    var kinveyConfig = kinvey_js_sdk_1.init({
        kinveyConfig: mergedConfig,
        httpAdapter: HttpAdapter,
        sessionStore: SessionStore,
        popup: Popup,
        storageAdapter: storage_1.getStorageAdapter(mergedConfig.storage),
        pubnub: nativescript_1.default
    });
    return Object.assign({}, kinveyConfig, { storage: mergedConfig.storage, _storage: mergedConfig.storage });
}
exports.init = init;
function initialize(config) {
    return init(config);
}
exports.initialize = initialize;
