// nativescript-ui-chart/initializers/chart-initializers.ios.js
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("tns-core-modules/color");
var chart_axis_common_1 = require("../visualization/views/chart-axis.common");
var font_1 = require("tns-core-modules/ui/styling/font");
var chart_public_enum_1 = require("../misc/chart-public-enum");
var chart_series_common_1 = require("../visualization/views/chart-series.common");
var utils = require("tns-core-modules/utils/utils");
var chart_native_helper_1 = require("../misc/chart-native-helper");
var ChartBaseValueMapper = /** @class */ (function () {
    function ChartBaseValueMapper() {
    }
    ChartBaseValueMapper.prototype.onLegendChanged = function (oldValue, newValue, chart) {
    };
    ChartBaseValueMapper.prototype.onPalettesChanged = function (data, chart) {
    };
    ChartBaseValueMapper.prototype.onPalettesPropertyChanged = function (oldValue, newValue, chart) {
    };
    ChartBaseValueMapper.prototype.onSeriesChanged = function (data, chart) {
    };
    ChartBaseValueMapper.prototype.onSeriesPropertyChanged = function (oldValue, newValue, chart) {
    };
    ChartBaseValueMapper.prototype.onAnnotationsChanged = function (data, chart) {
    };
    ChartBaseValueMapper.prototype.onAnnotationsPropertyChanged = function (oldValue, newValue, chart) {
    };
    ChartBaseValueMapper.prototype.onSelectionModeChanged = function (oldValue, newValue, chart) {
    };
    ChartBaseValueMapper.prototype.loadSeries = function (chart) {
    };
    ChartBaseValueMapper.prototype.loadAnnotations = function (chart) {
    };
    ChartBaseValueMapper.prototype.reloadPalettes = function (chart) {
    };
    return ChartBaseValueMapper;
}());
exports.ChartBaseValueMapper = ChartBaseValueMapper;
var CartesianAxisValueMapper = /** @class */ (function () {
    function CartesianAxisValueMapper() {
    }
    CartesianAxisValueMapper.prototype.onLineThicknessChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.ios.style.lineStroke = TKStroke.strokeWithColorWidth(axis.lineColor ? axis.lineColor.ios : axis.ios.style.lineStroke.color, newValue);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLineColorChanged = function (oldValue, newValue, axis) {
        if (newValue !== null && newValue !== undefined) {
            axis.ios.style.lineStroke = TKStroke.strokeWithColorWidth(newValue.ios, axis.lineThickness ? axis.lineThickness : axis.ios.style.lineStroke.width);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLineHiddenChanged = function (oldValue, newValue, axis) {
        if (!newValue && !isNaN(+axis.lineThickness)) {
            axis.ios.style.lineStroke.width = axis.lineThickness;
        }
        axis.ios.style.lineHidden = newValue;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onLabelTextColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.ios.style.labelStyle.textColor = newValue.ios;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBackgroundColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.ios.style.labelStyle.fill = TKSolidFill.solidFillWithColor(axis.labelBackgroundColor.ios);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBorderColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.ios.style.labelStyle.stroke = axis.labelBorderWidth ?
                TKStroke.strokeWithColorWidth(newValue.ios, axis.labelBorderWidth) :
                TKStroke.strokeWithColorWidth(newValue.ios, axis.ios.style.labelStyle.stroke.width);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBorderWidthChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.ios.style.labelStyle.stroke = axis.labelBorderColor ?
                TKStroke.strokeWithColorWidth(axis.labelBorderColor.ios, newValue) :
                TKStroke.strokeWithColorWidth(axis.ios.style.labelStyle.stroke.color, newValue);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelMarginChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            var horizontalOffset = axis.ios.isVertical ? newValue : 0;
            var verticalOffset = axis.ios.isVertical ? 0 : newValue;
            axis.ios.style.labelStyle.textOffset = { horizontal: horizontalOffset, vertical: verticalOffset };
            axis.ios.style.labelStyle.firstLabelTextOffset = { horizontal: horizontalOffset, vertical: verticalOffset };
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelRotationAngleChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.ios.style.labelStyle.rotationAngle = newValue;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelFitModeChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisLabelFitMode.Multiline === newValue) {
                axis.ios.style.labelStyle.fitMode = 1 /* Multiline */;
            }
            else if (chart_public_enum_1.ChartAxisLabelFitMode.Rotate === newValue) {
                axis.ios.style.labelStyle.fitMode = 2 /* Rotate */;
            }
            else {
                axis.ios.style.labelStyle.fitMode = 0 /* None */;
            }
            axis.update();
        }
    };
    // NOTE: initially we cannot set new mode to native object since initially we don't know if the axis is horizontal or vertical
    // that's why we set the mode in owner's axis changed call (see CartesianSeriesValueMapper::onHorizontalAxisChanged)
    // But, we need this update for cases when layout mode is changed from code after the chart is shown
    CartesianAxisValueMapper.prototype.onLabelLayoutModeChanged = function (oldValue, newValue, axis) {
        if (!newValue) {
            return;
        }
        var textAlignment;
        if (axis.ios.isVertical) {
            textAlignment = (newValue === chart_public_enum_1.ChartAxisLabelLayoutMode.Inner) ?
                2 /* Right */ : 1 /* Left */;
        }
        else {
            textAlignment = (newValue === chart_public_enum_1.ChartAxisLabelLayoutMode.Inner) ?
                4 /* Top */ : 8 /* Bottom */;
        }
        axis.ios.style.labelStyle.textAlignment = textAlignment;
        axis.ios.style.labelStyle.firstLabelTextAlignment = textAlignment;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onLabelFormatChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.ios.labelFormat = (newValue);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onTicksHiddenChanged = function (oldValue, newValue, axis) {
        axis.ios.style.majorTickStyle.ticksHidden = newValue;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onTicksThicknessChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.ios.style.majorTickStyle.ticksWidth = newValue;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onTicksLengthChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.ios.style.majorTickStyle.ticksLength = newValue;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onTicksOffsetChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.ios.style.majorTickStyle.ticksOffset = newValue;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onTicksColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            var newColor = newValue.ios;
            axis.ios.style.majorTickStyle.ticksFill = TKSolidFill.solidFillWithColor(newColor);
            axis.ios.style.majorTickStyle.ticksStroke = TKStroke.strokeWithColor(newColor);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onHorizontalLocationChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisHorizontalLocation.Left === newValue) {
                axis.ios.position = 0 /* Left */;
            }
            else if (chart_public_enum_1.ChartAxisHorizontalLocation.Right === newValue) {
                axis.ios.position = 1 /* Right */;
            }
            if (axis.owner && axis.owner.ios) {
                axis.owner.ios.update();
            }
        }
    };
    CartesianAxisValueMapper.prototype.onVerticalLocationChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisVerticalLocation.Top === newValue) {
                axis.ios.position = 2 /* Top */;
            }
            else if (chart_public_enum_1.ChartAxisVerticalLocation.Bottom === newValue) {
                axis.ios.position = 3 /* Bottom */;
            }
            if (axis.owner && axis.owner.ios) {
                axis.owner.ios.update();
            }
        }
    };
    CartesianAxisValueMapper.prototype.onFontInternalChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            var defaultLabelSize = 12;
            var labelFontSize = newValue.fontSize || defaultLabelSize;
            var iosFont = newValue.getUIFont(UIFont.systemFontOfSize(labelFontSize));
            axis.ios.style.labelStyle.font = iosFont;
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onLabelPaddingChanged = function (newValue, propertyName, axis) {
        var paddingLeftFromStyle = axis.label.style.paddingLeft;
        var paddingLeft = (!propertyName || propertyName === "paddingLeft") && !isNaN(+paddingLeftFromStyle) ?
            +paddingLeftFromStyle : axis.ios.style.labelStyle.insets.left;
        var paddingTopFromStyle = axis.label.style.paddingTop;
        var paddingTop = (!propertyName || propertyName === "paddingTop") && !isNaN(+paddingTopFromStyle) ?
            +paddingTopFromStyle : axis.ios.style.labelStyle.insets.top;
        var paddingRightFromStyle = axis.label.style.paddingRight;
        var paddingRight = (!propertyName || propertyName === "paddingRight") && !isNaN(+paddingRightFromStyle) ?
            +paddingRightFromStyle : axis.ios.style.labelStyle.insets.right;
        var paddingBottomFromStyle = axis.label.style.paddingBottom;
        var paddingBottom = (!propertyName || propertyName === "paddingBottom") && !isNaN(+paddingBottomFromStyle) ?
            +paddingBottomFromStyle : axis.ios.style.labelStyle.insets.bottom;
        var insets = new UIEdgeInsets({
            left: paddingLeft,
            top: paddingTop,
            right: paddingRight,
            bottom: paddingBottom
        });
        axis.ios.style.labelStyle.insets = insets;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onLabelSizeChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            var fontName = axis.ios.style.labelStyle.font.fontName;
            axis.ios.style.labelStyle.font = UIFont.fontWithNameSize(fontName, newValue);
            axis.update();
        }
    };
    CartesianAxisValueMapper.prototype.onAllowZoomChanged = function (oldValue, newValue, axis) {
        axis.ios.allowZoom = newValue;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onAllowPanChanged = function (oldValue, newValue, axis) {
        axis.ios.allowPan = newValue;
        axis.update();
    };
    CartesianAxisValueMapper.prototype.onHiddenChanged = function (oldValue, newValue, axis) {
        axis.ios.hidden = newValue;
        axis.update();
    };
    return CartesianAxisValueMapper;
}());
exports.CartesianAxisValueMapper = CartesianAxisValueMapper;
var CategoricalAxisValueMapper = /** @class */ (function (_super) {
    __extends(CategoricalAxisValueMapper, _super);
    function CategoricalAxisValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CategoricalAxisValueMapper.prototype.onMajorTickIntervalChanged = function (oldValue, newValue, axis) {
        if (!isNaN(+newValue)) {
            axis.ios.majorTickInterval = newValue;
            axis.update();
        }
    };
    CategoricalAxisValueMapper.prototype.onPlotModeChanged = function (oldValue, newValue, axis) {
        switch (newValue) {
            case chart_public_enum_1.ChartAxisPlotMode.BetweenTicks:
                axis.ios.setPlotMode(1 /* BetweenTicks */);
                break;
            case chart_public_enum_1.ChartAxisPlotMode.OnTicks:
                axis.ios.setPlotMode(0 /* OnTicks */);
                break;
        }
        axis.update();
    };
    CategoricalAxisValueMapper.prototype.onLastLabelVisibilityChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (axis.plotMode && axis.plotMode !== chart_public_enum_1.ChartAxisPlotMode.OnTicks) {
                console.log("WARNING: Setting the 'lastLabelVisibility' to 'Hidden' with 'plotMode' set to " + axis.plotMode + " is not supported.");
                return;
            }
            switch (newValue) {
                case chart_public_enum_1.ChartAxisLabelVisibility.Hidden: {
                    axis.ios.style.majorTickStyle.maxTickClippingMode = 1 /* Hidden */;
                    axis.ios.style.labelStyle.maxLabelClippingMode = 1 /* Hidden */;
                    axis.update();
                    break;
                }
                case chart_public_enum_1.ChartAxisLabelVisibility.Clip:
                case chart_public_enum_1.ChartAxisLabelVisibility.Visible: {
                    axis.ios.style.majorTickStyle.maxTickClippingMode = 0 /* Visible */;
                    axis.ios.style.labelStyle.maxLabelClippingMode = 0 /* Visible */;
                    axis.update();
                    break;
                }
                default: {
                    console.log("WARNING: Unsupported 'lastLabelVisibility' set: " + newValue);
                    break;
                }
            }
        }
    };
    CategoricalAxisValueMapper.prototype.onFirstLabelVisibilityChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (axis.plotMode && axis.plotMode !== chart_public_enum_1.ChartAxisPlotMode.OnTicks) {
                console.log("WARNING: Setting the 'firstLabelVisibility' to 'Hidden' with 'plotMode' set to " + axis.plotMode + " is not supported.");
                return;
            }
            switch (newValue) {
                case chart_public_enum_1.ChartAxisLabelVisibility.Hidden: {
                    axis.ios.style.majorTickStyle.minTickClippingMode = 1 /* Hidden */;
                    axis.ios.style.labelStyle.minLabelClippingMode = 1 /* Hidden */;
                    axis.update();
                    break;
                }
                case chart_public_enum_1.ChartAxisLabelVisibility.Clip:
                case chart_public_enum_1.ChartAxisLabelVisibility.Visible: {
                    axis.ios.style.majorTickStyle.minTickClippingMode = 0 /* Visible */;
                    axis.ios.style.labelStyle.minLabelClippingMode = 0 /* Visible */;
                    axis.update();
                    break;
                }
                default: {
                    console.log("WARNING: Unsupported 'firstLabelVisibility' set: " + newValue);
                    break;
                }
            }
        }
    };
    return CategoricalAxisValueMapper;
}(CartesianAxisValueMapper));
exports.CategoricalAxisValueMapper = CategoricalAxisValueMapper;
////////////////////////////////////////////////////////////////////////////////////////////
//  Series value mappers
////////////////////////////////////////////////////////////////////////////////////////////
var ChartSeriesValueMapper = /** @class */ (function () {
    function ChartSeriesValueMapper() {
    }
    ChartSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) { };
    ChartSeriesValueMapper.prototype.onValuePropertyChanged = function (oldValue, newValue, series) { };
    ChartSeriesValueMapper.prototype.onShowLabelsChanged = function (oldValue, newValue, series) {
        if (series.ios && newValue) {
            this.applyLabelStyle(series);
        }
    };
    ChartSeriesValueMapper.prototype.onLegendTitleChanged = function (newValue, series) {
        if (series.ios && newValue) {
            series.ios.title = newValue;
            if (series.owner) {
                series.owner.nativeView.legend.reloadItems();
            }
        }
    };
    ChartSeriesValueMapper.prototype.onSelectionModeChanged = function (oldValue, newValue, series) {
        if (series.ios && newValue) {
            this.applySelectionMode(series);
            if (series.owner) {
                series.updateOwnerChart();
            }
        }
    };
    ChartSeriesValueMapper.prototype.onLabelPaddingChanged = function (newValue, propertyName, series) {
        this.applyLabelInsetsFromStyle(series);
    };
    ChartSeriesValueMapper.prototype.onFontInternalChanged = function (oldValue, newValue, series) {
        if (series.ios && newValue) {
            this.applyLabelFontFromStyle(series);
        }
    };
    ChartSeriesValueMapper.prototype.applySelectionMode = function (series) {
        switch (series.selectionMode) {
            case chart_public_enum_1.ChartSeriesSelectionMode.Series:
                series.ios.selection = 2 /* Series */;
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.DataPoint:
                series.ios.selection = 3 /* DataPoint */;
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.None:
                series.ios.selection = 1 /* None */;
                break;
            case (chart_public_enum_1.ChartSeriesSelectionMode.NotSet):
                series.ios.selection = 0 /* NotSet */;
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.DataPointMultiple:
                series.ios.selection = 4 /* DataPointMultiple */;
                break;
        }
    };
    ChartSeriesValueMapper.prototype.onLabelStyleChanged = function (oldValue, newValue, series) {
        if (series.ios && newValue) {
            this.applyLabelStyle(series);
        }
    };
    ChartSeriesValueMapper.prototype.applyLabelFontFromStyle = function (series) {
        if (series.ios && series.label && series.label.style.fontInternal !== font_1.Font.default) {
            var defaultLabelSize = 10;
            var labelFontSize = series.label.style.fontInternal.fontSize || defaultLabelSize;
            var iosFont = series.label.style.fontInternal.getUIFont(UIFont.systemFontOfSize(labelFontSize));
            series.ios.style.pointLabelStyle.font = iosFont;
            if (series.owner) {
                series.updateOwnerChart();
            }
        }
    };
    ChartSeriesValueMapper.prototype.applyLabelInsetsFromStyle = function (series) {
        if (!series.ios || !series.label) {
            return;
        }
        var paddingLeft = !isNaN(+series.label.style.paddingLeft) ? +series.label.style.paddingLeft : series.ios.style.pointLabelStyle.insets.left;
        var paddingTop = !isNaN(+series.label.style.paddingTop) ? +series.label.style.paddingTop : series.ios.style.pointLabelStyle.insets.top;
        var paddingRight = !isNaN(+series.label.style.paddingRight) ? +series.label.style.paddingRight : series.ios.style.pointLabelStyle.insets.right;
        var paddingBottom = !isNaN(+series.label.style.paddingBottom) ? +series.label.style.paddingBottom : series.ios.style.pointLabelStyle.insets.bottom;
        var insets = new UIEdgeInsets({
            left: paddingLeft,
            top: paddingTop,
            right: paddingRight,
            bottom: paddingBottom
        });
        series.ios.style.pointLabelStyle.insets = insets;
        if (series.owner) {
            series.updateOwnerChart();
        }
    };
    ChartSeriesValueMapper.prototype.applyLabelStyle = function (series) {
        if (!series || !series.ios) {
            return;
        }
        series.ios.style.pointLabelStyle.textHidden = !series.showLabels;
        if (series.labelStyle) {
            if (series.labelStyle.textColor) {
                series.ios.style.pointLabelStyle.textColor = series.labelStyle.textColor.ios;
            }
            if (series.labelStyle.fillColor) {
                series.ios.style.pointLabelStyle.fill = TKSolidFill.solidFillWithColor(series.labelStyle.fillColor.ios);
            }
            if (series.labelStyle.strokeWidth || series.labelStyle.strokeColor) {
                var strokeColor = series.labelStyle.strokeColor ? series.labelStyle.strokeColor.ios : new color_1.Color("Black").ios;
                var stroke = TKStroke.strokeWithColor(strokeColor);
                if (series.labelStyle.strokeWidth) {
                    stroke.width = series.labelStyle.strokeWidth;
                }
                series.ios.style.pointLabelStyle.stroke = stroke;
            }
            if (series.labelStyle.margin) {
                var marginIsVertical = false;
                if (series instanceof chart_series_common_1.CartesianSeries) {
                    if (series.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis ||
                        (series.owner && series.owner.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
                        marginIsVertical = true;
                    }
                }
                var horizontalOffset = marginIsVertical ? 0 : series.labelStyle.margin;
                var verticalOffset = marginIsVertical ? series.labelStyle.margin : 0;
                series.ios.style.pointLabelStyle.labelOffset = { horizontal: horizontalOffset, vertical: verticalOffset };
            }
            if (series.labelStyle.textFormat) {
                series.ios.style.pointLabelStyle.stringFormat = series.labelStyle.textFormat;
            }
            if (series.labelStyle.fontStyle || series.labelStyle.fontStyle || series.labelStyle.textSize) {
                var font = null;
                var fontSize = !isNaN(+series.labelStyle.textSize) ? series.labelStyle.textSize : 10;
                if (series.labelStyle.fontName) {
                    font = UIFont.fontWithNameSize(series.labelStyle.fontName, fontSize);
                    if (!font) {
                        console.log("WARNING: Cannot create font with given name: " + series.labelStyle.fontName);
                        return;
                    }
                }
                if (!font) {
                    font = UIFont.systemFontOfSize(fontSize);
                }
                if (series.labelStyle.fontStyle) {
                    var traits = 0 /* ClassUnknown */;
                    var fontStyle = series.labelStyle.fontStyle;
                    switch (fontStyle) {
                        case chart_public_enum_1.ChartFontStyle.Bold:
                            traits = 2 /* TraitBold */;
                            break;
                        case chart_public_enum_1.ChartFontStyle.Italic:
                            traits = 1 /* TraitItalic */;
                            break;
                        case chart_public_enum_1.ChartFontStyle.BoldItalic:
                            traits = 2 /* TraitBold */ | 1 /* TraitItalic */;
                            break;
                    }
                    var newFont = UIFont.fontWithDescriptorSize(font.fontDescriptor.fontDescriptorWithSymbolicTraits(traits), fontSize);
                    if (newFont) {
                        font = newFont;
                    }
                }
                if (font) {
                    series.ios.style.pointLabelStyle.font = font;
                }
            }
        }
        this.applyLabelFontFromStyle(series);
        this.applyLabelInsetsFromStyle(series);
        if (series.owner) {
            series.updateOwnerChart();
        }
    };
    ChartSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return series && series.items;
    };
    return ChartSeriesValueMapper;
}());
exports.ChartSeriesValueMapper = ChartSeriesValueMapper;
var PieSeriesValueMapper = /** @class */ (function (_super) {
    __extends(PieSeriesValueMapper, _super);
    function PieSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PieSeriesValueMapper.prototype.onShowLabelsChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            this.applyLabelStyle(series);
        }
    };
    PieSeriesValueMapper.prototype.onValuePropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    PieSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    PieSeriesValueMapper.prototype.onLabelPropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    PieSeriesValueMapper.prototype.onLabelStyleChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            this.applyLabelStyle(series);
        }
    };
    PieSeriesValueMapper.prototype.onExpandRadiusChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            var radius = newValue;
            if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
                console.log("WARNING : Expand radius factor must be number in range [0,1]");
            }
            else {
                series.ios.expandRadius = radius + 1;
            }
        }
    };
    PieSeriesValueMapper.prototype.onOuterRadiusFactorChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            var radius = newValue;
            if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
                console.log("WARNING : Outer radius factor must be number in range [0,1]");
            }
            else {
                series.ios.outerRadius = radius;
                series.updateOwnerChart();
            }
        }
    };
    PieSeriesValueMapper.prototype.onShowPercentageChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            series.ios.displayPercentage = (newValue) ? true : false;
            series.updateOwnerChart();
        }
    };
    PieSeriesValueMapper.prototype.onStartAngleChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            if (!isNaN(+newValue)) {
                series.ios.startAngle = newValue * Math.PI / 180;
                series.updateOwnerChart();
            }
        }
    };
    PieSeriesValueMapper.prototype.onEndAngleChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            if (!isNaN(+newValue)) {
                series.ios.endAngle = series.endAngle * Math.PI / 180;
                series.updateOwnerChart();
            }
        }
    };
    PieSeriesValueMapper.prototype.onLegendTitleChanged = function (newValue, series) {
        console.log("WARNING: legendTitle property is not applicable to Pie series. Use 'legendLabel' instead.");
    };
    PieSeriesValueMapper.prototype.onFillColorsChanged = function (oldValue, newValue, series) {
        this.applyFills(series);
    };
    PieSeriesValueMapper.prototype.onStrokeWidthChanged = function (oldValue, newValue, series) {
        this.applyStrokes(series);
    };
    PieSeriesValueMapper.prototype.onStrokeColorsChanged = function (oldValue, newValue, series) {
        this.applyStrokes(series);
    };
    PieSeriesValueMapper.prototype.applyFills = function (series) {
        if (series.ios && series.fillColors) {
            var nativeArray = NSMutableArray.alloc().initWithCapacity(series.fillColors.length);
            for (var i = 0; i < series.fillColors.length; i++) {
                nativeArray.addObject(TKSolidFill.solidFillWithColor(series.fillColors[i].ios));
            }
            series.ios.style.fills = nativeArray;
        }
    };
    PieSeriesValueMapper.prototype.applyStrokes = function (series) {
        if (series.ios && series.strokeColors && series.strokeWidth) {
            var nativeArray = NSMutableArray.alloc().initWithCapacity(series.strokeColors.length);
            for (var i = 0; i < series.strokeColors.length; i++) {
                nativeArray.addObject(TKStroke.strokeWithColorWidth(series.strokeColors[i].ios, series.strokeWidth));
            }
            series.ios.style.strokes = nativeArray;
        }
    };
    PieSeriesValueMapper.prototype.applyColors = function (series) {
        this.applyFills(series);
        this.applyStrokes(series);
    };
    PieSeriesValueMapper.prototype.getNativeData = function (series) {
        var length = series.items.length;
        var nativeSource = NSMutableArray.new();
        var item, name, value;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            value = item[series.valueProperty];
            if (series.legendLabel) {
                name = item[series.legendLabel];
            }
            else {
                name = "Series " + (i + 1);
            }
            nativeSource.addObject(TKChartDataPoint.alloc().initWithNameValue(name, value));
        }
        return nativeSource;
    };
    PieSeriesValueMapper.prototype.setCommonProperties = function (series) {
        var radius = series.expandRadius;
        if (radius) {
            if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
                console.log("WARNING : Expand radius factor must be number in range [0,1]");
            }
            else {
                series.ios.expandRadius = radius + 1;
            }
        }
        radius = series.outerRadiusFactor;
        if (radius) {
            if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
                console.log("WARNING : Outer radius factor must be number in range [0,1]");
            }
            else {
                series.ios.outerRadius = radius;
            }
        }
        if (!isNaN(series.startAngle)) {
            series.ios.startAngle = series.startAngle * Math.PI / 180;
        }
        if (!isNaN(series.endAngle)) {
            series.ios.endAngle = series.endAngle * Math.PI / 180;
        }
    };
    PieSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.valueProperty;
    };
    PieSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        series.ios = TKChartPieSeries.alloc().initWithItems(this.getNativeData(series));
        series.ios.adjustSizeToFit = false;
        this.setCommonProperties(series);
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyColors(series);
        series.ios.labelDisplayMode = 2 /* Outside */;
        series.ios.visibleInLegend = true;
        series.ios.rotationEnabled = false;
        series.ios.displayPercentage = (series.showPercentage === undefined) ? true : series.showPercentage;
        if (series.owner && series.owner['addSeries']) {
            // only series from xml have owner, manually created series are added on series changed handler
            series.owner['addSeries'](series.ios);
        }
    };
    return PieSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.PieSeriesValueMapper = PieSeriesValueMapper;
var DonutSeriesValueMapper = /** @class */ (function (_super) {
    __extends(DonutSeriesValueMapper, _super);
    function DonutSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DonutSeriesValueMapper.prototype.onInnerRadiusFactorChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            var radius = newValue;
            if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
                console.log("WARNING : Inner radius factor must be number in range (0,1)");
            }
            else {
                series.ios.innerRadius = (series.innerRadiusFactor) ? series.innerRadiusFactor : 0.5;
                series.updateOwnerChart();
            }
        }
    };
    DonutSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        series.ios = TKChartDonutSeries.alloc().initWithItems(this.getNativeData(series));
        series.ios.adjustSizeToFit = false;
        var radius = series.innerRadiusFactor;
        if (isNaN(+radius) || radius < 0.0 || radius > 1.0) {
            console.log("WARNING : Inner radius factor must be number in range (0,1)");
        }
        else {
            series.ios.innerRadius = (series.innerRadiusFactor) ? series.innerRadiusFactor : 0.5;
        }
        this.setCommonProperties(series);
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyColors(series);
        series.ios.labelDisplayMode = 2 /* Outside */;
        series.ios.visibleInLegend = true;
        series.ios.rotationEnabled = false;
        series.ios.displayPercentage = (series.showPercentage === undefined) ? true : series.showPercentage;
        if (series.owner && series.owner['addSeries']) {
            series.owner['addSeries'](series.ios);
        }
    };
    return DonutSeriesValueMapper;
}(PieSeriesValueMapper));
exports.DonutSeriesValueMapper = DonutSeriesValueMapper;
var CartesianSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CartesianSeriesValueMapper, _super);
    function CartesianSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CartesianSeriesValueMapper.prototype.onHorizontalAxisChanged = function (oldValue, newValue, series) {
        // since label alignment requires info about horizontal or vertical the axis is, we set appropriate values here
        var labelLayoutMode = series.horizontalAxis.labelLayoutMode;
        if (labelLayoutMode) {
            if (labelLayoutMode === chart_public_enum_1.ChartAxisLabelLayoutMode.Inner) {
                series.horizontalAxis.ios.style.labelStyle.textAlignment = 4 /* Top */;
                series.horizontalAxis.ios.style.labelStyle.firstLabelTextAlignment = 4 /* Top */;
            }
            else {
                series.horizontalAxis.ios.style.labelStyle.textAlignment = 8 /* Bottom */;
                series.horizontalAxis.ios.style.labelStyle.firstLabelTextAlignment = 8 /* Bottom */;
            }
        }
        this.updateNativeSeriesAxis(series, oldValue, newValue, false);
    };
    CartesianSeriesValueMapper.prototype.onVerticalAxisChanged = function (oldValue, newValue, series) {
        var labelLayoutMode = series.verticalAxis.labelLayoutMode;
        if (labelLayoutMode) {
            if (labelLayoutMode === chart_public_enum_1.ChartAxisLabelLayoutMode.Inner) {
                series.verticalAxis.ios.style.labelStyle.textAlignment = 2 /* Right */;
                series.verticalAxis.ios.style.labelStyle.firstLabelTextAlignment = 2 /* Right */;
            }
            else {
                series.verticalAxis.ios.style.labelStyle.textAlignment = 1 /* Left */;
                series.verticalAxis.ios.style.labelStyle.firstLabelTextAlignment = 1 /* Left */;
            }
        }
        this.updateNativeSeriesAxis(series, oldValue, newValue, true);
    };
    CartesianSeriesValueMapper.prototype.onPaletteModeChanged = function (oldValue, newValue, series) {
        if (series.ios) {
            this.applyPaletteMode(series);
        }
    };
    CartesianSeriesValueMapper.prototype.applyPaletteMode = function (series) {
        if (series.ios) {
            var nativePaletteMode = chart_native_helper_1.convertToIOSPaletteMode(series.paletteMode);
            if (nativePaletteMode !== null) {
                series.ios.style.paletteMode = nativePaletteMode;
                series.updateOwnerChart();
            }
        }
    };
    CartesianSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        this.applyFillColor(series);
    };
    CartesianSeriesValueMapper.prototype.onStrokeWidthChanged = function (oldValue, newValue, series) {
        this.applyStrokeWidth(series);
    };
    CartesianSeriesValueMapper.prototype.onStrokeColorChanged = function (oldValue, newValue, series) {
        this.applyStrokeColor(series);
    };
    CartesianSeriesValueMapper.prototype.applyFillColor = function (series) {
        if (series.ios && series.fillColor) {
            series.ios.style.fill = TKSolidFill.solidFillWithColor(series.fillColor.ios);
            if (series.owner) {
                series.updateOwnerChart();
            }
        }
    };
    CartesianSeriesValueMapper.prototype.applyStrokeWidth = function (series) {
        this.applyStroke(series);
    };
    CartesianSeriesValueMapper.prototype.applyStrokeColor = function (series) {
        this.applyStroke(series);
    };
    CartesianSeriesValueMapper.prototype.applyStroke = function (series) {
        if (series.ios && series.strokeColor && series.strokeWidth) {
            series.ios.style.stroke = TKStroke.strokeWithColorWidth(series.strokeColor.ios, series.strokeWidth);
            if (series.owner) {
                series.updateOwnerChart();
            }
        }
    };
    CartesianSeriesValueMapper.prototype.applyColors = function (series) {
        this.applyFillColor(series);
        this.applyStroke(series);
    };
    CartesianSeriesValueMapper.prototype.updateNativeSeriesAxis = function (series, oldAxis, newAxis, isVertical) {
        if (!series.ios || !series.ios.chart) {
            return;
        }
        var nativeSeries = series.ios;
        var nativeOldAxis = oldAxis ? oldAxis.ios : null;
        var nativeNewAxis = newAxis ? newAxis.ios : null;
        series.ios.chart.replaceSeriesAxisOldNewVertical(nativeSeries, nativeOldAxis, nativeNewAxis, isVertical);
    };
    CartesianSeriesValueMapper.prototype.bindSeriesAxes = function (series) {
        if (series.horizontalAxis) {
            this.updateNativeSeriesAxis(series, null, series.horizontalAxis, false);
        }
        if (series.verticalAxis) {
            this.updateNativeSeriesAxis(series, null, series.verticalAxis, true);
        }
    };
    return CartesianSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.CartesianSeriesValueMapper = CartesianSeriesValueMapper;
var CategoricalSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CategoricalSeriesValueMapper, _super);
    function CategoricalSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CategoricalSeriesValueMapper.prototype.onCategoryPropertyChanged = function (oldValue, newValue, series) {
        if (series.horizontalAxis && series.verticalAxis && series.ios) {
            series.ios.setHorizontalAxis(series.horizontalAxis.ios);
            series.ios.setVerticalAxis(series.verticalAxis.ios);
        }
        this.bind(series);
    };
    CategoricalSeriesValueMapper.prototype.onValuePropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    CategoricalSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    CategoricalSeriesValueMapper.prototype.onShowLabelsChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            this.applyLabelStyle(series);
        }
    };
    CategoricalSeriesValueMapper.prototype.onStackModePropertyChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            if (series.stackMode) {
                this.bind(series);
            }
        }
        else {
            this.setStackInfo(series);
        }
    };
    CategoricalSeriesValueMapper.prototype.onLabelStyleChanged = function (oldValue, newValue, series) {
        if (series.ios && newValue) {
            this.applyLabelStyle(series);
        }
    };
    CategoricalSeriesValueMapper.prototype.updateNative = function (series) {
        this.bind(series);
    };
    CategoricalSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.categoryProperty;
    };
    CategoricalSeriesValueMapper.prototype.bind = function (series) {
        console.log("WARNING: You should override this bind in class specific initializer");
    };
    // This method must be called by every bind override at an appropriate place where the series should be given a combine mode
    CategoricalSeriesValueMapper.prototype.setStackInfo = function (series) {
        if (!series.stackMode) {
            return;
        }
        switch (series.stackMode) {
            case chart_public_enum_1.ChartSeriesStackMode.Stack:
                series.ios.stackInfo = TKChartStackInfo.alloc().initWithIDWithStackMode(1, 0 /* Stack */);
                break;
            case chart_public_enum_1.ChartSeriesStackMode.Stack100:
                series.ios.stackInfo = TKChartStackInfo.alloc().initWithIDWithStackMode(1, 1 /* Stack100 */);
                break;
            case chart_public_enum_1.ChartSeriesStackMode.None:
                series.ios.stackInfo = null;
                break;
        }
        if (series.ios.chart) {
            series.ios.chart.reloadData();
        }
    };
    /**
    *  Cleans the axes from previous data categories if any.
    *  This method should be called if series items collection is changed.
    */
    CategoricalSeriesValueMapper.prototype.cleanCategoryAxes = function (series) {
        if (series.ios && series.ios.items && series.owner) {
            series.ios.items.removeAllObjects();
            if (series.horizontalAxis &&
                (series.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
                series.horizontalAxis.ios.removeAllCategories();
            }
            else if (series.owner.horizontalAxis &&
                (series.owner.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
                series.owner.horizontalAxis.ios.removeAllCategories();
            }
            if (series.verticalAxis &&
                (series.verticalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
                series.verticalAxis.ios.removeAllCategories();
            }
            else if (series.owner.verticalAxis &&
                (series.owner.verticalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
                series.owner.verticalAxis.ios.removeAllCategories();
            }
        }
    };
    return CategoricalSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.CategoricalSeriesValueMapper = CategoricalSeriesValueMapper;
var BarSeriesValueMapper = /** @class */ (function (_super) {
    __extends(BarSeriesValueMapper, _super);
    function BarSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BarSeriesValueMapper.prototype.onHorizontalAxisChanged = function (oldValue, newValue, series) {
        series.updateNative();
        _super.prototype.onHorizontalAxisChanged.call(this, oldValue, newValue, series);
    };
    BarSeriesValueMapper.prototype.onVerticalAxisChanged = function (oldValue, newValue, series) {
        series.updateNative();
        _super.prototype.onVerticalAxisChanged.call(this, oldValue, newValue, series);
    };
    BarSeriesValueMapper.prototype.onMinBarSizeChanged = function (oldValue, newValue, series) {
        if (series.ios) {
            if (!isNaN(newValue)) {
                series.ios.minBarHeight = utils.layout.toDevicePixels(newValue);
                series.ios.minColumnWidth = utils.layout.toDevicePixels(newValue);
            }
            else {
                series.ios.maxBarHeight = 0;
                series.ios.maxColumnWidth = 0;
            }
        }
    };
    BarSeriesValueMapper.prototype.onMaxBarSizeChanged = function (oldValue, newValue, series) {
        if (series.ios) {
            if (!isNaN(newValue)) {
                series.ios.maxBarHeight = utils.layout.toDevicePixels(newValue);
                series.ios.maxColumnWidth = utils.layout.toDevicePixels(newValue);
            }
            else {
                series.ios.maxBarHeight = 0;
                series.ios.maxColumnWidth = 0;
            }
        }
    };
    BarSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.valueProperty;
    };
    BarSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var useColumnSeries;
        if (series.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis ||
            (series.owner && series.owner.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
            useColumnSeries = true;
        }
        this.cleanCategoryAxes(series);
        var nativeSource = NSMutableArray.new();
        if (series && series.items) {
            var length_1 = series.items.length;
            for (var i = 0; i < length_1; i++) {
                var item = series.getItemAtIndex(i);
                var value = item[series.valueProperty];
                var category = item[series.categoryProperty];
                nativeSource.addObject(useColumnSeries ?
                    TKChartDataPoint.alloc().initWithXY(category, value) :
                    TKChartDataPoint.alloc().initWithXY(value, category));
            }
        }
        series.ios = useColumnSeries ?
            TKChartColumnSeries.alloc().initWithItems(nativeSource) :
            TKChartBarSeries.alloc().initWithItems(nativeSource);
        if (!isNaN(series.minBarSize)) {
            series.ios.minColumnWidth = utils.layout.toDevicePixels(series.minBarSize);
            series.ios.minBarHeight = utils.layout.toDevicePixels(series.minBarSize);
        }
        else {
            series.ios.minColumnWidth = 0;
            series.ios.minBarHeight = 0;
        }
        if (!isNaN(series.maxBarSize)) {
            series.ios.maxColumnWidth = utils.layout.toDevicePixels(series.maxBarSize);
            series.ios.maxBarHeight = utils.layout.toDevicePixels(series.maxBarSize);
        }
        else {
            series.ios.maxColumnWidth = 0;
            series.ios.maxBarHeight = 0;
        }
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return BarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BarSeriesValueMapper = BarSeriesValueMapper;
var RangeBarSeriesValueMapper = /** @class */ (function (_super) {
    __extends(RangeBarSeriesValueMapper, _super);
    function RangeBarSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RangeBarSeriesValueMapper.prototype.onHorizontalAxisChanged = function (oldValue, newValue, series) {
        series.updateNative();
        _super.prototype.onHorizontalAxisChanged.call(this, oldValue, newValue, series);
    };
    RangeBarSeriesValueMapper.prototype.onVerticalAxisChanged = function (oldValue, newValue, series) {
        series.updateNative();
        _super.prototype.onVerticalAxisChanged.call(this, oldValue, newValue, series);
    };
    RangeBarSeriesValueMapper.prototype.onHighPropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    RangeBarSeriesValueMapper.prototype.onLowPropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    RangeBarSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.highPropertyName && series.lowPropertyName;
    };
    RangeBarSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var useColumnSeries = false;
        if (series.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis ||
            (series.owner && series.owner.horizontalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
            useColumnSeries = true;
        }
        var length = series.items.length;
        var nativeSource = NSMutableArray.new();
        var item = null;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            nativeSource.addObject(useColumnSeries ?
                TKChartRangeDataPoint.alloc().initWithXLowHigh(item[series.categoryProperty], item[series.lowPropertyName], item[series.highPropertyName]) :
                TKChartRangeDataPoint.alloc().initWithYLowHigh(item[series.categoryProperty], item[series.lowPropertyName], item[series.highPropertyName]));
        }
        this.cleanCategoryAxes(series);
        series.ios = useColumnSeries ?
            TKChartRangeColumnSeries.alloc().initWithItems(nativeSource) :
            TKChartRangeBarSeries.alloc().initWithItems(nativeSource);
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return RangeBarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.RangeBarSeriesValueMapper = RangeBarSeriesValueMapper;
var LineSeriesValueMapper = /** @class */ (function (_super) {
    __extends(LineSeriesValueMapper, _super);
    function LineSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LineSeriesValueMapper.prototype.getDataPointsForSeries = function (series) {
        var isCatAxisVertical = false;
        if (series.verticalAxis instanceof chart_axis_common_1.CategoricalAxis ||
            (series.owner && series.owner.verticalAxis instanceof chart_axis_common_1.CategoricalAxis)) {
            isCatAxisVertical = true;
        }
        var nativeSource = NSMutableArray.new();
        var length = series.items.length;
        var item = null;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            nativeSource.addObject(TKChartDataPoint.alloc().initWithXY((isCatAxisVertical) ? item[series.valueProperty] : item[series.categoryProperty], (isCatAxisVertical) ? item[series.categoryProperty] : item[series.valueProperty]));
        }
        return nativeSource;
    };
    LineSeriesValueMapper.prototype.applyPaletteMode = function (series) {
        if (series.paletteMode !== undefined) {
            console.log('WARNING: Property paletteMode is not applicable for series of type: ' + series.constructor.name);
        }
    };
    LineSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.valueProperty;
    };
    LineSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartLineSeries.alloc().initWithItems(this.getDataPointsForSeries(series));
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return LineSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.LineSeriesValueMapper = LineSeriesValueMapper;
var SplineSeriesValueMapper = /** @class */ (function (_super) {
    __extends(SplineSeriesValueMapper, _super);
    function SplineSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SplineSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartSplineSeries.alloc().initWithItems(this.getDataPointsForSeries(series));
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return SplineSeriesValueMapper;
}(LineSeriesValueMapper));
exports.SplineSeriesValueMapper = SplineSeriesValueMapper;
var AreaSeriesValueMapper = /** @class */ (function (_super) {
    __extends(AreaSeriesValueMapper, _super);
    function AreaSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AreaSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartAreaSeries.alloc().initWithItems(this.getDataPointsForSeries(series));
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return AreaSeriesValueMapper;
}(LineSeriesValueMapper));
exports.AreaSeriesValueMapper = AreaSeriesValueMapper;
var SplineAreaSeriesValueMapper = /** @class */ (function (_super) {
    __extends(SplineAreaSeriesValueMapper, _super);
    function SplineAreaSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SplineAreaSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartSplineAreaSeries.alloc().initWithItems(this.getDataPointsForSeries(series));
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.setStackInfo(series);
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return SplineAreaSeriesValueMapper;
}(AreaSeriesValueMapper));
exports.SplineAreaSeriesValueMapper = SplineAreaSeriesValueMapper;
var BubbleSeriesValueMapper = /** @class */ (function (_super) {
    __extends(BubbleSeriesValueMapper, _super);
    function BubbleSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue, series) {
        // todo: this property update should recreate the data point, that's why we will remove the current native instance and must create a new one
        this.bind(series);
    };
    BubbleSeriesValueMapper.prototype.onBubbleScalePropertyChanged = function (oldValue, newValue, series) {
        if (!series.ios) {
            this.bind(series);
        }
        else {
            if (!isNaN(+newValue))
                series.ios.scale = newValue;
        }
    };
    BubbleSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.valueProperty && series.bubbleSizeProperty;
    };
    BubbleSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var nativeSource = NSMutableArray.new();
        var length = series.items.length;
        var item = null;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            nativeSource.addObject(TKChartBubbleDataPoint.alloc().initWithXYArea(item[series.categoryProperty], item[series.valueProperty], item[series.bubbleSizeProperty]));
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartBubbleSeries.alloc().initWithItems(nativeSource);
        series.ios.scale = series.bubbleScale;
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return BubbleSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BubbleSeriesValueMapper = BubbleSeriesValueMapper;
var OhlcSeriesValueMapper = /** @class */ (function (_super) {
    __extends(OhlcSeriesValueMapper, _super);
    function OhlcSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OhlcSeriesValueMapper.prototype.onHighPropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    OhlcSeriesValueMapper.prototype.onLowPropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    OhlcSeriesValueMapper.prototype.onOpenPropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    OhlcSeriesValueMapper.prototype.onClosePropertyNameChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    OhlcSeriesValueMapper.prototype.onValuePropertyChanged = function (oldValue, newValue, series) {
        console.log("WARNING: OHLC series doesn't use valueProperty property.");
    };
    OhlcSeriesValueMapper.prototype.onStrokeColorsChanged = function (oldValue, newValue, series) {
        this.applyStrokes(series);
    };
    OhlcSeriesValueMapper.prototype.applyColors = function (series) {
        _super.prototype.applyColors.call(this, series);
        this.applyStrokes(series);
    };
    OhlcSeriesValueMapper.prototype.applyStrokes = function (series) {
        if (series.ios && series.strokeColors) {
            var nativeArray = NSMutableArray.alloc().initWithCapacity(series.strokeColors.length);
            for (var i = 0; i < series.strokeColors.length; i++) {
                nativeArray.addObject(TKStroke.strokeWithColorWidth(series.strokeColors[i].ios, series.strokeWidth));
            }
            series.ios.style.strokes = nativeArray;
        }
    };
    OhlcSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.openPropertyName && series.highPropertyName && series.lowPropertyName && series.closePropertyName;
    };
    // public onShowLabelsChanged(data: PropertyChangeData, series: chartCommonModule.OhlcSeries) {
    //     if (series.items && series.categoryProperty) {
    //         this.bind(series);
    //     }
    // }
    OhlcSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var nativeSource = NSMutableArray.new();
        var currentItem = null;
        for (var i = 0; i < series.items.length; i++) {
            currentItem = series.getItemAtIndex(i);
            var categoryValue = currentItem[series.categoryProperty];
            var date = getDateObject(categoryValue);
            nativeSource.addObject(TKChartFinancialDataPoint.dataPointWithXOpenHighLowClose(date, currentItem[series.openPropertyName], currentItem[series.highPropertyName], currentItem[series.lowPropertyName], currentItem[series.closePropertyName]));
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartOhlcSeries.alloc().initWithItems(nativeSource);
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return OhlcSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.OhlcSeriesValueMapper = OhlcSeriesValueMapper;
var CandlestickSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CandlestickSeriesValueMapper, _super);
    function CandlestickSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CandlestickSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var nativeSource = NSMutableArray.new();
        var item = null;
        for (var i = 0; i < series.items.length; i++) {
            item = series.getItemAtIndex(i);
            var categoryValue = item[series.categoryProperty];
            var date = getDateObject(categoryValue);
            nativeSource.addObject(TKChartFinancialDataPoint.dataPointWithXOpenHighLowClose(date, item[series.openPropertyName], item[series.highPropertyName], item[series.lowPropertyName], item[series.closePropertyName]));
        }
        this.cleanCategoryAxes(series);
        series.ios = TKChartCandlestickSeries.alloc().initWithItems(nativeSource);
        this.applyLabelStyle(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    CandlestickSeriesValueMapper.prototype.onFillColorsChanged = function (oldValue, newValue, series) {
        this.applyFills(series);
    };
    CandlestickSeriesValueMapper.prototype.applyColors = function (series) {
        _super.prototype.applyColors.call(this, series);
        this.applyFills(series);
    };
    CandlestickSeriesValueMapper.prototype.applyFills = function (series) {
        if (series.ios && series.fillColors) {
            var nativeArray = NSMutableArray.alloc().initWithCapacity(series.fillColors.length);
            for (var i = 0; i < series.fillColors.length; i++) {
                nativeArray.addObject(TKSolidFill.solidFillWithColor(series.fillColors[i].ios));
            }
            series.ios.style.fills = nativeArray;
            series.ios.style.paletteMode = 1 /* UseItemIndex */;
        }
    };
    return CandlestickSeriesValueMapper;
}(OhlcSeriesValueMapper));
exports.CandlestickSeriesValueMapper = CandlestickSeriesValueMapper;
var ScatterSeriesValueMapper = /** @class */ (function (_super) {
    __extends(ScatterSeriesValueMapper, _super);
    function ScatterSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScatterSeriesValueMapper.prototype.onXPropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    ScatterSeriesValueMapper.prototype.onYPropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    ScatterSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    ScatterSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.xProperty && series.yProperty;
    };
    ScatterSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var nativeSource = NSMutableArray.new();
        var length = series.items.length;
        var item = null;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            // NOTE: scatter series are intended for numeric data on X & Y
            nativeSource.addObject(TKChartDataPoint.alloc().initWithXY(item[series.xProperty], item[series.yProperty]));
        }
        series.ios = TKChartScatterSeries.alloc().initWithItems(nativeSource);
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return ScatterSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.ScatterSeriesValueMapper = ScatterSeriesValueMapper;
var ScatterBubbleSeriesValueMapper = /** @class */ (function (_super) {
    __extends(ScatterBubbleSeriesValueMapper, _super);
    function ScatterBubbleSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScatterBubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue, series) {
        this.bind(series);
    };
    ScatterBubbleSeriesValueMapper.prototype.onBubbleScaleChanged = function (oldValue, newValue, series) {
        if (series.ios) {
            if (!isNaN(+newValue))
                series.ios.scale = newValue;
        }
        else {
            this.bind(series);
        }
    };
    ScatterBubbleSeriesValueMapper.prototype.seriesCanBeBound = function (series) {
        return _super.prototype.seriesCanBeBound.call(this, series) && series.xProperty && series.yProperty && series.bubbleSizeProperty;
    };
    ScatterBubbleSeriesValueMapper.prototype.bind = function (series) {
        if (!this.seriesCanBeBound(series)) {
            return;
        }
        var nativeSource = NSMutableArray.new();
        var length = series.items.length;
        var item = null;
        for (var i = 0; i < length; i++) {
            item = series.getItemAtIndex(i);
            nativeSource.addObject(TKChartBubbleDataPoint.alloc().initWithXYArea(item[series.xProperty], item[series.yProperty], item[series.bubbleSizeProperty]));
        }
        series.ios = TKChartBubbleSeries.alloc().initWithItems(nativeSource);
        series.ios.scale = series.bubbleScale;
        this.applyLabelStyle(series);
        this.applySelectionMode(series);
        this.applyPaletteMode(series);
        this.applyColors(series);
        series.ios.title = series.legendTitle;
        this.bindSeriesAxes(series);
        if (series.owner) {
            series.owner['addSeries'](series);
        }
    };
    return ScatterBubbleSeriesValueMapper;
}(ScatterSeriesValueMapper));
exports.ScatterBubbleSeriesValueMapper = ScatterBubbleSeriesValueMapper;
function getDateObject(categoryValue) {
    var dateFormatter = NSDateFormatter.alloc().init();
    // todo: this date format have to be added as property
    dateFormatter.dateFormat = "dd/MM/yyyy";
    if (typeof categoryValue === "string") {
        return dateFormatter.dateFromString(categoryValue);
    }
    else {
        if (categoryValue instanceof Date) {
            return categoryValue;
        }
    }
    console.log("Error: Not supported value set to 'categoryProperty'");
    return undefined;
}
