// nativescript-ui-chart/initializers/chart-initializers.android.js
Object.defineProperty(exports, "__esModule", { value: true });
var observable_array_1 = require("tns-core-modules/data/observable-array");
var chartCommonModule = require("../ui-chart.common");
var color_1 = require("tns-core-modules/color");
var utilsModule = require("tns-core-modules/utils/utils");
var chart_public_enum_1 = require("../misc/chart-public-enum");
var chart_series_common_1 = require("../visualization/views/chart-series.common");
var chart_native_helper_1 = require("../misc/chart-native-helper");
var ChartBaseValueMapper = /** @class */ (function () {
    function ChartBaseValueMapper() {
    }
    ChartBaseValueMapper.prototype.onLegendChanged = function (oldValue, newValue, chart) {
        if (!chart.legend) {
            if (chart.rootLayout) {
                chart.rootLayout.removeAllViews();
            }
            if (chart.androidView) {
                chart.rootLayout.addView(chart.androidView);
            }
        }
    };
    ChartBaseValueMapper.prototype.onPalettesChanged = function (data, chart) {
        if (data.eventName && data.eventName.toLowerCase() === "change") {
            this.loadPalette(chart.palettes, chart);
        }
    };
    ChartBaseValueMapper.prototype.onPalettesPropertyChanged = function (oldValue, newValue, chart) {
        for (var i = 0; i < newValue.length; ++i) {
            newValue.getItem(i).owner = chart;
        }
        this.loadPalette(newValue, chart);
    };
    ChartBaseValueMapper.prototype.loadPalette = function (palettes, chart) {
        if (!chart.androidView) {
            return;
        }
        this._currentChart = chart;
        var normalPalette = chart.androidView.getPalette().clone();
        var selectionPalette = chart.androidView.getSelectionPalette().clone();
        if (chart.palettes) {
            for (var i = 0; i < chart.series.length; i++) {
                var palettesForSeries = this.getPalettesForSeries(palettes, chart.series.getItem(i));
                if (palettesForSeries.length > 0) {
                    this.applyPalettesToSeries(palettesForSeries, chart.series.getItem(i), normalPalette, selectionPalette);
                }
            }
        }
        chart.androidView.setPalette(normalPalette);
        chart.androidView.setSelectionPalette(selectionPalette);
    };
    ChartBaseValueMapper.prototype.getPalettesForSeries = function (source, series) {
        var palettes = [];
        for (var i = 0; i < source.length; i++) {
            var palette = source.getItem(i);
            if (palette.seriesName === series[chart_public_enum_1.seriesName]) {
                palettes.push(palette);
            }
        }
        return palettes;
    };
    ChartBaseValueMapper.prototype.applyPalettesToSeries = function (palettes, series, normalPalette, selectionPalette) {
        var nativeNormalEntries = normalPalette.seriesEntries();
        var nativeSelectionEntries = selectionPalette.seriesEntries();
        if (!nativeNormalEntries || !nativeSelectionEntries) {
            return;
        }
        for (var i = 0; i < palettes.length; ++i) {
            if (palettes[i].seriesState === chart_public_enum_1.ChartPaletteSeriesState.Selected) {
                this.buildNewPaletteForSeries(palettes[i], series, nativeSelectionEntries);
            }
            else {
                this.buildNewPaletteForSeries(palettes[i], series, nativeNormalEntries);
            }
        }
    };
    ChartBaseValueMapper.prototype.buildNewPaletteForSeries = function (palette, series, nativePaletteEntries) {
        var currentEntryCollection = null;
        var sizeOfCurrentEntryCollection = 0;
        for (var i = 0; i < nativePaletteEntries.size(); i++) {
            if (nativePaletteEntries.get(i).getFamily() === series.android.paletteFamily()) {
                currentEntryCollection = nativePaletteEntries.get(i);
                sizeOfCurrentEntryCollection = currentEntryCollection.size();
                break;
            }
        }
        var newNativeEntries = new com.telerik.widget.palettes.PaletteEntryCollection();
        newNativeEntries.setFamily(series.android.paletteFamily());
        if (palette.entries && palette.entries.length > 0) {
            for (var i = 0; i < palette.entries.length; i++) {
                var nativeEntry = new com.telerik.widget.palettes.PaletteEntry();
                var paletteEntry = palette.entries.getItem(i);
                var currentNativeEntry = (currentEntryCollection ? currentEntryCollection.get(i % sizeOfCurrentEntryCollection) : null);
                if (paletteEntry) {
                    if (!isNaN(+paletteEntry.strokeWidth)) {
                        nativeEntry.setStrokeWidth(paletteEntry.strokeWidth);
                    }
                    else {
                        nativeEntry.setStrokeWidth((currentNativeEntry ? currentNativeEntry.getStrokeWidth() : 1));
                    }
                    if (paletteEntry.strokeColor) {
                        nativeEntry.setStroke(paletteEntry.strokeColor.android);
                    }
                    else {
                        nativeEntry.setStroke(currentNativeEntry ? currentNativeEntry.getStroke() : (new color_1.Color("Black")).android);
                    }
                    if (paletteEntry.fillColor) {
                        nativeEntry.setFill(paletteEntry.fillColor.android);
                    }
                    else {
                        nativeEntry.setFill(currentNativeEntry ? currentNativeEntry.getFill() : (new color_1.Color("Blue")).android);
                    }
                    newNativeEntries.add(nativeEntry);
                }
            }
        }
        else {
            return;
        }
        if (series instanceof chart_series_common_1.PieSeries) {
            nativePaletteEntries.remove(currentEntryCollection);
            nativePaletteEntries.add(newNativeEntries);
            return;
        }
        if (series instanceof chart_series_common_1.CartesianSeries) {
            if (series.paletteMode === chart_public_enum_1.ChartSeriesPaletteMode.Item) {
                currentEntryCollection.clear();
                currentEntryCollection.addAll(newNativeEntries);
                return;
            }
        }
        var index = this.getPaletteEntryIndexForSeries(series);
        if (index !== -1) {
            var entry = newNativeEntries.get(index % newNativeEntries.size());
            var i = currentEntryCollection.size();
            while (i < series.android.getCollectionIndex()) {
                // ensure that we have at least as many palette entries as number of series we have
                currentEntryCollection.add(i, entry);
                ++i;
            }
            currentEntryCollection.add(series.android.getCollectionIndex(), entry);
        }
        else {
            nativePaletteEntries.add(newNativeEntries);
        }
    };
    ChartBaseValueMapper.prototype.getPaletteEntryIndexForSeries = function (series) {
        if (!this._currentChart) {
            return -1;
        }
        var itemIndex = 0;
        for (var i = 0; i < this._currentChart.series.length; ++i) {
            if ((this._currentChart.series.getItem(i).seriesName === series.seriesName) && (this._currentChart.series.getItem(i).android.getCollectionIndex() < series.android.getCollectionIndex())) {
                itemIndex++;
            }
        }
        return itemIndex;
    };
    ChartBaseValueMapper.prototype.onSeriesChanged = function (data, chart) {
        if (data.eventName && data.eventName.toLowerCase() === "change") {
            var collectionData = data;
            if (collectionData.action && collectionData.action.toLowerCase() === "add") {
                if (chart.androidView && chart.series) {
                    for (var i = 0; i < collectionData.addedCount; i++) {
                        var currentSeries = chart.series.getItem(collectionData.index + i);
                        chart._addView(currentSeries);
                        chart.androidView.getSeries().add(currentSeries.android);
                        currentSeries.owner = chart;
                        currentSeries.initializer.applyLabelStyle(currentSeries, chart);
                    }
                }
                return;
            }
            if (collectionData.action && collectionData.action.toLowerCase() === "splice") {
                if (collectionData.removed instanceof observable_array_1.ObservableArray) {
                    for (var serieIndex = 0; serieIndex < collectionData.removed.length; serieIndex++) {
                        var currentSeries = collectionData.removed.getItem(serieIndex);
                        chart._removeView(currentSeries);
                        if (chart.androidView.getSeries().indexOf(currentSeries.android) !== -1) {
                            chart.androidView.getSeries().remove(currentSeries.android);
                        }
                    }
                }
                else {
                    for (var serieIndex = 0; serieIndex < collectionData.removed.length; serieIndex++) {
                        var currentSeries = collectionData.removed[serieIndex];
                        chart._removeView(currentSeries);
                        if (chart.androidView.getSeries().indexOf(currentSeries.android) !== -1) {
                            chart.androidView.getSeries().remove(currentSeries.android);
                        }
                    }
                }
                return;
            }
        }
    };
    ChartBaseValueMapper.prototype.onSeriesPropertyChanged = function (oldValue, newValue, chart) {
        if (oldValue) {
            if (chart.androidView) {
                for (var serieIndex = 0; serieIndex < oldValue.length; serieIndex++) {
                    var currentSeries = oldValue.getItem(serieIndex);
                    chart._removeView(currentSeries);
                    if (chart.androidView.getSeries().indexOf(currentSeries.android) !== -1) {
                        chart.androidView.getSeries().remove(currentSeries.android);
                    }
                }
            }
        }
        this.loadSeries(chart);
    };
    ChartBaseValueMapper.prototype.loadSeries = function (chart) {
        if (chart.androidView && chart.series) {
            for (var i = 0; i < chart.series.length; i++) {
                var series = chart.series.getItem(i);
                chart.androidView.getSeries().add(series.android);
                series.owner = chart;
                chart._addView(series);
            }
            if (chart.palettes) {
                this.loadPalette(chart.palettes, chart);
            }
            for (var i = 0; i < chart.series.length; i++) {
                chart.series.getItem(i).initializer.applyLabelStyle(chart.series.getItem(i), chart);
            }
        }
    };
    ChartBaseValueMapper.prototype.reloadPalettes = function (chart) {
        if (chart.palettes) {
            this.loadPalette(chart.palettes, chart);
        }
    };
    ChartBaseValueMapper.prototype.loadAnnotations = function (chart) {
        if (chart.androidView && chart.annotations) {
            for (var i = 0; i < chart.annotations.length; i++) {
                chart.androidView.getAnnotations().add(chart.annotations.getItem(i).android);
            }
        }
    };
    ChartBaseValueMapper.prototype.onAnnotationsChanged = function (data, chart) {
        if (data.eventName && data.eventName.toLowerCase() === "change") {
            // var collectionData = <ChangedData<any>>data;
            if (data.action && data.action.toLowerCase() === "add") {
                if (chart.androidView && chart.annotations) {
                    for (var i = 0; i < data.addedCount; i++) {
                        chart.annotations.getItem(data.index + i)._init(chart);
                        chart.annotations.getItem(data.index + i)._createNative();
                        chart.androidView.getAnnotations().add(chart.annotations.getItem(data.index + i).android);
                    }
                }
                return;
            }
            if (data.action && data.action.toLowerCase() === "splice") {
                for (var annIdx = 0; annIdx < data.removed.length; annIdx++) {
                    if (chart.androidView.getAnnotations().indexOf(data.removed[annIdx].android) !== -1) {
                        chart.androidView.getAnnotations().remove(data.removed[annIdx].android);
                    }
                }
                return;
            }
        }
    };
    ChartBaseValueMapper.prototype.onAnnotationsPropertyChanged = function (oldValue, newValue, chart) {
        if (chart && chart.androidView) {
            chart.androidView.getAnnotations().clear();
            for (var i = 0; i < chart.annotations.length; i++) {
                chart.annotations.getItem(i)._init(chart);
                chart.annotations.getItem(i)._createNative();
            }
            this.loadAnnotations(chart);
        }
    };
    ChartBaseValueMapper.prototype.resolveVerticalAxisForSeries = function (series) {
        var chart = series.owner;
        if (series.verticalAxis) {
            return series.verticalAxis;
        }
        return chart.verticalAxis;
    };
    return ChartBaseValueMapper;
}());
exports.ChartBaseValueMapper = ChartBaseValueMapper;
var CartesianAxisValueMapper = /** @class */ (function () {
    function CartesianAxisValueMapper() {
    }
    CartesianAxisValueMapper.prototype.onLineThicknessChanged = function (oldValue, newValue, axis) {
        if (!isNaN(+newValue)) {
            axis.android.setLineThickness(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onLineColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLineColor(newValue.android);
        }
    };
    CartesianAxisValueMapper.prototype.onLineHiddenChanged = function (oldValue, newValue, axis) {
        axis.android.setShowLine(!newValue);
    };
    CartesianAxisValueMapper.prototype.onLabelTextColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelTextColor(newValue.android);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBackgroundColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelFillColor(newValue.android);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBorderColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelStrokeColor(newValue.android);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelBorderWidthChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelStrokeWidth(newValue);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelMarginChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            axis.android.setLabelMargin(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onLabelRotationAngleChanged = function (oldValue, newValue, axis) {
        if (!isNaN(newValue)) {
            var M_PI = 3.14159265358979323846264338327950288;
            axis.android.setLabelRotationAngle(newValue * 180 / M_PI);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelFitModeChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (axis.labelLayoutMode === chart_public_enum_1.ChartAxisLabelLayoutMode.Inner && (newValue !== chart_public_enum_1.ChartAxisLabelFitMode.None)) {
                console.log("WARNING: Label fit mode is not supported when label layout mode is set to Inner");
                return;
            }
            if (chart_public_enum_1.ChartAxisLabelFitMode.Multiline === newValue) {
                axis.android.setLabelFitMode(com.telerik.widget.chart.engine.axes.common.AxisLabelFitMode.MULTI_LINE);
            }
            else if (chart_public_enum_1.ChartAxisLabelFitMode.Rotate === newValue) {
                axis.android.setLabelFitMode(com.telerik.widget.chart.engine.axes.common.AxisLabelFitMode.ROTATE);
            }
            else {
                axis.android.setLabelFitMode(com.telerik.widget.chart.engine.axes.common.AxisLabelFitMode.NONE);
            }
        }
    };
    CartesianAxisValueMapper.prototype.onLabelLayoutModeChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisLabelLayoutMode.Outer === newValue) {
                axis.android.setLabelLayoutMode(com.telerik.widget.chart.engine.axes.AxisLabelLayoutMode.OUTER);
            }
            else if (chart_public_enum_1.ChartAxisLabelLayoutMode.Inner === newValue) {
                if (axis.android.getLabelFitMode() !== com.telerik.widget.chart.engine.axes.common.AxisLabelFitMode.NONE) {
                    console.log("WARNING: 'Inner' layout mode for axis labels can be combined only with fit mode 'None'.");
                }
                else {
                    axis.android.setLabelLayoutMode(com.telerik.widget.chart.engine.axes.AxisLabelLayoutMode.INNER);
                }
            }
        }
    };
    CartesianAxisValueMapper.prototype.onLabelFormatChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelFormat(java.lang.String.valueOf(newValue));
        }
    };
    CartesianAxisValueMapper.prototype.onTicksHiddenChanged = function (oldValue, newValue, axis) {
        axis.android.getModel().setMajorTickHidden(newValue);
    };
    CartesianAxisValueMapper.prototype.onTicksThicknessChanged = function (oldValue, newValue, axis) {
        if (!isNaN(+newValue)) {
            axis.android.setTickThickness(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onTicksLengthChanged = function (oldValue, newValue, axis) {
        if (!isNaN(+newValue)) {
            axis.android.getModel().setMajorTickLength(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onTicksOffsetChanged = function (oldValue, newValue, axis) {
        if (!isNaN(+newValue)) {
            axis.android.getModel().setMajorTickTranslate(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onTicksColorChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setTickColor(newValue.android);
        }
    };
    CartesianAxisValueMapper.prototype.onHorizontalLocationChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisHorizontalLocation.Left === newValue) {
                axis.android.setHorizontalLocation(com.telerik.widget.chart.engine.axes.common.AxisHorizontalLocation.LEFT);
            }
            else if (chart_public_enum_1.ChartAxisHorizontalLocation.Right === newValue) {
                axis.android.setHorizontalLocation(com.telerik.widget.chart.engine.axes.common.AxisHorizontalLocation.RIGHT);
            }
        }
    };
    CartesianAxisValueMapper.prototype.onVerticalLocationChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            if (chart_public_enum_1.ChartAxisVerticalLocation.Top === newValue) {
                axis.android.setVerticalLocation(com.telerik.widget.chart.engine.axes.common.AxisVerticalLocation.TOP);
            }
            else if (chart_public_enum_1.ChartAxisVerticalLocation.Bottom === newValue) {
                axis.android.setVerticalLocation(com.telerik.widget.chart.engine.axes.common.AxisVerticalLocation.BOTTOM);
            }
        }
    };
    CartesianAxisValueMapper.prototype.onFontInternalChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            var androidFont = newValue.getAndroidTypeface();
            axis.android.setLabelFont(androidFont);
            var androidSize = newValue.fontSize * utilsModule.layout.getDisplayDensity();
            axis.android.setLabelSize(androidSize);
        }
    };
    CartesianAxisValueMapper.prototype.onLabelPaddingChanged = function (newValue, propertyName, axis) {
        var paddingLeftFromStyle = axis.label.style.paddingLeft;
        var paddingLeft = (!propertyName || propertyName === "paddingLeft") && !isNaN(+paddingLeftFromStyle) ?
            +paddingLeftFromStyle : axis.android.getLabelPaddingLeft();
        var paddingTopFromStyle = axis.label.style.paddingTop;
        var paddingTop = (!propertyName || propertyName === "paddingTop") && !isNaN(+paddingTopFromStyle) ?
            +paddingTopFromStyle : axis.android.getLabelPaddingTop();
        var paddingBottomFromStyle = axis.label.style.paddingBottom;
        var paddingBottom = (!propertyName || propertyName === "paddingBottom") && !isNaN(+paddingBottomFromStyle) ?
            +paddingBottomFromStyle : axis.android.getLabelPaddingBottom();
        var paddingRightFromStyle = axis.label.style.paddingRight;
        var paddingRight = (!propertyName || propertyName === "paddingRight") && !isNaN(+paddingRightFromStyle) ?
            +paddingRightFromStyle : axis.android.getLabelPaddingRight();
        axis.android.setLabelPadding(paddingLeft, paddingTop, paddingRight, paddingBottom);
    };
    CartesianAxisValueMapper.prototype.onLabelSizeChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setLabelSize(newValue * utilsModule.layout.getDisplayDensity());
        }
    };
    CartesianAxisValueMapper.prototype.onAllowZoomChanged = function (oldValue, newValue, axis) {
        this.updatePanZoomBehavior(axis);
    };
    CartesianAxisValueMapper.prototype.onAllowPanChanged = function (oldValue, newValue, axis) {
        this.updatePanZoomBehavior(axis);
    };
    CartesianAxisValueMapper.prototype.updatePanZoomBehavior = function (axis) {
        if (!axis.owner) {
            return;
        }
        if (axis.owner instanceof chartCommonModule.RadCartesianChart) {
            axis.owner['updatePanZoomBehavior']();
            return;
        }
        // if owner is series we call its owner
        if (axis.owner.owner instanceof chartCommonModule.RadCartesianChart) {
            axis.owner.owner.updatePanZoomBehavior();
            return;
        }
    };
    CartesianAxisValueMapper.prototype.onHiddenChanged = function (oldValue, newValue, axis) {
        axis.android.setVisible(newValue ? false : true);
    };
    return CartesianAxisValueMapper;
}());
exports.CartesianAxisValueMapper = CartesianAxisValueMapper;
var CategoricalAxisValueMapper = /** @class */ (function (_super) {
    __extends(CategoricalAxisValueMapper, _super);
    function CategoricalAxisValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CategoricalAxisValueMapper.prototype.onMajorTickIntervalChanged = function (oldValue, newValue, axis) {
        if (newValue) {
            axis.android.setMajorTickInterval(java.lang.Integer.valueOf(newValue));
        }
    };
    CategoricalAxisValueMapper.prototype.onPlotModeChanged = function (oldValue, newValue, axis) {
        switch (newValue) {
            case chart_public_enum_1.ChartAxisPlotMode.BetweenTicks:
                axis.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.BETWEEN_TICKS);
                break;
            case chart_public_enum_1.ChartAxisPlotMode.OnTicks:
                axis.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.ON_TICKS);
                break;
            // case ChartAxisPlotMode.OnTicksPadded:
            //     axis.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.ON_TICKS_PADDED);
            //     break;
        }
    };
    CategoricalAxisValueMapper.prototype.onLastLabelVisibilityChanged = function (oldValue, newValue, axis) {
        switch (newValue) {
            case chart_public_enum_1.ChartAxisLabelVisibility.Clip: {
                axis.android.setLastLabelVisibility(com.telerik.widget.chart.engine.axes.common.AxisLastLabelVisibility.CLIP);
                break;
            }
            case chart_public_enum_1.ChartAxisLabelVisibility.Hidden: {
                axis.android.setLastLabelVisibility(com.telerik.widget.chart.engine.axes.common.AxisLastLabelVisibility.HIDDEN);
                break;
            }
            case chart_public_enum_1.ChartAxisLabelVisibility.Visible: {
                axis.android.setLastLabelVisibility(com.telerik.widget.chart.engine.axes.common.AxisLastLabelVisibility.VISIBLE);
                break;
            }
        }
    };
    CategoricalAxisValueMapper.prototype.onFirstLabelVisibilityChanged = function (oldValue, newValue, axis) {
        console.log("Error: Changing the 'firstLabelVisibility' property is not supported in Android.");
    };
    return CategoricalAxisValueMapper;
}(CartesianAxisValueMapper));
exports.CategoricalAxisValueMapper = CategoricalAxisValueMapper;
var ChartSeriesValueMapper = /** @class */ (function () {
    function ChartSeriesValueMapper() {
    }
    ChartSeriesValueMapper.prototype.onShowLabelsChanged = function (oldValue, newValue, series) {
        if (newValue === null || newValue === undefined) {
            return;
        }
        series.android.setShowLabels(newValue);
    };
    ChartSeriesValueMapper.prototype.onLegendTitleChanged = function (newValue, series) {
        series.android.setLegendTitle(newValue);
    };
    ChartSeriesValueMapper.prototype.onLabelPaddingChanged = function (newValue, propertyName, series) {
        var paddingLeftFromStyle = series.label.style.paddingLeft;
        var paddingLeft = (!propertyName || propertyName === "paddingLeft") && !isNaN(+paddingLeftFromStyle) ?
            +paddingLeftFromStyle : series.android.getLabelPaddingLeft();
        var paddingTopFromStyle = series.label.style.paddingTop;
        var paddingTop = (!propertyName || propertyName === "paddingTop") && !isNaN(+paddingTopFromStyle) ?
            +paddingTopFromStyle : series.android.getLabelPaddingTop();
        var paddingBottomFromStyle = series.label.style.paddingBottom;
        var paddingBottom = (!propertyName || propertyName === "paddingBottom") && !isNaN(+paddingBottomFromStyle) ?
            +paddingBottomFromStyle : series.android.getLabelPaddingBottom();
        var paddingRightFromStyle = series.label.style.paddingRight;
        var paddingRight = (!propertyName || propertyName === "paddingRight") && !isNaN(+paddingRightFromStyle) ?
            +paddingRightFromStyle : series.android.getLabelPaddingRight();
        series.android.setLabelPadding(paddingLeft, paddingTop, paddingRight, paddingBottom);
    };
    ChartSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) {
        if (!series.items || !series.items.length) {
            return;
        }
        var length = series.items.length;
        var nativeSource = new java.util.ArrayList();
        for (var i = 0; i < length; i++) {
            var item = series.getItemAtIndex(i);
            if (series['categoryProperty'] && item[series['categoryProperty']] instanceof Date) {
                item[series['categoryProperty']] = item[series['categoryProperty']].getTime();
            }
            nativeSource.add(java.lang.String.valueOf(JSON.stringify(item)));
        }
        series.android.setData(nativeSource);
    };
    ChartSeriesValueMapper.prototype.onValuePropertyChanged = function (oldValue, newValue, series) {
        if (!series.valueProperty) {
            return;
        }
        series.android.setValueBinding(new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[series.valueProperty];
            }
        })));
    };
    ChartSeriesValueMapper.prototype.onLabelStyleChanged = function (oldValue, newValue, series) {
        if (newValue) {
            this.applyLabelStyle(series);
        }
    };
    ChartSeriesValueMapper.prototype.onSelectionModeChanged = function (oldValue, newValue, series) {
        if (!series.android) {
            return;
        }
        switch (newValue) {
            case chart_public_enum_1.ChartSeriesSelectionMode.NotSet:
                series.android.setSelectionMode(com.telerik.widget.chart.visualization.common.SeriesSelectionMode.NOT_SET);
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.None:
                series.android.setSelectionMode(com.telerik.widget.chart.visualization.common.SeriesSelectionMode.NONE);
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.Series:
                series.android.setSelectionMode(com.telerik.widget.chart.visualization.common.SeriesSelectionMode.SERIES);
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.DataPoint:
                series.android.setSelectionMode(com.telerik.widget.chart.visualization.common.SeriesSelectionMode.DATA_POINT_SINGLE);
                break;
            case chart_public_enum_1.ChartSeriesSelectionMode.DataPointMultiple:
                series.android.setSelectionMode(com.telerik.widget.chart.visualization.common.SeriesSelectionMode.DATA_POINT_MULTIPLE);
                break;
        }
    };
    ChartSeriesValueMapper.prototype.onFontInternalChanged = function (oldValue, newValue, series) {
        this.applyFontInternal(newValue, series);
    };
    ChartSeriesValueMapper.prototype.applyFontInternal = function (fontInternal, series) {
        if (fontInternal) {
            var androidFont = fontInternal.getAndroidTypeface();
            if (androidFont) {
                series.android.setLabelFont(androidFont);
            }
            if (!isNaN(fontInternal.fontSize)) {
                var androidSize = fontInternal.fontSize * utilsModule.layout.getDisplayDensity();
                series.android.setLabelSize(androidSize);
            }
        }
    };
    ChartSeriesValueMapper.prototype.applyLabelStyle = function (series, chart) {
        if (!series || !series.labelStyle) {
            return;
        }
        if (!isNaN(+series.labelStyle.textSize)) {
            series.android.setLabelSize(series.labelStyle.textSize * utilsModule.layout.getDisplayDensity());
        }
        if (series.labelStyle.textColor) {
            series.android.setLabelTextColor(series.labelStyle.textColor.android);
        }
        if (series.labelStyle.fillColor) {
            series.android.setLabelFillColor(series.labelStyle.fillColor.android);
        }
        if (series.labelStyle.strokeWidth) {
            series.android.setLabelStrokeWidth(series.labelStyle.strokeWidth);
        }
        if (series.labelStyle.strokeColor) {
            series.android.setLabelStrokeColor(series.labelStyle.strokeColor.android);
        }
        if (!isNaN(+series.labelStyle.margin)) {
            series.android.setLabelMargin(series.labelStyle.margin * utilsModule.layout.getDisplayDensity());
        }
        if (series.labelStyle.textFormat) {
            series.android.setLabelFormat(series.labelStyle.textFormat);
        }
        if (series.label && series.label.style.fontInternal) {
            this.applyFontInternal(series.label.style.fontInternal, series);
        }
        var fontStyle = android.graphics.Typeface.NORMAL;
        switch (series.labelStyle.fontStyle) {
            case chart_public_enum_1.ChartFontStyle.Bold:
                fontStyle = android.graphics.Typeface.BOLD;
                break;
            case chart_public_enum_1.ChartFontStyle.Italic:
                fontStyle = android.graphics.Typeface.ITALIC;
                break;
            case chart_public_enum_1.ChartFontStyle.BoldItalic:
                fontStyle = android.graphics.Typeface.BOLD_ITALIC;
                break;
        }
        if (series.labelStyle.fontName) {
            series.android.setLabelFont(android.graphics.Typeface.create(series.labelStyle.fontName, fontStyle));
        }
        else {
            series.android.setLabelFontStyle(fontStyle);
        }
    };
    return ChartSeriesValueMapper;
}());
exports.ChartSeriesValueMapper = ChartSeriesValueMapper;
var CartesianSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CartesianSeriesValueMapper, _super);
    function CartesianSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CartesianSeriesValueMapper.prototype.onHorizontalAxisChanged = function (oldValue, newValue, series) {
        series.android.setHorizontalAxis(null);
        series.android.setHorizontalAxis(newValue.android);
        newValue.owner = series;
    };
    CartesianSeriesValueMapper.prototype.onVerticalAxisChanged = function (oldValue, newValue, series) {
        series.android.setVerticalAxis(null);
        series.android.setVerticalAxis(newValue.android);
        newValue.owner = series;
    };
    CartesianSeriesValueMapper.prototype.onItemsChanged = function (oldValue, newValue, series) {
        _super.prototype.onItemsChanged.call(this, oldValue, newValue, series);
        if (series.owner) {
            series.owner['_syncZoomValues']();
            // Addresses an issue with ChartPalettes overriding applied label styles after
            // a new items collection is assigned. See ticket: 1105637
            this.applyLabelStyle(series, series.owner);
        }
    };
    CartesianSeriesValueMapper.prototype.onPaletteModeChanged = function (oldValue, newValue, series) {
        var nativePaletteMode = chart_native_helper_1.convertToAndroidPaletteMode(series.paletteMode);
        if (series.android && nativePaletteMode) {
            series.android.setPaletteMode(nativePaletteMode);
        }
    };
    CartesianSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        series.android.setFillColor(newValue.android);
    };
    CartesianSeriesValueMapper.prototype.onStrokeWidthChanged = function (oldValue, newValue, series) {
        var devicePixels = utilsModule.layout.toDevicePixels(newValue);
        series.android.setStrokeWidth(devicePixels);
    };
    CartesianSeriesValueMapper.prototype.onStrokeColorChanged = function (oldValue, newValue, series) {
        series.android.setStrokeColor(newValue.android);
    };
    return CartesianSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.CartesianSeriesValueMapper = CartesianSeriesValueMapper;
var ScatterSeriesValueMapper = /** @class */ (function (_super) {
    __extends(ScatterSeriesValueMapper, _super);
    function ScatterSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScatterSeriesValueMapper.prototype.onXPropertyChanged = function (oldValue, newValue, series) {
        if (!series.xProperty) {
            return;
        }
        var xPropName = series.xProperty;
        series.android.setXValueBinding(new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[xPropName];
            }
        })));
    };
    ScatterSeriesValueMapper.prototype.onYPropertyChanged = function (oldValue, newValue, series) {
        if (!series.yProperty) {
            return;
        }
        var yPropName = series.yProperty;
        series.android.setYValueBinding(new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[yPropName];
            }
        })));
    };
    return ScatterSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.ScatterSeriesValueMapper = ScatterSeriesValueMapper;
var ScatterBubbleSeriesValueMapper = /** @class */ (function (_super) {
    __extends(ScatterBubbleSeriesValueMapper, _super);
    function ScatterBubbleSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScatterBubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue, series) {
        if (!series.bubbleSizeProperty) {
            return;
        }
        var propertyName = series.bubbleSizeProperty;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[propertyName];
            }
        }));
        series.android.setBubbleSizeBinding(binding);
    };
    ScatterBubbleSeriesValueMapper.prototype.onBubbleScaleChanged = function (oldValue, newValue, series) {
        if (newValue) {
            // todo: we use (scale^2) because of bug in Android scale calculation. Update this hack when it is fixed.
            series.android.setBubbleScale(Math.pow(newValue * utilsModule.layout.getDisplayDensity(), 2));
        }
    };
    return ScatterBubbleSeriesValueMapper;
}(ScatterSeriesValueMapper));
exports.ScatterBubbleSeriesValueMapper = ScatterBubbleSeriesValueMapper;
var CategoricalSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CategoricalSeriesValueMapper, _super);
    function CategoricalSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CategoricalSeriesValueMapper.prototype.updateNative = function (series) {
    };
    CategoricalSeriesValueMapper.prototype.onCategoryPropertyChanged = function (oldValue, newValue, series) {
        if (!series.categoryProperty) {
            return;
        }
        series.android.setCategoryBinding(new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[series.categoryProperty];
            }
        })));
    };
    CategoricalSeriesValueMapper.prototype.onStackModePropertyChanged = function (oldValue, newValue, series) {
        if (!newValue) {
            return;
        }
        switch (newValue) {
            case chart_public_enum_1.ChartSeriesStackMode.None:
                series.android.setCombineMode(com.telerik.widget.chart.engine.series.combination.ChartSeriesCombineMode.CLUSTER);
                break;
            case chart_public_enum_1.ChartSeriesStackMode.Stack:
                series.android.setCombineMode(com.telerik.widget.chart.engine.series.combination.ChartSeriesCombineMode.STACK);
                break;
            case chart_public_enum_1.ChartSeriesStackMode.Stack100:
                series.android.setCombineMode(com.telerik.widget.chart.engine.series.combination.ChartSeriesCombineMode.STACK_100);
                break;
        }
    };
    return CategoricalSeriesValueMapper;
}(CartesianSeriesValueMapper));
exports.CategoricalSeriesValueMapper = CategoricalSeriesValueMapper;
var BarSeriesValueMapper = /** @class */ (function (_super) {
    __extends(BarSeriesValueMapper, _super);
    function BarSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BarSeriesValueMapper.prototype.onMinBarSizeChanged = function (oldValue, newValue, series) {
        if (series.android && !isNaN(newValue)) {
            series.android.setMinBarWidth(utilsModule.layout.toDevicePixels(newValue));
        }
        else {
            series.android.setMinBarWidth(0);
        }
    };
    BarSeriesValueMapper.prototype.onMaxBarSizeChanged = function (oldValue, newValue, series) {
        if (series.android && !isNaN(newValue)) {
            series.android.setMaxBarWidth(utilsModule.layout.toDevicePixels(newValue));
        }
        else {
            series.android.setMaxBarWidth(0);
        }
    };
    return BarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BarSeriesValueMapper = BarSeriesValueMapper;
var PieSeriesValueMapper = /** @class */ (function (_super) {
    __extends(PieSeriesValueMapper, _super);
    function PieSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PieSeriesValueMapper.prototype.onLabelPropertyChanged = function (oldValue, newValue, series) {
        if (!series.legendLabel) {
            return;
        }
        series.android.setNameBinding(new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[series.legendLabel];
            }
        })));
    };
    PieSeriesValueMapper.prototype.onLegendTitleChanged = function (newValue, series) {
    };
    PieSeriesValueMapper.prototype.onExpandRadiusChanged = function (oldValue, newValue, series) {
        if (isNaN(+newValue) || newValue < 0.0 || newValue > 1.0) {
            console.log("WARNING: Expand radius factor must be number in range [0,1]");
            return;
        }
        series.android.setSelectedPointOffset(newValue);
    };
    PieSeriesValueMapper.prototype.onOuterRadiusFactorChanged = function (oldValue, newValue, series) {
        if (isNaN(+newValue) || newValue < 0.0 || newValue > 1.0) {
            console.log("WARNING: Outer radius factor must be number in range [0,1]");
            return;
        }
        series.android.setRadiusFactor(newValue);
        series.android.requestLayout();
    };
    PieSeriesValueMapper.prototype.onStartAngleChanged = function (oldValue, newValue, series) {
        if (!isNaN(+newValue)) {
            var sweepAngle = series.endAngle ? series.endAngle - newValue : 360;
            series.android.setAngleRange(new com.telerik.widget.chart.engine.chartAreas.AngleRange(newValue, sweepAngle));
        }
    };
    PieSeriesValueMapper.prototype.onEndAngleChanged = function (oldValue, newValue, series) {
        if (!isNaN(+newValue)) {
            var startAngle = series.startAngle ? series.startAngle : 0;
            var sweepAngle = newValue - startAngle;
            series.android.setAngleRange(new com.telerik.widget.chart.engine.chartAreas.AngleRange(startAngle, sweepAngle));
        }
    };
    PieSeriesValueMapper.prototype.onShowPercentageChanged = function (oldValue, newValue, series) {
        series.android.setDisplayPercentage(series.showPercentage);
    };
    PieSeriesValueMapper.prototype.onFillColorsChanged = function (oldValue, newValue, series) {
        this.applyColors(series);
    };
    PieSeriesValueMapper.prototype.onStrokeWidthChanged = function (oldValue, newValue, series) {
        this.applyColors(series);
    };
    PieSeriesValueMapper.prototype.onStrokeColorsChanged = function (oldValue, newValue, series) {
        this.applyColors(series);
    };
    PieSeriesValueMapper.prototype.applyColors = function (series) {
        var fillColors = series.fillColors;
        var fillColorsLength = fillColors ? fillColors.length : 0;
        var strokeColors = series.strokeColors;
        var strokeColorsLength = strokeColors ? strokeColors.length : 0;
        var length = Math.max(fillColorsLength, strokeColorsLength);
        var nativeSliceStyles = new java.util.ArrayList();
        for (var i = 0; i < length; i++) {
            var style = new com.telerik.widget.chart.visualization.pieChart.SliceStyle();
            if (i < fillColorsLength) {
                style.setFillColor(fillColors[i].android);
            }
            if (i < strokeColorsLength) {
                style.setStrokeColor(strokeColors[i].android);
            }
            if (!isNaN(+series.strokeWidth)) {
                style.setStrokeWidth(+series.strokeWidth);
            }
            nativeSliceStyles.add(style);
        }
        series.android.setSliceStyles(nativeSliceStyles);
    };
    return PieSeriesValueMapper;
}(ChartSeriesValueMapper));
exports.PieSeriesValueMapper = PieSeriesValueMapper;
var DonutSeriesValueMapper = /** @class */ (function (_super) {
    __extends(DonutSeriesValueMapper, _super);
    function DonutSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DonutSeriesValueMapper.prototype.onInnerRadiusFactorChanged = function (oldValue, newValue, series) {
        if (!newValue) {
            return;
        }
        series.android.setInnerRadiusFactor(newValue);
    };
    return DonutSeriesValueMapper;
}(PieSeriesValueMapper));
exports.DonutSeriesValueMapper = DonutSeriesValueMapper;
var RangeBarSeriesValueMapper = /** @class */ (function (_super) {
    __extends(RangeBarSeriesValueMapper, _super);
    function RangeBarSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RangeBarSeriesValueMapper.prototype.onHighPropertyNameChanged = function (oldValue, newValue, series) { };
    RangeBarSeriesValueMapper.prototype.onLowPropertyNameChanged = function (oldValue, newValue, series) { };
    return RangeBarSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.RangeBarSeriesValueMapper = RangeBarSeriesValueMapper;
var LineSeriesValueMapper = /** @class */ (function (_super) {
    __extends(LineSeriesValueMapper, _super);
    function LineSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LineSeriesValueMapper.prototype.onPaletteModeChanged = function (oldValue, newValue, series) {
        console.log('WARNING: Property paletteMode is not applicable for series of type: ' + series.constructor.name);
    };
    LineSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        console.log('WARNING: Property fillColor is not applicable for series of type: ' + series.constructor.name);
    };
    return LineSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.LineSeriesValueMapper = LineSeriesValueMapper;
var SplineSeriesValueMapper = /** @class */ (function (_super) {
    __extends(SplineSeriesValueMapper, _super);
    function SplineSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SplineSeriesValueMapper;
}(LineSeriesValueMapper));
exports.SplineSeriesValueMapper = SplineSeriesValueMapper;
var AreaSeriesValueMapper = /** @class */ (function (_super) {
    __extends(AreaSeriesValueMapper, _super);
    function AreaSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AreaSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        series.android.setFillColor(newValue.android);
    };
    return AreaSeriesValueMapper;
}(LineSeriesValueMapper));
exports.AreaSeriesValueMapper = AreaSeriesValueMapper;
var SplineAreaSeriesValueMapper = /** @class */ (function (_super) {
    __extends(SplineAreaSeriesValueMapper, _super);
    function SplineAreaSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SplineAreaSeriesValueMapper;
}(AreaSeriesValueMapper));
exports.SplineAreaSeriesValueMapper = SplineAreaSeriesValueMapper;
var OhlcSeriesValueMapper = /** @class */ (function (_super) {
    __extends(OhlcSeriesValueMapper, _super);
    function OhlcSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OhlcSeriesValueMapper.prototype.onHighPropertyNameChanged = function (oldValue, newValue, series) { };
    OhlcSeriesValueMapper.prototype.onLowPropertyNameChanged = function (oldValue, newValue, series) { };
    OhlcSeriesValueMapper.prototype.onOpenPropertyNameChanged = function (oldValue, newValue, series) { };
    OhlcSeriesValueMapper.prototype.onClosePropertyNameChanged = function (oldValue, newValue, series) { };
    OhlcSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        console.log('WARNING: Property fillColor is not applicable for series of type: ' + series.constructor.name);
    };
    OhlcSeriesValueMapper.prototype.onStrokeColorsChanged = function (oldValue, newValue, series) {
        if (newValue && newValue.length > 0) {
            series.android.setStrokeColor(newValue[0].android);
            if (newValue.length > 1) {
                series.android.setStrokeSecondColor(newValue[1].android);
            }
        }
    };
    return OhlcSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.OhlcSeriesValueMapper = OhlcSeriesValueMapper;
var CandlestickSeriesValueMapper = /** @class */ (function (_super) {
    __extends(CandlestickSeriesValueMapper, _super);
    function CandlestickSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CandlestickSeriesValueMapper.prototype.onFillColorChanged = function (oldValue, newValue, series) {
        series.android.setFillColor(newValue.android);
    };
    CandlestickSeriesValueMapper.prototype.onFillColorsChanged = function (oldValue, newValue, series) {
        if (newValue && newValue.length > 0) {
            series.android.setFillColor(newValue[0].android);
            if (newValue.length > 1) {
                series.android.setFillSecondColor(newValue[1].android);
            }
        }
    };
    return CandlestickSeriesValueMapper;
}(OhlcSeriesValueMapper));
exports.CandlestickSeriesValueMapper = CandlestickSeriesValueMapper;
var BubbleSeriesValueMapper = /** @class */ (function (_super) {
    __extends(BubbleSeriesValueMapper, _super);
    function BubbleSeriesValueMapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BubbleSeriesValueMapper.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue, series) { };
    BubbleSeriesValueMapper.prototype.onBubbleScalePropertyChanged = function (oldValue, newValue, series) { };
    return BubbleSeriesValueMapper;
}(CategoricalSeriesValueMapper));
exports.BubbleSeriesValueMapper = BubbleSeriesValueMapper;
