// nativescript-ui-chart/visualization/views/chart-axis.android.js
Object.defineProperty(exports, "__esModule", { value: true });
var axisCommonModule = require("./chart-axis.common");
var publicEnumModule = require("../../misc/chart-public-enum");
var dateHelperModule = require("../../misc/chart-date-helper");
var CategoricalAxis = /** @class */ (function (_super) {
    __extends(CategoricalAxis, _super);
    function CategoricalAxis() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(CategoricalAxis.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.axes.CategoricalAxis();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    CategoricalAxis.prototype.onMajorTickIntervalChanged = function (oldValue, newValue) {
        if (newValue) {
            this.android.setMajorTickInterval(newValue);
        }
    };
    return CategoricalAxis;
}(axisCommonModule.CategoricalAxis));
exports.CategoricalAxis = CategoricalAxis;
var DateTimeContinuousAxis = /** @class */ (function (_super) {
    __extends(DateTimeContinuousAxis, _super);
    function DateTimeContinuousAxis() {
        return _super.call(this) || this;
    }
    Object.defineProperty(DateTimeContinuousAxis.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.axes.DateTimeContinuousAxis();
                this._android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.BETWEEN_TICKS);
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    DateTimeContinuousAxis.prototype.onDateFormatChanged = function (oldValue, newValue) {
        if (newValue) {
            this.android.setDateTimeFormat(new java.text.SimpleDateFormat(newValue));
        }
    };
    DateTimeContinuousAxis.prototype.onLabelFormatChanged = function (oldValue, newValue) {
        console.log("WARNING: labelFormat property is not supported for DateTimeCategoricalAxis. Use dateFormat instead.");
    };
    DateTimeContinuousAxis.prototype.onSourceDateFormatChanged = function (oldValue, newValue) {
        if (newValue) {
            this.android.setSourceDateTimeFormat(new java.text.SimpleDateFormat(newValue));
        }
    };
    DateTimeContinuousAxis.prototype.onPlotModeChanged = function (oldValue, newValue) {
        switch (newValue) {
            case publicEnumModule.ChartAxisPlotMode.BetweenTicks:
                this.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.BETWEEN_TICKS);
                break;
            case publicEnumModule.ChartAxisPlotMode.OnTicks:
                this.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.ON_TICKS);
                break;
            // case publicEnumModule.ChartAxisPlotMode.OnTicksPadded:
            //     axis.android.setPlotMode(com.telerik.widget.chart.engine.axes.common.AxisPlotMode.ON_TICKS_PADDED);
            //     break;
        }
    };
    DateTimeContinuousAxis.prototype.onMinimumChanged = function (oldValue, newValue) {
        if (newValue === undefined) {
            this.android.setMinimum(undefined);
            return;
        }
        if (typeof newValue === "string") {
            var formatter = new java.text.SimpleDateFormat("dd/MM/yyyy");
            var parsedDate = formatter.parse(newValue);
            var millis = parsedDate.getTime();
            var nativeValue = java.util.Calendar.getInstance();
            nativeValue.setTimeInMillis(millis);
            this.android.setMinimum(nativeValue);
        }
        else {
            this.android.setMinimum(dateHelperModule.getCalendarFromDate(newValue));
        }
    };
    DateTimeContinuousAxis.prototype.onMaximumChanged = function (oldValue, newValue) {
        if (newValue === undefined) {
            this.android.setMaximum(undefined);
            return;
        }
        if (typeof newValue === "string") {
            var formatter = new java.text.SimpleDateFormat("dd/MM/yyyy");
            var parsedDate = formatter.parse(newValue);
            var millis = parsedDate.getTime();
            var nativeValue = java.util.Calendar.getInstance();
            nativeValue.setTimeInMillis(millis);
            this.android.setMaximum(nativeValue);
        }
        else {
            this.android.setMaximum(dateHelperModule.getCalendarFromDate(newValue));
        }
    };
    DateTimeContinuousAxis.prototype.onMajorStepUnitChanged = function (oldValue, newValue) {
        switch (newValue) {
            case publicEnumModule.ChartAxisDateTimeComponent.Second:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.SECOND);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Minute:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.MINUTE);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Hour:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.HOUR);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Day:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.DAY);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Week:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.WEEK);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Month:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.MONTH);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Year:
                this.android.setMajorStepUnit(com.telerik.widget.chart.engine.axes.common.TimeInterval.YEAR);
                break;
        }
    };
    DateTimeContinuousAxis.prototype.onMajorStepChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMajorStep(+newValue);
        }
    };
    return DateTimeContinuousAxis;
}(axisCommonModule.DateTimeContinuousAxis));
exports.DateTimeContinuousAxis = DateTimeContinuousAxis;
var DateTimeCategoricalAxis = /** @class */ (function (_super) {
    __extends(DateTimeCategoricalAxis, _super);
    function DateTimeCategoricalAxis() {
        return _super.call(this) || this;
    }
    Object.defineProperty(DateTimeCategoricalAxis.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.axes.DateTimeCategoricalAxis();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    DateTimeCategoricalAxis.prototype.onDateFormatChanged = function (oldValue, newValue) {
        if (newValue) {
            this.android.setDateTimeFormat(new java.text.SimpleDateFormat(newValue));
        }
    };
    DateTimeCategoricalAxis.prototype.onLabelFormatChanged = function (oldValue, newValue) {
        console.log("WARNING: labelFormat property is not supported for DateTimeCategoricalAxis. Use dateFormat instead.");
    };
    DateTimeCategoricalAxis.prototype.onDateTimeComponentChanged = function (oldValue, newValue) {
        switch (newValue) {
            case publicEnumModule.ChartAxisDateTimeComponent.Second:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.SECOND);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Minute:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.MINUTE);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Hour:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.HOUR);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Day:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.DAY);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Week:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.WEEK);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Month:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.MONTH);
                break;
            case publicEnumModule.ChartAxisDateTimeComponent.Year:
                this.android.setDateTimeComponent(com.telerik.widget.chart.engine.axes.common.DateTimeComponent.YEAR);
                break;
        }
    };
    return DateTimeCategoricalAxis;
}(axisCommonModule.DateTimeCategoricalAxis));
exports.DateTimeCategoricalAxis = DateTimeCategoricalAxis;
////////////////////////////////////////////////////////////////////////
// LogarithmicAxis
////////////////////////////////////////////////////////////////////////
var LogarithmicAxis = /** @class */ (function (_super) {
    __extends(LogarithmicAxis, _super);
    function LogarithmicAxis() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(LogarithmicAxis.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.axes.LogarithmicAxis();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    LogarithmicAxis.prototype.onExponentStepChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue) && newValue > 0) {
            this.android.setExponentStep(newValue);
        }
    };
    LogarithmicAxis.prototype.onLogarithmBaseChanged = function (oldValue, newValue) {
        if (!isNaN(newValue)) {
            this.android.setLogarithmBase(newValue);
        }
    };
    LogarithmicAxis.prototype.onMajorStepChanged = function (oldValue, newValue) {
        console.log("WARNING: majorStep property is not used for LogarithmicAxis. Use exponentStep property instead.");
    };
    LogarithmicAxis.prototype.onMinimumChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMinimum(parseFloat(newValue));
        }
    };
    LogarithmicAxis.prototype.onMaximumChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMaximum(parseFloat(newValue));
        }
    };
    return LogarithmicAxis;
}(axisCommonModule.LogarithmicAxis));
exports.LogarithmicAxis = LogarithmicAxis;
var LinearAxis = /** @class */ (function (_super) {
    __extends(LinearAxis, _super);
    function LinearAxis() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(LinearAxis.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.axes.LinearAxis();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    LinearAxis.prototype.onMajorStepChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMajorStep(+newValue);
        }
    };
    LinearAxis.prototype.onMinimumChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMinimum(parseFloat(newValue));
        }
    };
    LinearAxis.prototype.onMaximumChanged = function (oldValue, newValue) {
        if (!isNaN(+newValue)) {
            this.android.setMaximum(parseFloat(newValue));
        }
    };
    return LinearAxis;
}(axisCommonModule.LinearAxis));
exports.LinearAxis = LinearAxis;
