// nativescript-ui-chart/visualization/views/chart-legend-view.common.js
Object.defineProperty(exports, "__esModule", { value: true });
var view_1 = require("tns-core-modules/ui/core/view");
var chart_public_enum_1 = require("../../misc/chart-public-enum");
var color_1 = require("tns-core-modules/color");
var RadLegendView = /** @class */ (function (_super) {
    __extends(RadLegendView, _super);
    function RadLegendView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RadLegendView.prototype.onTitleSizeChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onTitleColorChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onEnableSelectionChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onPositionPropertyChanged = function (oldValue, newValue) {
        this.onPositionChanged(oldValue, newValue);
    };
    RadLegendView.prototype.onPositionChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onOffsetOriginPropertyChanged = function (oldValue, newValue) {
        this.onOffsetOriginChanged(oldValue, newValue);
    };
    RadLegendView.prototype.onOffsetOriginChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onHorizontalOffsetPropertyChanged = function (oldValue, newValue) {
        this.onHorizontalOffsetChanged(oldValue, newValue);
    };
    RadLegendView.prototype.onHorizontalOffsetChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onVerticalOffsetPropertyChanged = function (oldValue, newValue) {
        this.onVerticalOffsetChanged(oldValue, newValue);
    };
    RadLegendView.prototype.onVerticalOffsetChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.onTitlePropertyChanged = function (oldValue, newValue) {
        this.onTitleChanged(oldValue, newValue);
    };
    RadLegendView.prototype.onTitleChanged = function (oldValue, newValue) {
    };
    RadLegendView.prototype.updateLegendView = function (chartView) {
    };
    RadLegendView.prototype.updateLegendTitle = function () {
    };
    RadLegendView.prototype.updateListViewForTheme = function () {
    };
    RadLegendView.titleSizeProperty = new view_1.Property({
        name: 'titleSize',
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onTitleSizeChanged(oldValue, newValue);
        }
    });
    RadLegendView.titleColorProperty = new view_1.Property({
        name: 'titleColor',
        defaultValue: undefined,
        equalityComparer: color_1.Color.equals,
        valueConverter: function (v) { return new color_1.Color(v); },
        valueChanged: function (target, oldValue, newValue) {
            target.onTitleColorChanged(oldValue, newValue);
        }
    });
    RadLegendView.enableSelectionProperty = new view_1.Property({
        name: 'enableSelection',
        defaultValue: false,
        valueConverter: view_1.booleanConverter,
        valueChanged: function (target, oldValue, newValue) {
            target.onEnableSelectionChanged(oldValue, newValue);
        }
    });
    RadLegendView.positionProperty = new view_1.Property({
        name: "position",
        defaultValue: chart_public_enum_1.ChartLegendPosition.Bottom,
        valueConverter: function (value) { return chart_public_enum_1.ChartLegendPosition[chart_public_enum_1.toPascalCase(value)]; },
        valueChanged: function (target, oldValue, newValue) {
            target.onPositionPropertyChanged(oldValue, newValue);
        },
    });
    RadLegendView.offsetOriginProperty = new view_1.Property({
        name: "offsetOrigin",
        defaultValue: chart_public_enum_1.ChartLegendOffsetOrigin.TopLeft,
        valueConverter: function (value) { return chart_public_enum_1.ChartLegendOffsetOrigin[chart_public_enum_1.toPascalCase(value)]; },
        valueChanged: function (target, oldValue, newValue) {
            target.onOffsetOriginPropertyChanged(oldValue, newValue);
        },
    });
    RadLegendView.horizontalOffsetProperty = new view_1.Property({
        name: "horizontalOffset",
        defaultValue: 0,
        valueConverter: parseFloat,
        valueChanged: function (target, oldValue, newValue) {
            target.onHorizontalOffsetPropertyChanged(oldValue, newValue);
        },
    });
    RadLegendView.verticalOffsetProperty = new view_1.Property({
        name: "verticalOffset",
        defaultValue: 0,
        valueConverter: parseFloat,
        valueChanged: function (target, oldValue, newValue) {
            target.onVerticalOffsetPropertyChanged(oldValue, newValue);
        },
    });
    RadLegendView.titleProperty = new view_1.Property({
        name: "title",
        defaultValue: undefined,
        valueChanged: function (target, oldValue, newValue) {
            target.onTitlePropertyChanged(oldValue, newValue);
        },
    });
    return RadLegendView;
}(view_1.View));
exports.RadLegendView = RadLegendView;
RadLegendView.positionProperty.register(RadLegendView);
RadLegendView.offsetOriginProperty.register(RadLegendView);
RadLegendView.horizontalOffsetProperty.register(RadLegendView);
RadLegendView.verticalOffsetProperty.register(RadLegendView);
RadLegendView.titleProperty.register(RadLegendView);
RadLegendView.titleSizeProperty.register(RadLegendView);
RadLegendView.titleColorProperty.register(RadLegendView);
RadLegendView.enableSelectionProperty.register(RadLegendView);
