// nativescript-ui-chart/visualization/views/chart-series.android.js
Object.defineProperty(exports, "__esModule", { value: true });
var seriesCommonModule = require("./chart-series.common");
var utilsModule = require("tns-core-modules/utils/utils");
var initializersImpl = require("../../initializers/chart-initializers");
var view_1 = require("tns-core-modules/ui/core/view");
var color_1 = require("tns-core-modules/color");
var PieSeries = /** @class */ (function (_super) {
    __extends(PieSeries, _super);
    function PieSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PieSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.pieChart.PieSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    PieSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return PieSeries;
}(seriesCommonModule.PieSeries));
exports.PieSeries = PieSeries;
var DonutSeries = /** @class */ (function (_super) {
    __extends(DonutSeries, _super);
    function DonutSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(DonutSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.pieChart.DoughnutSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    DonutSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return DonutSeries;
}(seriesCommonModule.DonutSeries));
exports.DonutSeries = DonutSeries;
var LineSeries = /** @class */ (function (_super) {
    __extends(LineSeries, _super);
    function LineSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(LineSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.LineSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    LineSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    Object.defineProperty(LineSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.LineSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    LineSeries = __decorate([
        view_1.CSSType("LineSeries"),
        __metadata("design:paramtypes", [])
    ], LineSeries);
    return LineSeries;
}(seriesCommonModule.CategoricalSeries));
exports.LineSeries = LineSeries;
var OhlcSeries = /** @class */ (function (_super) {
    __extends(OhlcSeries, _super);
    function OhlcSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(OhlcSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.OhlcSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    OhlcSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    // TODO: 'setNative' currently is not raised on Android
    // [seriesCommonModule.OhlcSeries.legendTitleProperty.setNative](newValue: string): void {
    //     this.initializer.onLegendTitleChanged(newValue, this);
    // }
    OhlcSeries.prototype.onValuePropertyChanged = function (oldValue, newValue) {
    };
    OhlcSeries.prototype.onHighPropertyNameChanged = function (oldValue, newValue) {
        if (!newValue) {
            return;
        }
        var highPropertyName = this.highPropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[highPropertyName];
            }
        }));
        this.android.setHighBinding(binding);
    };
    OhlcSeries.prototype.onLowPropertyNameChanged = function (oldValue, newValue) {
        var lowPropertyName = this.lowPropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[lowPropertyName];
            }
        }));
        this.android.setLowBinding(binding);
    };
    OhlcSeries.prototype.onOpenPropertyNameChanged = function (oldValue, newValue) {
        var openPropertyName = this.openPropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[openPropertyName];
            }
        }));
        this.android.setOpenBinding(binding);
    };
    OhlcSeries.prototype.onClosePropertyNameChanged = function (oldValue, newValue) {
        var closePropertyName = this.closePropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[closePropertyName];
            }
        }));
        this.android.setCloseBinding(binding);
    };
    return OhlcSeries;
}(seriesCommonModule.OhlcSeries));
exports.OhlcSeries = OhlcSeries;
var CandlestickSeries = /** @class */ (function (_super) {
    __extends(CandlestickSeries, _super);
    function CandlestickSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(CandlestickSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.CandlestickSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    CandlestickSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    CandlestickSeries.prototype.onFillColorsPropertyChanged = function (oldValue, newValue) {
        this.onFillColorsChanged(oldValue, newValue);
    };
    CandlestickSeries.prototype.onFillColorsChanged = function (oldValue, newValue) {
        this.initializer.onFillColorsChanged(oldValue, newValue, this);
    };
    Object.defineProperty(CandlestickSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.CandlestickSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    CandlestickSeries.fillColorsProperty = new view_1.Property({
        name: "fillColors",
        defaultValue: undefined,
        valueConverter: function (v) { return v.split(",").map(function (v1) { return new color_1.Color(v1); }); },
        valueChanged: function (target, oldValue, newValue) {
            target.onFillColorsPropertyChanged(oldValue, newValue);
        },
    });
    CandlestickSeries = __decorate([
        view_1.CSSType("CandlestickSeries"),
        __metadata("design:paramtypes", [])
    ], CandlestickSeries);
    return CandlestickSeries;
}(OhlcSeries));
exports.CandlestickSeries = CandlestickSeries;
CandlestickSeries.fillColorsProperty.register(CandlestickSeries);
var SplineSeries = /** @class */ (function (_super) {
    __extends(SplineSeries, _super);
    function SplineSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(SplineSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.SplineSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    SplineSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    SplineSeries = __decorate([
        view_1.CSSType("SplineSeries"),
        __metadata("design:paramtypes", [])
    ], SplineSeries);
    return SplineSeries;
}(LineSeries));
exports.SplineSeries = SplineSeries;
var AreaSeries = /** @class */ (function (_super) {
    __extends(AreaSeries, _super);
    function AreaSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(AreaSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.AreaSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AreaSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.AreaSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    AreaSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    AreaSeries = __decorate([
        view_1.CSSType("AreaSeries"),
        __metadata("design:paramtypes", [])
    ], AreaSeries);
    return AreaSeries;
}(seriesCommonModule.CategoricalSeries));
exports.AreaSeries = AreaSeries;
var SplineAreaSeries = /** @class */ (function (_super) {
    __extends(SplineAreaSeries, _super);
    function SplineAreaSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(SplineAreaSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.SplineAreaSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    SplineAreaSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    SplineAreaSeries = __decorate([
        view_1.CSSType("SplineAreaSeries"),
        __metadata("design:paramtypes", [])
    ], SplineAreaSeries);
    return SplineAreaSeries;
}(AreaSeries));
exports.SplineAreaSeries = SplineAreaSeries;
var ScatterBubbleSeries = /** @class */ (function (_super) {
    __extends(ScatterBubbleSeries, _super);
    function ScatterBubbleSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(ScatterBubbleSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.scatter.ScatterBubbleSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    ScatterBubbleSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return ScatterBubbleSeries;
}(seriesCommonModule.ScatterBubbleSeries));
exports.ScatterBubbleSeries = ScatterBubbleSeries;
var BubbleSeries = /** @class */ (function (_super) {
    __extends(BubbleSeries, _super);
    function BubbleSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(BubbleSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.BubbleSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    BubbleSeries.prototype.onBubbleScaleChanged = function (oldValue, newValue) {
        if (newValue) {
            // todo: we use (scale^2) because of bug in Android scale calculation. Update this hack when it is fixed.
            this.android.setBubbleScale(Math.pow(newValue * utilsModule.layout.getDisplayDensity(), 2));
        }
    };
    BubbleSeries.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue) {
        if (!this.bubbleSizeProperty) {
            return;
        }
        var propertyName = this.bubbleSizeProperty;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[propertyName];
            }
        }));
        this.android.setBubbleSizeBinding(binding);
    };
    BubbleSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return BubbleSeries;
}(seriesCommonModule.BubbleSeries));
exports.BubbleSeries = BubbleSeries;
var ScatterSeries = /** @class */ (function (_super) {
    __extends(ScatterSeries, _super);
    function ScatterSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ScatterSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.scatter.ScatterPointSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    ScatterSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return ScatterSeries;
}(seriesCommonModule.ScatterSeries));
exports.ScatterSeries = ScatterSeries;
var BarSeries = /** @class */ (function (_super) {
    __extends(BarSeries, _super);
    function BarSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(BarSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.BarSeries();
                this._android.setCombineMode(com.telerik.widget.chart.engine.series.combination.ChartSeriesCombineMode.CLUSTER);
                if (!isNaN(this.minBarSize)) {
                    this._android.setMinBarWidth(utilsModule.layout.toDevicePixels(this.minBarSize));
                }
                else {
                    this._android.setMinBarWidth(0);
                }
                if (!isNaN(this.maxBarSize)) {
                    this._android.setMaxBarWidth(utilsModule.layout.toDevicePixels(this.maxBarSize));
                }
                else {
                    this._android.setMaxBarWidth(0);
                }
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    BarSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return BarSeries;
}(seriesCommonModule.BarSeries));
exports.BarSeries = BarSeries;
var RangeBarSeries = /** @class */ (function (_super) {
    __extends(RangeBarSeries, _super);
    function RangeBarSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(RangeBarSeries.prototype, "android", {
        get: function () {
            if (!this._android) {
                this._android = new com.telerik.widget.chart.visualization.cartesianChart.series.categorical.RangeBarSeries();
            }
            return this._android;
        },
        enumerable: true,
        configurable: true
    });
    RangeBarSeries.prototype.onValuePropertyChanged = function (oldValue, newValue) {
        console.log("WARNING: Range bar series doesn't use valueProperty property.");
    };
    RangeBarSeries.prototype.onHighPropertyNameChanged = function (oldValue, newValue) {
        if (!newValue) {
            return;
        }
        var highPropertyName = this.highPropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[highPropertyName];
            }
        }));
        this.android.setHighBinding(binding);
    };
    RangeBarSeries.prototype.onLowPropertyNameChanged = function (oldValue, newValue) {
        var lowPropertyName = this.lowPropertyName;
        var binding = new com.telerik.widget.chart.engine.databinding.GenericDataPointBinding(new com.telerik.android.common.Function({
            apply: function (arg) {
                var item = JSON.parse(arg);
                return item[lowPropertyName];
            }
        }));
        this.android.setLowBinding(binding);
    };
    RangeBarSeries.prototype.onLegendTitleChanged = function (oldValue, newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return RangeBarSeries;
}(seriesCommonModule.RangeBarSeries));
exports.RangeBarSeries = RangeBarSeries;
