// nativescript-ui-chart/visualization/views/chart-series.ios.js
Object.defineProperty(exports, "__esModule", { value: true });
var seriesCommonModule = require("./chart-series.common");
var initializersImpl = require("../../initializers/chart-initializers");
var commonModule = require("../../ui-chart.common");
var view_1 = require("tns-core-modules/ui/core/view");
var view_2 = require("tns-core-modules/ui/core/view");
var color_1 = require("tns-core-modules/color");
var PieSeries = /** @class */ (function (_super) {
    __extends(PieSeries, _super);
    function PieSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PieSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    PieSeries.prototype[seriesCommonModule.PieSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return PieSeries;
}(seriesCommonModule.PieSeries));
exports.PieSeries = PieSeries;
var DonutSeries = /** @class */ (function (_super) {
    __extends(DonutSeries, _super);
    function DonutSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(DonutSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    DonutSeries.prototype.updateOwnerChart = function () {
        if (this.owner && (this.owner instanceof commonModule.RadPieChart)) {
            this.owner.updateChart();
        }
    };
    Object.defineProperty(DonutSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.DonutSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    DonutSeries.prototype.onInnerRadiusFactorChanged = function (oldValue, newValue) {
        this.initializer.onInnerRadiusFactorChanged(oldValue, newValue, this);
    };
    DonutSeries.prototype[seriesCommonModule.DonutSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return DonutSeries;
}(seriesCommonModule.DonutSeries));
exports.DonutSeries = DonutSeries;
////////////////////////////////////////////////////////////////////////
// BarSeries
////////////////////////////////////////////////////////////////////////
var BarSeries = /** @class */ (function (_super) {
    __extends(BarSeries, _super);
    function BarSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(BarSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    BarSeries.prototype[seriesCommonModule.BarSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    BarSeries.prototype.updateNative = function () {
        this.initializer.updateNative(this);
    };
    return BarSeries;
}(seriesCommonModule.BarSeries));
exports.BarSeries = BarSeries;
////////////////////////////////////////////////////////////////////////
// RangeBarSeries
////////////////////////////////////////////////////////////////////////
var RangeBarSeries = /** @class */ (function (_super) {
    __extends(RangeBarSeries, _super);
    function RangeBarSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(RangeBarSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.RangeBarSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeBarSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    RangeBarSeries.prototype.updateNative = function () {
        this.initializer.updateNative(this);
    };
    RangeBarSeries.prototype[seriesCommonModule.RangeBarSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    RangeBarSeries.prototype.onHighPropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onHighPropertyNameChanged(oldValue, newValue, this);
    };
    RangeBarSeries.prototype.onLowPropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onLowPropertyNameChanged(oldValue, newValue, this);
    };
    RangeBarSeries.prototype.onValuePropertyChanged = function (oldValue, newValue) {
        console.log("WARNING: Range bar series doesn't use valueProperty property.");
    };
    return RangeBarSeries;
}(seriesCommonModule.RangeBarSeries));
exports.RangeBarSeries = RangeBarSeries;
////////////////////////////////////////////////////////////////////////
// LineSeries
////////////////////////////////////////////////////////////////////////
var LineSeries = /** @class */ (function (_super) {
    __extends(LineSeries, _super);
    function LineSeries() {
        return _super.call(this) || this;
    }
    Object.defineProperty(LineSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.LineSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LineSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    LineSeries.prototype[seriesCommonModule.CategoricalSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    LineSeries = __decorate([
        view_1.CSSType("LineSeries"),
        __metadata("design:paramtypes", [])
    ], LineSeries);
    return LineSeries;
}(seriesCommonModule.CategoricalSeries));
exports.LineSeries = LineSeries;
var SplineSeries = /** @class */ (function (_super) {
    __extends(SplineSeries, _super);
    function SplineSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(SplineSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.SplineSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    SplineSeries.prototype[LineSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    SplineSeries = __decorate([
        view_1.CSSType("SplineSeries")
    ], SplineSeries);
    return SplineSeries;
}(LineSeries));
exports.SplineSeries = SplineSeries;
var AreaSeries = /** @class */ (function (_super) {
    __extends(AreaSeries, _super);
    function AreaSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(AreaSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.AreaSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    AreaSeries.prototype[LineSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    AreaSeries = __decorate([
        view_1.CSSType("AreaSeries")
    ], AreaSeries);
    return AreaSeries;
}(LineSeries));
exports.AreaSeries = AreaSeries;
var SplineAreaSeries = /** @class */ (function (_super) {
    __extends(SplineAreaSeries, _super);
    function SplineAreaSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(SplineAreaSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.SplineAreaSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    SplineAreaSeries.prototype[LineSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    SplineAreaSeries = __decorate([
        view_1.CSSType("SplineAreaSeries")
    ], SplineAreaSeries);
    return SplineAreaSeries;
}(AreaSeries));
exports.SplineAreaSeries = SplineAreaSeries;
var BubbleSeries = /** @class */ (function (_super) {
    __extends(BubbleSeries, _super);
    function BubbleSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(BubbleSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BubbleSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.BubbleSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    BubbleSeries.prototype.onBubbleScaleChanged = function (oldValue, newValue) {
        this.initializer.onBubbleScalePropertyChanged(oldValue, newValue, this);
    };
    BubbleSeries.prototype.onBubbleSizePropertyChanged = function (oldValue, newValue) {
        this.initializer.onBubbleSizePropertyChanged(oldValue, newValue, this);
    };
    BubbleSeries.prototype[seriesCommonModule.BubbleSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return BubbleSeries;
}(seriesCommonModule.BubbleSeries));
exports.BubbleSeries = BubbleSeries;
var ScatterSeries = /** @class */ (function (_super) {
    __extends(ScatterSeries, _super);
    function ScatterSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ScatterSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    ScatterSeries.prototype[seriesCommonModule.ScatterSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return ScatterSeries;
}(seriesCommonModule.ScatterSeries));
exports.ScatterSeries = ScatterSeries;
var ScatterBubbleSeries = /** @class */ (function (_super) {
    __extends(ScatterBubbleSeries, _super);
    function ScatterBubbleSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ScatterBubbleSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    ScatterBubbleSeries.prototype[seriesCommonModule.ScatterBubbleSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    return ScatterBubbleSeries;
}(seriesCommonModule.ScatterBubbleSeries));
exports.ScatterBubbleSeries = ScatterBubbleSeries;
var OhlcSeries = /** @class */ (function (_super) {
    __extends(OhlcSeries, _super);
    function OhlcSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(OhlcSeries.prototype, "ios", {
        get: function () {
            return this._series;
        },
        set: function (value) {
            this._series = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OhlcSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.OhlcSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    OhlcSeries.prototype[seriesCommonModule.OhlcSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    OhlcSeries.prototype.onHighPropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onHighPropertyNameChanged(oldValue, newValue, this);
    };
    OhlcSeries.prototype.onLowPropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onLowPropertyNameChanged(oldValue, newValue, this);
    };
    OhlcSeries.prototype.onOpenPropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onOpenPropertyNameChanged(oldValue, newValue, this);
    };
    OhlcSeries.prototype.onClosePropertyNameChanged = function (oldValue, newValue) {
        this.initializer.onClosePropertyNameChanged(oldValue, newValue, this);
    };
    OhlcSeries.prototype.onValuePropertyChanged = function (oldValue, newValue) {
        console.log("WARNING: OHLC series doesn't use valueProperty property.");
    };
    return OhlcSeries;
}(seriesCommonModule.OhlcSeries));
exports.OhlcSeries = OhlcSeries;
var CandlestickSeries = /** @class */ (function (_super) {
    __extends(CandlestickSeries, _super);
    function CandlestickSeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(CandlestickSeries.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        set: function (value) {
            this._ios = value;
        },
        enumerable: true,
        configurable: true
    });
    CandlestickSeries.prototype.onFillColorsPropertyChanged = function (oldValue, newValue) {
        this.onFillColorsChanged(oldValue, newValue);
    };
    CandlestickSeries.prototype.onFillColorsChanged = function (oldValue, newValue) {
        this.initializer.onFillColorsChanged(oldValue, newValue, this);
    };
    Object.defineProperty(CandlestickSeries.prototype, "initializer", {
        get: function () {
            if (!this._initializer) {
                this._initializer = new initializersImpl.CandlestickSeriesValueMapper();
            }
            return this._initializer;
        },
        enumerable: true,
        configurable: true
    });
    CandlestickSeries.prototype[OhlcSeries.legendTitleProperty.setNative] = function (newValue) {
        this.initializer.onLegendTitleChanged(newValue, this);
    };
    CandlestickSeries.fillColorsProperty = new view_2.Property({
        name: "fillColors",
        defaultValue: undefined,
        valueConverter: function (v) { return v.split(",").map(function (v1) { return new color_1.Color(v1); }); },
        valueChanged: function (target, oldValue, newValue) {
            target.onFillColorsPropertyChanged(oldValue, newValue);
        },
    });
    CandlestickSeries = __decorate([
        view_1.CSSType("CandlestickSeries")
    ], CandlestickSeries);
    return CandlestickSeries;
}(OhlcSeries));
exports.CandlestickSeries = CandlestickSeries;
CandlestickSeries.fillColorsProperty.register(CandlestickSeries);
