// @nativescript/core/ui/index.js
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var action_bar_1 = require("./action-bar");
exports.ActionBar = action_bar_1.ActionBar;
exports.ActionItem = action_bar_1.ActionItem;
exports.ActionItems = action_bar_1.ActionItems;
exports.NavigationButton = action_bar_1.NavigationButton;
var activity_indicator_1 = require("./activity-indicator");
exports.ActivityIndicator = activity_indicator_1.ActivityIndicator;
var animation_1 = require("./animation");
exports.Animation = animation_1.Animation;
var bottom_navigation_1 = require("./bottom-navigation");
exports.BottomNavigation = bottom_navigation_1.BottomNavigation;
var builder_1 = require("./builder");
exports.Builder = builder_1.Builder;
var button_1 = require("./button");
exports.Button = button_1.Button;
var content_view_1 = require("./content-view");
exports.ContentView = content_view_1.ContentView;
var view_base_1 = require("./core/view-base");
exports.ViewBase = view_base_1.ViewBase;
var view_1 = require("./core/view");
exports.View = view_1.View;
var date_picker_1 = require("./date-picker");
exports.DatePicker = date_picker_1.DatePicker;
var editable_text_base_1 = require("./editable-text-base");
exports.EditableTextBase = editable_text_base_1.EditableTextBase;
var frame_1 = require("./frame");
exports.Frame = frame_1.Frame;
var gestures_1 = require("./gestures");
exports.GestureStateTypes = gestures_1.GestureStateTypes;
exports.GestureTypes = gestures_1.GestureTypes;
exports.GesturesObserver = gestures_1.GesturesObserver;
exports.SwipeDirection = gestures_1.SwipeDirection;
var html_view_1 = require("./html-view");
exports.HtmlView = html_view_1.HtmlView;
var image_1 = require("./image");
exports.Image = image_1.Image;
var image_cache_1 = require("./image-cache");
exports.ImageCache = image_cache_1.Cache;
var label_1 = require("./label");
exports.Label = label_1.Label;
__export(require("./layouts"));
var list_picker_1 = require("./list-picker");
exports.ListPicker = list_picker_1.ListPicker;
var list_view_1 = require("./list-view");
exports.ListView = list_view_1.ListView;
var page_1 = require("./page");
exports.Page = page_1.Page;
var placeholder_1 = require("./placeholder");
exports.Placeholder = placeholder_1.Placeholder;
var progress_1 = require("./progress");
exports.Progress = progress_1.Progress;
var proxy_view_container_1 = require("./proxy-view-container");
exports.ProxyViewContainer = proxy_view_container_1.ProxyViewContainer;
var repeater_1 = require("./repeater");
exports.Repeater = repeater_1.Repeater;
var scroll_view_1 = require("./scroll-view");
exports.ScrollView = scroll_view_1.ScrollView;
var search_bar_1 = require("./search-bar");
exports.SearchBar = search_bar_1.SearchBar;
var segmented_bar_1 = require("./segmented-bar");
exports.SegmentedBar = segmented_bar_1.SegmentedBar;
exports.SegmentedBarItem = segmented_bar_1.SegmentedBarItem;
var slider_1 = require("./slider");
exports.Slider = slider_1.Slider;
var style_1 = require("./styling/style");
exports.Style = style_1.Style;
var switch_1 = require("./switch");
exports.Switch = switch_1.Switch;
var tab_content_item_1 = require("./tab-navigation-base/tab-content-item");
exports.TabContentItem = tab_content_item_1.TabContentItem;
var tab_navigation_base_1 = require("./tab-navigation-base/tab-navigation-base");
exports.TabNavigationBase = tab_navigation_base_1.TabNavigationBase;
var tab_strip_1 = require("./tab-navigation-base/tab-strip");
exports.TabStrip = tab_strip_1.TabStrip;
var tab_strip_item_1 = require("./tab-navigation-base/tab-strip-item");
exports.TabStripItem = tab_strip_item_1.TabStripItem;
var tab_view_1 = require("./tab-view");
exports.TabView = tab_view_1.TabView;
exports.TabViewItem = tab_view_1.TabViewItem;
var tabs_1 = require("./tabs");
exports.Tabs = tabs_1.Tabs;
var text_base_1 = require("./text-base");
exports.TextBase = text_base_1.TextBase;
var formatted_string_1 = require("./text-base/formatted-string");
exports.FormattedString = formatted_string_1.FormattedString;
var span_1 = require("./text-base/span");
exports.Span = span_1.Span;
var text_field_1 = require("./text-field");
exports.TextField = text_field_1.TextField;
var text_view_1 = require("./text-view");
exports.TextView = text_view_1.TextView;
var time_picker_1 = require("./time-picker");
exports.TimePicker = time_picker_1.TimePicker;
var transition_1 = require("./transition");
exports.Transition = transition_1.Transition;
var web_view_1 = require("./web-view");
exports.WebView = web_view_1.WebView;
