// kinvey-nativescript-sdk/lib/nativescript/popup/popup.js
/* THIS IS A PLACEHOLDER FILE AND WILL BE REMOVED WHEN NATIVESCRIPT BUILDS */
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var Popup = /** @class */ (function (_super) {
    __extends(Popup, _super);
    function Popup() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Popup.prototype.isClosed = function () {
        return true;
    };
    Popup.prototype.onLoaded = function (listener) {
        return this;
    };
    Popup.prototype.onClosed = function (listener) {
        return this;
    };
    Popup.prototype.onError = function (listener) {
        return this;
    };
    Popup.prototype.close = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    Popup.open = function (url, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Popup does not work.');
            });
        });
    };
    return Popup;
}(events_1.EventEmitter));
exports.Popup = Popup;
